// @trustvc/trustvc/dist/esm/open-attestation/utils.js
import forge from 'node-forge';
import { utils } from '@tradetrust-tt/tradetrust';
export { SUPPORTED_SIGNING_ALGORITHM, SchemaId, getData as getDataV2, isSchemaValidationError, obfuscateDocument, validateSchema } from '@tradetrust-tt/tradetrust';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const ENCRYPTION_PARAMETERS = Object.freeze({
  algorithm: "AES-GCM",
  keyLength: 256,
  // Key length in bits
  ivLength: 96,
  // IV length in bits: NIST suggests 12 bytes
  tagLength: 128,
  // GCM authentication tag length in bits, see link above for explanation
  version: "OPEN-ATTESTATION-TYPE-1"
  // Type 1 using the above params without compression
});
const generateEncryptionKey = /* @__PURE__ */ __name((keyLengthInBits = ENCRYPTION_PARAMETERS.keyLength) => {
  const encryptionKey = forge.random.getBytesSync(keyLengthInBits / 8);
  return forge.util.bytesToHex(encryptionKey);
}, "generateEncryptionKey");
const encodeDocument = /* @__PURE__ */ __name((document) => {
  const bytes = forge.util.encodeUtf8(document);
  return forge.util.encode64(bytes);
}, "encodeDocument");
const decodeDocument = /* @__PURE__ */ __name((encoded) => {
  const decoded = forge.util.decode64(encoded);
  return forge.util.decodeUtf8(decoded);
}, "decodeDocument");
const {
  isTransferableAsset,
  isDocumentRevokable,
  getAssetId,
  isWrappedV2Document,
  isSignedWrappedV2Document,
  isWrappedV3Document,
  isSignedWrappedV3Document,
  isRawV2Document,
  isRawV3Document,
  getDocumentData,
  getIssuerAddress,
  diagnose,
  getTemplateURL
} = utils;

export { ENCRYPTION_PARAMETERS, decodeDocument, diagnose, encodeDocument, generateEncryptionKey, getAssetId, getDocumentData, getIssuerAddress, getTemplateURL, isDocumentRevokable, isRawV2Document, isRawV3Document, isSignedWrappedV2Document, isSignedWrappedV3Document, isTransferableAsset, isWrappedV2Document, isWrappedV3Document };
