// @trustvc/trustvc/dist/esm/deploy/token-registry.js
import { TDocDeployer__factory, TradeTrustToken__factory } from '@tradetrust-tt/token-registry/contracts';
import { getChainIdSafe, getDefaultContractAddress, isValidAddress, getTxOptions, isSupportedTitleEscrowFactory } from '../token-registry-functions/utils';
import { encodeInitParams } from '../token-registry-v5/utils';
import { getEthersContractFromProvider, getEthersContractFactoryFromProvider, isV6EthersProvider } from '../utils/ethers';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const deployTokenRegistry = /* @__PURE__ */ __name(async (registryName, registrySymbol, signer, options = {}) => {
  const { maxFeePerGas, maxPriorityFeePerGas } = options;
  let { chainId, standalone, factoryAddress, tokenRegistryImplAddress, deployerContractAddress } = options;
  const deployerAddress = await signer.getAddress();
  if (!chainId) {
    chainId = await getChainIdSafe(signer);
  }
  const {
    TitleEscrowFactory: defaultTitleEscrowFactoryAddress,
    TokenImplementation: defaultTokenImplementationContractAddress,
    Deployer: defaultDeployerContractAddress
  } = getDefaultContractAddress(chainId);
  if (!isValidAddress(deployerContractAddress)) {
    deployerContractAddress = defaultDeployerContractAddress;
  }
  if (!isValidAddress(tokenRegistryImplAddress)) {
    tokenRegistryImplAddress = defaultTokenImplementationContractAddress;
  }
  if (standalone !== false && (!deployerContractAddress || !tokenRegistryImplAddress)) {
    console.error(
      `Network ${chainId} does not support "quick-start" mode. Defaulting to standalone mode.`
    );
    standalone = true;
  }
  if (!standalone) {
    if (!deployerContractAddress || !tokenRegistryImplAddress) {
      throw new Error(`Network ${chainId} currently is not supported. Use --standalone instead.`);
    }
    const Contract = getEthersContractFromProvider(signer.provider);
    const deployerContract = new Contract(
      deployerContractAddress,
      TDocDeployer__factory.abi,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      signer
      // Type assertion needed for v5/v6 compatibility
    );
    const initParam = encodeInitParams({
      name: registryName,
      symbol: registrySymbol,
      deployer: deployerAddress
    });
    const txOptions = await getTxOptions(signer, chainId, maxFeePerGas, maxPriorityFeePerGas);
    return await deployerContract.deploy(tokenRegistryImplAddress, initParam, txOptions);
  } else {
    if (!factoryAddress || !isValidAddress(factoryAddress)) {
      factoryAddress = defaultTitleEscrowFactoryAddress;
      if (!factoryAddress) {
        throw new Error(`Network ${chainId} currently is not supported. Supply a factory address.`);
      }
    }
    const supportedTitleEscrowFactory = await isSupportedTitleEscrowFactory(
      factoryAddress,
      signer.provider
    );
    if (!supportedTitleEscrowFactory) {
      throw new Error(`Title Escrow Factory ${factoryAddress} is not supported.`);
    }
    const Contract = getEthersContractFactoryFromProvider(signer.provider);
    const tokenFactory = new Contract(
      TradeTrustToken__factory.abi,
      TradeTrustToken__factory.bytecode,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      signer
      // Type assertion needed for v5/v6 compatibility
    );
    const isV6 = isV6EthersProvider(signer.provider);
    const txOptions = await getTxOptions(signer, chainId, maxFeePerGas, maxPriorityFeePerGas);
    if (isV6) {
      const contract = await tokenFactory.deploy(
        registryName,
        registrySymbol,
        factoryAddress,
        txOptions
      );
      return await contract.deploymentTransaction().wait();
    } else {
      const contract = await tokenFactory.deploy(
        registryName,
        registrySymbol,
        factoryAddress,
        txOptions
      );
      return await contract.deployTransaction.wait();
    }
  }
}, "deployTokenRegistry");

export { deployTokenRegistry };
