// @tradetrust-tt/token-registry-v4/dist/contracts/contracts/interfaces/ITitleEscrowSignable.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type {
  FunctionFragment,
  Result,
  EventFragment,
} from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../common";

export type BeneficiaryTransferEndorsementStruct = {
  beneficiary: PromiseOrValue<string>;
  holder: PromiseOrValue<string>;
  nominee: PromiseOrValue<string>;
  registry: PromiseOrValue<string>;
  tokenId: PromiseOrValue<BigNumberish>;
  deadline: PromiseOrValue<BigNumberish>;
  nonce: PromiseOrValue<BigNumberish>;
};

export type BeneficiaryTransferEndorsementStructOutput = [
  string,
  string,
  string,
  string,
  BigNumber,
  BigNumber,
  BigNumber
] & {
  beneficiary: string;
  holder: string;
  nominee: string;
  registry: string;
  tokenId: BigNumber;
  deadline: BigNumber;
  nonce: BigNumber;
};

export declare namespace SigHelper {
  export type SigStruct = {
    r: PromiseOrValue<BytesLike>;
    s: PromiseOrValue<BytesLike>;
    v: PromiseOrValue<BigNumberish>;
  };

  export type SigStructOutput = [string, string, number] & {
    r: string;
    s: string;
    v: number;
  };
}

export interface ITitleEscrowSignableInterface extends utils.Interface {
  functions: {
    "active()": FunctionFragment;
    "beneficiary()": FunctionFragment;
    "cancelBeneficiaryTransfer((address,address,address,address,uint256,uint256,uint256))": FunctionFragment;
    "holder()": FunctionFragment;
    "isHoldingToken()": FunctionFragment;
    "nominate(address)": FunctionFragment;
    "nominee()": FunctionFragment;
    "onERC721Received(address,address,uint256,bytes)": FunctionFragment;
    "registry()": FunctionFragment;
    "shred()": FunctionFragment;
    "surrender()": FunctionFragment;
    "tokenId()": FunctionFragment;
    "transferBeneficiary(address)": FunctionFragment;
    "transferBeneficiaryWithSig((address,address,address,address,uint256,uint256,uint256),(bytes32,bytes32,uint8))": FunctionFragment;
    "transferHolder(address)": FunctionFragment;
    "transferOwners(address,address)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "active"
      | "active()"
      | "beneficiary"
      | "beneficiary()"
      | "cancelBeneficiaryTransfer"
      | "cancelBeneficiaryTransfer((address,address,address,address,uint256,uint256,uint256))"
      | "holder"
      | "holder()"
      | "isHoldingToken"
      | "isHoldingToken()"
      | "nominate"
      | "nominate(address)"
      | "nominee"
      | "nominee()"
      | "onERC721Received"
      | "onERC721Received(address,address,uint256,bytes)"
      | "registry"
      | "registry()"
      | "shred"
      | "shred()"
      | "surrender"
      | "surrender()"
      | "tokenId"
      | "tokenId()"
      | "transferBeneficiary"
      | "transferBeneficiary(address)"
      | "transferBeneficiaryWithSig"
      | "transferBeneficiaryWithSig((address,address,address,address,uint256,uint256,uint256),(bytes32,bytes32,uint8))"
      | "transferHolder"
      | "transferHolder(address)"
      | "transferOwners"
      | "transferOwners(address,address)"
  ): FunctionFragment;

  encodeFunctionData(functionFragment: "active", values?: undefined): string;
  encodeFunctionData(functionFragment: "active()", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "beneficiary",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "beneficiary()",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "cancelBeneficiaryTransfer",
    values: [BeneficiaryTransferEndorsementStruct]
  ): string;
  encodeFunctionData(
    functionFragment: "cancelBeneficiaryTransfer((address,address,address,address,uint256,uint256,uint256))",
    values: [BeneficiaryTransferEndorsementStruct]
  ): string;
  encodeFunctionData(functionFragment: "holder", values?: undefined): string;
  encodeFunctionData(functionFragment: "holder()", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "isHoldingToken",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "isHoldingToken()",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "nominate",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "nominate(address)",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(functionFragment: "nominee", values?: undefined): string;
  encodeFunctionData(functionFragment: "nominee()", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "onERC721Received",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BytesLike>
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "onERC721Received(address,address,uint256,bytes)",
    values: [
      PromiseOrValue<string>,
      PromiseOrValue<string>,
      PromiseOrValue<BigNumberish>,
      PromiseOrValue<BytesLike>
    ]
  ): string;
  encodeFunctionData(functionFragment: "registry", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "registry()",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "shred", values?: undefined): string;
  encodeFunctionData(functionFragment: "shred()", values?: undefined): string;
  encodeFunctionData(functionFragment: "surrender", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "surrender()",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "tokenId", values?: undefined): string;
  encodeFunctionData(functionFragment: "tokenId()", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "transferBeneficiary",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "transferBeneficiary(address)",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "transferBeneficiaryWithSig",
    values: [BeneficiaryTransferEndorsementStruct, SigHelper.SigStruct]
  ): string;
  encodeFunctionData(
    functionFragment: "transferBeneficiaryWithSig((address,address,address,address,uint256,uint256,uint256),(bytes32,bytes32,uint8))",
    values: [BeneficiaryTransferEndorsementStruct, SigHelper.SigStruct]
  ): string;
  encodeFunctionData(
    functionFragment: "transferHolder",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "transferHolder(address)",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "transferOwners",
    values: [PromiseOrValue<string>, PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "transferOwners(address,address)",
    values: [PromiseOrValue<string>, PromiseOrValue<string>]
  ): string;

  decodeFunctionResult(functionFragment: "active", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "active()", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "beneficiary",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "beneficiary()",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "cancelBeneficiaryTransfer",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "cancelBeneficiaryTransfer((address,address,address,address,uint256,uint256,uint256))",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "holder", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "holder()", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "isHoldingToken",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "isHoldingToken()",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "nominate", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "nominate(address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "nominee", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "nominee()", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "onERC721Received",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "onERC721Received(address,address,uint256,bytes)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "registry", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "registry()", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "shred", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "shred()", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "surrender", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "surrender()",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "tokenId", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "tokenId()", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "transferBeneficiary",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferBeneficiary(address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferBeneficiaryWithSig",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferBeneficiaryWithSig((address,address,address,address,uint256,uint256,uint256),(bytes32,bytes32,uint8))",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferHolder",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferHolder(address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferOwners",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferOwners(address,address)",
    data: BytesLike
  ): Result;

  events: {
    "BeneficiaryTransfer(address,address,address,uint256)": EventFragment;
    "CancelBeneficiaryTransferEndorsement(bytes32,address,uint256)": EventFragment;
    "HolderTransfer(address,address,address,uint256)": EventFragment;
    "Nomination(address,address,address,uint256)": EventFragment;
    "Shred(address,uint256)": EventFragment;
    "Surrender(address,address,uint256)": EventFragment;
    "TokenReceived(address,address,bool,address,uint256)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "BeneficiaryTransfer"): EventFragment;
  getEvent(
    nameOrSignatureOrTopic: "BeneficiaryTransfer(address,address,address,uint256)"
  ): EventFragment;
  getEvent(
    nameOrSignatureOrTopic: "CancelBeneficiaryTransferEndorsement"
  ): EventFragment;
  getEvent(
    nameOrSignatureOrTopic: "CancelBeneficiaryTransferEndorsement(bytes32,address,uint256)"
  ): EventFragment;
  getEvent(nameOrSignatureOrTopic: "HolderTransfer"): EventFragment;
  getEvent(
    nameOrSignatureOrTopic: "HolderTransfer(address,address,address,uint256)"
  ): EventFragment;
  getEvent(nameOrSignatureOrTopic: "Nomination"): EventFragment;
  getEvent(
    nameOrSignatureOrTopic: "Nomination(address,address,address,uint256)"
  ): EventFragment;
  getEvent(nameOrSignatureOrTopic: "Shred"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "Shred(address,uint256)"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "Surrender"): EventFragment;
  getEvent(
    nameOrSignatureOrTopic: "Surrender(address,address,uint256)"
  ): EventFragment;
  getEvent(nameOrSignatureOrTopic: "TokenReceived"): EventFragment;
  getEvent(
    nameOrSignatureOrTopic: "TokenReceived(address,address,bool,address,uint256)"
  ): EventFragment;
}

export interface BeneficiaryTransferEventObject {
  fromBeneficiary: string;
  toBeneficiary: string;
  registry: string;
  tokenId: BigNumber;
}
export type BeneficiaryTransferEvent = TypedEvent<
  [string, string, string, BigNumber],
  BeneficiaryTransferEventObject
>;

export type BeneficiaryTransferEventFilter =
  TypedEventFilter<BeneficiaryTransferEvent>;

export interface CancelBeneficiaryTransferEndorsementEventObject {
  hash: string;
  endorser: string;
  tokenId: BigNumber;
}
export type CancelBeneficiaryTransferEndorsementEvent = TypedEvent<
  [string, string, BigNumber],
  CancelBeneficiaryTransferEndorsementEventObject
>;

export type CancelBeneficiaryTransferEndorsementEventFilter =
  TypedEventFilter<CancelBeneficiaryTransferEndorsementEvent>;

export interface HolderTransferEventObject {
  fromHolder: string;
  toHolder: string;
  registry: string;
  tokenId: BigNumber;
}
export type HolderTransferEvent = TypedEvent<
  [string, string, string, BigNumber],
  HolderTransferEventObject
>;

export type HolderTransferEventFilter = TypedEventFilter<HolderTransferEvent>;

export interface NominationEventObject {
  prevNominee: string;
  nominee: string;
  registry: string;
  tokenId: BigNumber;
}
export type NominationEvent = TypedEvent<
  [string, string, string, BigNumber],
  NominationEventObject
>;

export type NominationEventFilter = TypedEventFilter<NominationEvent>;

export interface ShredEventObject {
  registry: string;
  tokenId: BigNumber;
}
export type ShredEvent = TypedEvent<[string, BigNumber], ShredEventObject>;

export type ShredEventFilter = TypedEventFilter<ShredEvent>;

export interface SurrenderEventObject {
  surrenderer: string;
  registry: string;
  tokenId: BigNumber;
}
export type SurrenderEvent = TypedEvent<
  [string, string, BigNumber],
  SurrenderEventObject
>;

export type SurrenderEventFilter = TypedEventFilter<SurrenderEvent>;

export interface TokenReceivedEventObject {
  beneficiary: string;
  holder: string;
  isMinting: boolean;
  registry: string;
  tokenId: BigNumber;
}
export type TokenReceivedEvent = TypedEvent<
  [string, string, boolean, string, BigNumber],
  TokenReceivedEventObject
>;

export type TokenReceivedEventFilter = TypedEventFilter<TokenReceivedEvent>;

export interface ITitleEscrowSignable extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: ITitleEscrowSignableInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    active(overrides?: CallOverrides): Promise<[boolean]>;

    "active()"(overrides?: CallOverrides): Promise<[boolean]>;

    beneficiary(overrides?: CallOverrides): Promise<[string]>;

    "beneficiary()"(overrides?: CallOverrides): Promise<[string]>;

    cancelBeneficiaryTransfer(
      endorsement: BeneficiaryTransferEndorsementStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    "cancelBeneficiaryTransfer((address,address,address,address,uint256,uint256,uint256))"(
      endorsement: BeneficiaryTransferEndorsementStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    holder(overrides?: CallOverrides): Promise<[string]>;

    "holder()"(overrides?: CallOverrides): Promise<[string]>;

    isHoldingToken(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    "isHoldingToken()"(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    nominate(
      nominee: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    "nominate(address)"(
      nominee: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    nominee(overrides?: CallOverrides): Promise<[string]>;

    "nominee()"(overrides?: CallOverrides): Promise<[string]>;

    onERC721Received(
      operator: PromiseOrValue<string>,
      from: PromiseOrValue<string>,
      tokenId: PromiseOrValue<BigNumberish>,
      data: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    "onERC721Received(address,address,uint256,bytes)"(
      operator: PromiseOrValue<string>,
      from: PromiseOrValue<string>,
      tokenId: PromiseOrValue<BigNumberish>,
      data: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    registry(overrides?: CallOverrides): Promise<[string]>;

    "registry()"(overrides?: CallOverrides): Promise<[string]>;

    shred(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    "shred()"(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    surrender(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    "surrender()"(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    tokenId(overrides?: CallOverrides): Promise<[BigNumber]>;

    "tokenId()"(overrides?: CallOverrides): Promise<[BigNumber]>;

    transferBeneficiary(
      nominee: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    "transferBeneficiary(address)"(
      nominee: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    transferBeneficiaryWithSig(
      endorsement: BeneficiaryTransferEndorsementStruct,
      sig: SigHelper.SigStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    "transferBeneficiaryWithSig((address,address,address,address,uint256,uint256,uint256),(bytes32,bytes32,uint8))"(
      endorsement: BeneficiaryTransferEndorsementStruct,
      sig: SigHelper.SigStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    transferHolder(
      newHolder: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    "transferHolder(address)"(
      newHolder: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    transferOwners(
      nominee: PromiseOrValue<string>,
      newHolder: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    "transferOwners(address,address)"(
      nominee: PromiseOrValue<string>,
      newHolder: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  active(overrides?: CallOverrides): Promise<boolean>;

  "active()"(overrides?: CallOverrides): Promise<boolean>;

  beneficiary(overrides?: CallOverrides): Promise<string>;

  "beneficiary()"(overrides?: CallOverrides): Promise<string>;

  cancelBeneficiaryTransfer(
    endorsement: BeneficiaryTransferEndorsementStruct,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  "cancelBeneficiaryTransfer((address,address,address,address,uint256,uint256,uint256))"(
    endorsement: BeneficiaryTransferEndorsementStruct,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  holder(overrides?: CallOverrides): Promise<string>;

  "holder()"(overrides?: CallOverrides): Promise<string>;

  isHoldingToken(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  "isHoldingToken()"(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  nominate(
    nominee: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  "nominate(address)"(
    nominee: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  nominee(overrides?: CallOverrides): Promise<string>;

  "nominee()"(overrides?: CallOverrides): Promise<string>;

  onERC721Received(
    operator: PromiseOrValue<string>,
    from: PromiseOrValue<string>,
    tokenId: PromiseOrValue<BigNumberish>,
    data: PromiseOrValue<BytesLike>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  "onERC721Received(address,address,uint256,bytes)"(
    operator: PromiseOrValue<string>,
    from: PromiseOrValue<string>,
    tokenId: PromiseOrValue<BigNumberish>,
    data: PromiseOrValue<BytesLike>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  registry(overrides?: CallOverrides): Promise<string>;

  "registry()"(overrides?: CallOverrides): Promise<string>;

  shred(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  "shred()"(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  surrender(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  "surrender()"(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  tokenId(overrides?: CallOverrides): Promise<BigNumber>;

  "tokenId()"(overrides?: CallOverrides): Promise<BigNumber>;

  transferBeneficiary(
    nominee: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  "transferBeneficiary(address)"(
    nominee: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  transferBeneficiaryWithSig(
    endorsement: BeneficiaryTransferEndorsementStruct,
    sig: SigHelper.SigStruct,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  "transferBeneficiaryWithSig((address,address,address,address,uint256,uint256,uint256),(bytes32,bytes32,uint8))"(
    endorsement: BeneficiaryTransferEndorsementStruct,
    sig: SigHelper.SigStruct,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  transferHolder(
    newHolder: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  "transferHolder(address)"(
    newHolder: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  transferOwners(
    nominee: PromiseOrValue<string>,
    newHolder: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  "transferOwners(address,address)"(
    nominee: PromiseOrValue<string>,
    newHolder: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    active(overrides?: CallOverrides): Promise<boolean>;

    "active()"(overrides?: CallOverrides): Promise<boolean>;

    beneficiary(overrides?: CallOverrides): Promise<string>;

    "beneficiary()"(overrides?: CallOverrides): Promise<string>;

    cancelBeneficiaryTransfer(
      endorsement: BeneficiaryTransferEndorsementStruct,
      overrides?: CallOverrides
    ): Promise<void>;

    "cancelBeneficiaryTransfer((address,address,address,address,uint256,uint256,uint256))"(
      endorsement: BeneficiaryTransferEndorsementStruct,
      overrides?: CallOverrides
    ): Promise<void>;

    holder(overrides?: CallOverrides): Promise<string>;

    "holder()"(overrides?: CallOverrides): Promise<string>;

    isHoldingToken(overrides?: CallOverrides): Promise<boolean>;

    "isHoldingToken()"(overrides?: CallOverrides): Promise<boolean>;

    nominate(
      nominee: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;

    "nominate(address)"(
      nominee: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;

    nominee(overrides?: CallOverrides): Promise<string>;

    "nominee()"(overrides?: CallOverrides): Promise<string>;

    onERC721Received(
      operator: PromiseOrValue<string>,
      from: PromiseOrValue<string>,
      tokenId: PromiseOrValue<BigNumberish>,
      data: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<string>;

    "onERC721Received(address,address,uint256,bytes)"(
      operator: PromiseOrValue<string>,
      from: PromiseOrValue<string>,
      tokenId: PromiseOrValue<BigNumberish>,
      data: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<string>;

    registry(overrides?: CallOverrides): Promise<string>;

    "registry()"(overrides?: CallOverrides): Promise<string>;

    shred(overrides?: CallOverrides): Promise<void>;

    "shred()"(overrides?: CallOverrides): Promise<void>;

    surrender(overrides?: CallOverrides): Promise<void>;

    "surrender()"(overrides?: CallOverrides): Promise<void>;

    tokenId(overrides?: CallOverrides): Promise<BigNumber>;

    "tokenId()"(overrides?: CallOverrides): Promise<BigNumber>;

    transferBeneficiary(
      nominee: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;

    "transferBeneficiary(address)"(
      nominee: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;

    transferBeneficiaryWithSig(
      endorsement: BeneficiaryTransferEndorsementStruct,
      sig: SigHelper.SigStruct,
      overrides?: CallOverrides
    ): Promise<void>;

    "transferBeneficiaryWithSig((address,address,address,address,uint256,uint256,uint256),(bytes32,bytes32,uint8))"(
      endorsement: BeneficiaryTransferEndorsementStruct,
      sig: SigHelper.SigStruct,
      overrides?: CallOverrides
    ): Promise<void>;

    transferHolder(
      newHolder: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;

    "transferHolder(address)"(
      newHolder: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;

    transferOwners(
      nominee: PromiseOrValue<string>,
      newHolder: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;

    "transferOwners(address,address)"(
      nominee: PromiseOrValue<string>,
      newHolder: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {
    "BeneficiaryTransfer(address,address,address,uint256)"(
      fromBeneficiary?: PromiseOrValue<string> | null,
      toBeneficiary?: PromiseOrValue<string> | null,
      registry?: null,
      tokenId?: null
    ): BeneficiaryTransferEventFilter;
    BeneficiaryTransfer(
      fromBeneficiary?: PromiseOrValue<string> | null,
      toBeneficiary?: PromiseOrValue<string> | null,
      registry?: null,
      tokenId?: null
    ): BeneficiaryTransferEventFilter;

    "CancelBeneficiaryTransferEndorsement(bytes32,address,uint256)"(
      hash?: PromiseOrValue<BytesLike> | null,
      endorser?: PromiseOrValue<string> | null,
      tokenId?: PromiseOrValue<BigNumberish> | null
    ): CancelBeneficiaryTransferEndorsementEventFilter;
    CancelBeneficiaryTransferEndorsement(
      hash?: PromiseOrValue<BytesLike> | null,
      endorser?: PromiseOrValue<string> | null,
      tokenId?: PromiseOrValue<BigNumberish> | null
    ): CancelBeneficiaryTransferEndorsementEventFilter;

    "HolderTransfer(address,address,address,uint256)"(
      fromHolder?: PromiseOrValue<string> | null,
      toHolder?: PromiseOrValue<string> | null,
      registry?: null,
      tokenId?: null
    ): HolderTransferEventFilter;
    HolderTransfer(
      fromHolder?: PromiseOrValue<string> | null,
      toHolder?: PromiseOrValue<string> | null,
      registry?: null,
      tokenId?: null
    ): HolderTransferEventFilter;

    "Nomination(address,address,address,uint256)"(
      prevNominee?: PromiseOrValue<string> | null,
      nominee?: PromiseOrValue<string> | null,
      registry?: null,
      tokenId?: null
    ): NominationEventFilter;
    Nomination(
      prevNominee?: PromiseOrValue<string> | null,
      nominee?: PromiseOrValue<string> | null,
      registry?: null,
      tokenId?: null
    ): NominationEventFilter;

    "Shred(address,uint256)"(registry?: null, tokenId?: null): ShredEventFilter;
    Shred(registry?: null, tokenId?: null): ShredEventFilter;

    "Surrender(address,address,uint256)"(
      surrenderer?: PromiseOrValue<string> | null,
      registry?: null,
      tokenId?: null
    ): SurrenderEventFilter;
    Surrender(
      surrenderer?: PromiseOrValue<string> | null,
      registry?: null,
      tokenId?: null
    ): SurrenderEventFilter;

    "TokenReceived(address,address,bool,address,uint256)"(
      beneficiary?: PromiseOrValue<string> | null,
      holder?: PromiseOrValue<string> | null,
      isMinting?: PromiseOrValue<boolean> | null,
      registry?: null,
      tokenId?: null
    ): TokenReceivedEventFilter;
    TokenReceived(
      beneficiary?: PromiseOrValue<string> | null,
      holder?: PromiseOrValue<string> | null,
      isMinting?: PromiseOrValue<boolean> | null,
      registry?: null,
      tokenId?: null
    ): TokenReceivedEventFilter;
  };

  estimateGas: {
    active(overrides?: CallOverrides): Promise<BigNumber>;

    "active()"(overrides?: CallOverrides): Promise<BigNumber>;

    beneficiary(overrides?: CallOverrides): Promise<BigNumber>;

    "beneficiary()"(overrides?: CallOverrides): Promise<BigNumber>;

    cancelBeneficiaryTransfer(
      endorsement: BeneficiaryTransferEndorsementStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    "cancelBeneficiaryTransfer((address,address,address,address,uint256,uint256,uint256))"(
      endorsement: BeneficiaryTransferEndorsementStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    holder(overrides?: CallOverrides): Promise<BigNumber>;

    "holder()"(overrides?: CallOverrides): Promise<BigNumber>;

    isHoldingToken(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    "isHoldingToken()"(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    nominate(
      nominee: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    "nominate(address)"(
      nominee: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    nominee(overrides?: CallOverrides): Promise<BigNumber>;

    "nominee()"(overrides?: CallOverrides): Promise<BigNumber>;

    onERC721Received(
      operator: PromiseOrValue<string>,
      from: PromiseOrValue<string>,
      tokenId: PromiseOrValue<BigNumberish>,
      data: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    "onERC721Received(address,address,uint256,bytes)"(
      operator: PromiseOrValue<string>,
      from: PromiseOrValue<string>,
      tokenId: PromiseOrValue<BigNumberish>,
      data: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    registry(overrides?: CallOverrides): Promise<BigNumber>;

    "registry()"(overrides?: CallOverrides): Promise<BigNumber>;

    shred(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    "shred()"(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    surrender(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    "surrender()"(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    tokenId(overrides?: CallOverrides): Promise<BigNumber>;

    "tokenId()"(overrides?: CallOverrides): Promise<BigNumber>;

    transferBeneficiary(
      nominee: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    "transferBeneficiary(address)"(
      nominee: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    transferBeneficiaryWithSig(
      endorsement: BeneficiaryTransferEndorsementStruct,
      sig: SigHelper.SigStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    "transferBeneficiaryWithSig((address,address,address,address,uint256,uint256,uint256),(bytes32,bytes32,uint8))"(
      endorsement: BeneficiaryTransferEndorsementStruct,
      sig: SigHelper.SigStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    transferHolder(
      newHolder: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    "transferHolder(address)"(
      newHolder: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    transferOwners(
      nominee: PromiseOrValue<string>,
      newHolder: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    "transferOwners(address,address)"(
      nominee: PromiseOrValue<string>,
      newHolder: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    active(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    "active()"(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    beneficiary(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    "beneficiary()"(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    cancelBeneficiaryTransfer(
      endorsement: BeneficiaryTransferEndorsementStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    "cancelBeneficiaryTransfer((address,address,address,address,uint256,uint256,uint256))"(
      endorsement: BeneficiaryTransferEndorsementStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    holder(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    "holder()"(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    isHoldingToken(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    "isHoldingToken()"(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    nominate(
      nominee: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    "nominate(address)"(
      nominee: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    nominee(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    "nominee()"(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    onERC721Received(
      operator: PromiseOrValue<string>,
      from: PromiseOrValue<string>,
      tokenId: PromiseOrValue<BigNumberish>,
      data: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    "onERC721Received(address,address,uint256,bytes)"(
      operator: PromiseOrValue<string>,
      from: PromiseOrValue<string>,
      tokenId: PromiseOrValue<BigNumberish>,
      data: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    registry(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    "registry()"(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    shred(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    "shred()"(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    surrender(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    "surrender()"(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    tokenId(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    "tokenId()"(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    transferBeneficiary(
      nominee: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    "transferBeneficiary(address)"(
      nominee: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    transferBeneficiaryWithSig(
      endorsement: BeneficiaryTransferEndorsementStruct,
      sig: SigHelper.SigStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    "transferBeneficiaryWithSig((address,address,address,address,uint256,uint256,uint256),(bytes32,bytes32,uint8))"(
      endorsement: BeneficiaryTransferEndorsementStruct,
      sig: SigHelper.SigStruct,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    transferHolder(
      newHolder: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    "transferHolder(address)"(
      newHolder: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    transferOwners(
      nominee: PromiseOrValue<string>,
      newHolder: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    "transferOwners(address,address)"(
      nominee: PromiseOrValue<string>,
      newHolder: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
