// @trustvc/trustvc/dist/cjs/token-registry-functions/utils.js
'use strict';

var ethers = require('../utils/ethers');
var utils = require('../utils');
var ethersV6 = require('ethersV6');
var tokenRegistryV5 = require('../token-registry-v5');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const getTxOptions = /* @__PURE__ */ __name(async (signer, chainId, maxFeePerGas, maxPriorityFeePerGas) => {
  if (!maxFeePerGas || !maxPriorityFeePerGas) {
    chainId = chainId ?? await getChainIdSafe(signer);
    const gasStation = utils.SUPPORTED_CHAINS[chainId]?.gasStation;
    if (gasStation) {
      const gasFees = await gasStation();
      maxFeePerGas = gasFees?.maxFeePerGas ?? 0;
      maxPriorityFeePerGas = gasFees?.maxPriorityFeePerGas ?? 0;
    }
  }
  return maxFeePerGas && maxPriorityFeePerGas ? { maxFeePerGas, maxPriorityFeePerGas } : {};
}, "getTxOptions");
const getChainIdSafe = /* @__PURE__ */ __name(async (signer) => {
  if (ethers.isV6EthersProvider(signer.provider)) {
    const network = await signer.provider?.getNetwork();
    if (!network?.chainId) throw new Error("Cannot determine chainId: provider is missing");
    return network.chainId;
  }
  return await signer.getChainId();
}, "getChainIdSafe");
const getSignerAddressSafe = /* @__PURE__ */ __name(async (signer) => {
  if (ethers.isV6EthersProvider(signer.provider)) {
    return await signer.getAddress();
  }
  return await signer.getAddress();
}, "getSignerAddressSafe");
const { contractInterfaceId: CONTRACT_INTERFACE_ID, contractAddress: CONTRACT_ADDRESS } = tokenRegistryV5.constants;
const getDefaultContractAddress = /* @__PURE__ */ __name((chainId) => {
  const { TitleEscrowFactory, TokenImplementation, Deployer } = CONTRACT_ADDRESS;
  const chainTitleEscrowFactory = TitleEscrowFactory[chainId];
  const chainTokenImplementation = TokenImplementation[chainId];
  const chainDeployer = Deployer[chainId];
  return {
    TitleEscrowFactory: chainTitleEscrowFactory,
    TokenImplementation: chainTokenImplementation,
    Deployer: chainDeployer
  };
}, "getDefaultContractAddress");
const isValidAddress = /* @__PURE__ */ __name((address) => {
  if (!address) return false;
  return ethersV6.isAddress(address);
}, "isValidAddress");
const isSupportedTitleEscrowFactory = /* @__PURE__ */ __name(async (factoryAddress, provider) => {
  const Contract = ethers.getEthersContractFromProvider(provider);
  const titleEscrowFactoryContract = new Contract(
    factoryAddress,
    ["function implementation() view returns (address)"],
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    provider
  );
  const implAddr = await titleEscrowFactoryContract.implementation();
  const implContract = new Contract(
    implAddr,
    ["function supportsInterface(bytes4 interfaceId) view returns (bool)"],
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    provider
  );
  const { TitleEscrow: titleEscrowInterfaceId } = CONTRACT_INTERFACE_ID;
  return implContract.supportsInterface(titleEscrowInterfaceId);
}, "isSupportedTitleEscrowFactory");

exports.getChainIdSafe = getChainIdSafe;
exports.getDefaultContractAddress = getDefaultContractAddress;
exports.getSignerAddressSafe = getSignerAddressSafe;
exports.getTxOptions = getTxOptions;
exports.isSupportedTitleEscrowFactory = isSupportedTitleEscrowFactory;
exports.isValidAddress = isValidAddress;
