// @trustvc/trustvc/dist/cjs/open-attestation/utils.js
'use strict';

var forge = require('node-forge');
var tradetrust = require('@tradetrust-tt/tradetrust');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var forge__default = /*#__PURE__*/_interopDefault(forge);

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const ENCRYPTION_PARAMETERS = Object.freeze({
  algorithm: "AES-GCM",
  keyLength: 256,
  // Key length in bits
  ivLength: 96,
  // IV length in bits: NIST suggests 12 bytes
  tagLength: 128,
  // GCM authentication tag length in bits, see link above for explanation
  version: "OPEN-ATTESTATION-TYPE-1"
  // Type 1 using the above params without compression
});
const generateEncryptionKey = /* @__PURE__ */ __name((keyLengthInBits = ENCRYPTION_PARAMETERS.keyLength) => {
  const encryptionKey = forge__default.default.random.getBytesSync(keyLengthInBits / 8);
  return forge__default.default.util.bytesToHex(encryptionKey);
}, "generateEncryptionKey");
const encodeDocument = /* @__PURE__ */ __name((document) => {
  const bytes = forge__default.default.util.encodeUtf8(document);
  return forge__default.default.util.encode64(bytes);
}, "encodeDocument");
const decodeDocument = /* @__PURE__ */ __name((encoded) => {
  const decoded = forge__default.default.util.decode64(encoded);
  return forge__default.default.util.decodeUtf8(decoded);
}, "decodeDocument");
const {
  isTransferableAsset,
  isDocumentRevokable,
  getAssetId,
  isWrappedV2Document,
  isSignedWrappedV2Document,
  isWrappedV3Document,
  isSignedWrappedV3Document,
  isRawV2Document,
  isRawV3Document,
  getDocumentData,
  getIssuerAddress,
  diagnose,
  getTemplateURL
} = tradetrust.utils;

Object.defineProperty(exports, "SUPPORTED_SIGNING_ALGORITHM", {
  enumerable: true,
  get: function () { return tradetrust.SUPPORTED_SIGNING_ALGORITHM; }
});
Object.defineProperty(exports, "SchemaId", {
  enumerable: true,
  get: function () { return tradetrust.SchemaId; }
});
Object.defineProperty(exports, "getDataV2", {
  enumerable: true,
  get: function () { return tradetrust.getData; }
});
Object.defineProperty(exports, "isSchemaValidationError", {
  enumerable: true,
  get: function () { return tradetrust.isSchemaValidationError; }
});
Object.defineProperty(exports, "obfuscateDocument", {
  enumerable: true,
  get: function () { return tradetrust.obfuscateDocument; }
});
Object.defineProperty(exports, "validateSchema", {
  enumerable: true,
  get: function () { return tradetrust.validateSchema; }
});
exports.ENCRYPTION_PARAMETERS = ENCRYPTION_PARAMETERS;
exports.decodeDocument = decodeDocument;
exports.diagnose = diagnose;
exports.encodeDocument = encodeDocument;
exports.generateEncryptionKey = generateEncryptionKey;
exports.getAssetId = getAssetId;
exports.getDocumentData = getDocumentData;
exports.getIssuerAddress = getIssuerAddress;
exports.getTemplateURL = getTemplateURL;
exports.isDocumentRevokable = isDocumentRevokable;
exports.isRawV2Document = isRawV2Document;
exports.isRawV3Document = isRawV3Document;
exports.isSignedWrappedV2Document = isSignedWrappedV2Document;
exports.isSignedWrappedV3Document = isSignedWrappedV3Document;
exports.isTransferableAsset = isTransferableAsset;
exports.isWrappedV2Document = isWrappedV2Document;
exports.isWrappedV3Document = isWrappedV3Document;
