// @trustvc/trustvc/dist/cjs/open-attestation/wrap.js
'use strict';

var tradetrust = require('@tradetrust-tt/tradetrust');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const wrapOADocumentV2 = /* @__PURE__ */ __name(async (document) => {
  return tradetrust.wrapDocument(document);
}, "wrapOADocumentV2");
const wrapOADocumentsV2 = /* @__PURE__ */ __name(async (documents) => {
  return tradetrust.wrapDocuments(documents);
}, "wrapOADocumentsV2");
async function wrapOADocument(document) {
  if (tradetrust.utils.isRawV2Document(document)) {
    return wrapOADocumentV2(document);
  } else if (tradetrust.utils.isRawV3Document(document)) {
    return tradetrust.__unsafe__use__it__at__your__own__risks__wrapDocument(document);
  } else {
    throw new Error("Unsupported document version");
  }
}
__name(wrapOADocument, "wrapOADocument");
async function wrapOADocuments(documents) {
  if (documents.every((s) => tradetrust.utils.isRawV2Document(s))) {
    return wrapOADocumentsV2(documents);
  } else if (documents.every((s) => tradetrust.utils.isRawV3Document(s))) {
    return tradetrust.__unsafe__use__it__at__your__own__risks__wrapDocuments(documents);
  } else {
    throw new Error("Unsupported documents version");
  }
}
__name(wrapOADocuments, "wrapOADocuments");

Object.defineProperty(exports, "wrapOADocumentV3", {
  enumerable: true,
  get: function () { return tradetrust.__unsafe__use__it__at__your__own__risks__wrapDocument; }
});
Object.defineProperty(exports, "wrapOADocumentsV3", {
  enumerable: true,
  get: function () { return tradetrust.__unsafe__use__it__at__your__own__risks__wrapDocuments; }
});
exports.wrapOADocument = wrapOADocument;
exports.wrapOADocumentV2 = wrapOADocumentV2;
exports.wrapOADocuments = wrapOADocuments;
exports.wrapOADocumentsV2 = wrapOADocumentsV2;
