// @trustvc/trustvc/dist/cjs/verify/fragments/index.js
'use strict';

var ttVerify = require('@tradetrust-tt/tt-verify');
var w3cSignatureIntegrity = require('./document-integrity/w3cSignatureIntegrity');
var ecdsaW3CSignatureIntegrity = require('./document-integrity/ecdsaW3CSignatureIntegrity');
var transferableRecordVerifier = require('./document-status/transferableRecords/transferableRecordVerifier');
var w3cCredentialStatus = require('./document-status/w3cCredentialStatus');
var w3cIssuerIdentity = require('./issuer-identity/w3cIssuerIdentity');
var w3cEmptyCredentialStatus = require('./document-status/w3cEmptyCredentialStatus');



Object.defineProperty(exports, "openAttestationDidSignedDocumentStatus", {
  enumerable: true,
  get: function () { return ttVerify.openAttestationDidSignedDocumentStatus; }
});
Object.defineProperty(exports, "openAttestationDnsDidIdentityProof", {
  enumerable: true,
  get: function () { return ttVerify.openAttestationDnsDidIdentityProof; }
});
Object.defineProperty(exports, "openAttestationDnsTxtIdentityProof", {
  enumerable: true,
  get: function () { return ttVerify.openAttestationDnsTxtIdentityProof; }
});
Object.defineProperty(exports, "openAttestationEthereumDocumentStoreStatus", {
  enumerable: true,
  get: function () { return ttVerify.openAttestationEthereumDocumentStoreStatus; }
});
Object.defineProperty(exports, "openAttestationEthereumTokenRegistryStatus", {
  enumerable: true,
  get: function () { return ttVerify.openAttestationEthereumTokenRegistryStatus; }
});
Object.defineProperty(exports, "openAttestationHash", {
  enumerable: true,
  get: function () { return ttVerify.openAttestationHash; }
});
Object.defineProperty(exports, "w3cSignatureIntegrity", {
  enumerable: true,
  get: function () { return w3cSignatureIntegrity.w3cSignatureIntegrity; }
});
Object.defineProperty(exports, "ecdsaW3CSignatureIntegrity", {
  enumerable: true,
  get: function () { return ecdsaW3CSignatureIntegrity.ecdsaW3CSignatureIntegrity; }
});
Object.defineProperty(exports, "TRANSFERABLE_RECORDS_TYPE", {
  enumerable: true,
  get: function () { return transferableRecordVerifier.TRANSFERABLE_RECORDS_TYPE; }
});
Object.defineProperty(exports, "credentialStatusTransferableRecordVerifier", {
  enumerable: true,
  get: function () { return transferableRecordVerifier.credentialStatusTransferableRecordVerifier; }
});
Object.defineProperty(exports, "w3cCredentialStatus", {
  enumerable: true,
  get: function () { return w3cCredentialStatus.w3cCredentialStatus; }
});
Object.defineProperty(exports, "w3cIssuerIdentity", {
  enumerable: true,
  get: function () { return w3cIssuerIdentity.w3cIssuerIdentity; }
});
Object.defineProperty(exports, "w3cEmptyCredentialStatus", {
  enumerable: true,
  get: function () { return w3cEmptyCredentialStatus.w3cEmptyCredentialStatus; }
});
