// @tradetrust-tt/tt-verify/dist/cjs/verifiers/documentIntegrity/hash/openAttestationHash.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.openAttestationHash = void 0;
const tradetrust_1 = require("@tradetrust-tt/tradetrust");
const error_1 = require("../../../types/error");
const errorHandler_1 = require("../../../common/errorHandler");
const name = "OpenAttestationHash";
const type = "DOCUMENT_INTEGRITY";
const skip = () => {
    return Promise.resolve({
        status: "SKIPPED",
        type,
        name,
        reason: {
            code: error_1.OpenAttestationHashCode.SKIPPED,
            codeString: error_1.OpenAttestationHashCode[error_1.OpenAttestationHashCode.SKIPPED],
            message: `Document does not have merkle root, target hash or data.`,
        },
    });
};
const test = (document) => {
    return tradetrust_1.utils.isWrappedV3Document(document) || tradetrust_1.utils.isWrappedV2Document(document);
};
const verify = (document) => __awaiter(void 0, void 0, void 0, function* () {
    const hash = yield (0, tradetrust_1.verifySignature)(document);
    if (!hash) {
        return {
            type,
            name,
            data: hash,
            reason: {
                code: error_1.OpenAttestationHashCode.DOCUMENT_TAMPERED,
                codeString: error_1.OpenAttestationHashCode[error_1.OpenAttestationHashCode.DOCUMENT_TAMPERED],
                message: "Document has been tampered with",
            },
            status: "INVALID",
        };
    }
    return {
        type,
        name,
        data: hash,
        status: "VALID",
    };
});
exports.openAttestationHash = {
    skip,
    test,
    verify: (0, errorHandler_1.withCodedErrorHandler)(verify, {
        name,
        type,
        unexpectedErrorCode: error_1.OpenAttestationHashCode.UNEXPECTED_ERROR,
        unexpectedErrorString: error_1.OpenAttestationHashCode[error_1.OpenAttestationHashCode.UNEXPECTED_ERROR],
    }),
};
