// @trustvc/trustvc/dist/esm/open-cert/verify.js
import fetch from 'node-fetch';
import { Record, Optional, Boolean, String, Literal, Array, Union } from 'runtypes';
import { Reason, utils, CodedError, isValid, verificationBuilder, openAttestationVerifiers } from '@tradetrust-tt/tt-verify';
import { utils as utils$1, getData, v3 } from '@tradetrust-tt/tradetrust';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const RegistryEntry = Record({
  name: String,
  displayCard: Boolean,
  website: Optional(String),
  email: Optional(String),
  phone: Optional(String),
  logo: Optional(String),
  id: Optional(String)
});
const OpencertsRegistryVerificationValidData = RegistryEntry.And(
  Record({
    value: String,
    status: Literal("VALID")
  })
);
const OpencertsRegistryVerificationInvalidData = Record({
  value: String,
  status: Literal("SKIPPED"),
  reason: Reason
});
const OpencertsRegistryVerificationValidDataArray = Array(
  Union(OpencertsRegistryVerificationValidData, OpencertsRegistryVerificationInvalidData)
);
const OpencertsRegistryVerificationInvalidDataArray = Array(
  OpencertsRegistryVerificationInvalidData
);
const type = "ISSUER_IDENTITY";
const name = "OpencertsRegistryVerifier";
var OpencertsRegistryCode = /* @__PURE__ */ ((OpencertsRegistryCode2) => {
  OpencertsRegistryCode2[OpencertsRegistryCode2["INVALID_IDENTITY"] = 0] = "INVALID_IDENTITY";
  OpencertsRegistryCode2[OpencertsRegistryCode2["SKIPPED"] = 1] = "SKIPPED";
  OpencertsRegistryCode2[OpencertsRegistryCode2["UNEXPECTED_ERROR"] = 2] = "UNEXPECTED_ERROR";
  return OpencertsRegistryCode2;
})(OpencertsRegistryCode || {});
const storeToData = /* @__PURE__ */ __name((registry, store) => {
  const key = Object.keys(registry.issuers).find((k) => k.toLowerCase() === store.toLowerCase());
  if (key) {
    return {
      status: "VALID",
      value: store,
      ...registry.issuers[key]
    };
  }
  return {
    value: store,
    status: "SKIPPED",
    reason: {
      code: 0 /* INVALID_IDENTITY */,
      codeString: OpencertsRegistryCode[0 /* INVALID_IDENTITY */],
      message: `Document store ${store} not found in the registry`
    }
  };
}, "storeToData");
const registryVerifier = {
  test: /* @__PURE__ */ __name((document) => {
    if (utils$1.isWrappedV3Document(document)) {
      return document.openAttestationMetadata.proof.method === v3.Method.DocumentStore;
    } else if (utils$1.isWrappedV2Document(document)) {
      const documentData = getData(document);
      return documentData.issuers.some(
        (issuer) => "documentStore" in issuer || "certificateStore" in issuer
      );
    }
    return false;
  }, "test"),
  skip: /* @__PURE__ */ __name(() => {
    return Promise.resolve({
      status: "SKIPPED",
      type,
      name,
      reason: {
        code: 1 /* SKIPPED */,
        codeString: OpencertsRegistryCode[1 /* SKIPPED */],
        message: `Document issuers doesn't have "documentStore" or "certificateStore" property or ${v3.Method.DocumentStore} method`
      }
    });
  }, "skip"),
  verify: /* @__PURE__ */ __name(async (document) => {
    let registry;
    try {
      const res = await fetch("https://opencerts.io/static/registry.json");
      if (!res.ok) {
        throw new CodedError(
          `Non-OK response while fetching registry.json (status ${res.status})`,
          2 /* UNEXPECTED_ERROR */,
          "UNEXPECTED_ERROR"
        );
      }
      const registryJson = await res.json();
      registry = registryJson;
    } catch (error) {
      throw new CodedError(
        `Failed to retrieve opencerts registry: ${error.message}`,
        2 /* UNEXPECTED_ERROR */,
        "UNEXPECTED_ERROR"
      );
    }
    if (utils$1.isWrappedV3Document(document)) {
      const data = storeToData(registry, document.openAttestationMetadata.proof.value);
      if (OpencertsRegistryVerificationValidData.guard(data)) {
        return {
          type,
          name,
          status: "VALID",
          data
        };
      } else {
        return {
          type,
          name,
          status: "SKIPPED",
          reason: data.reason
        };
      }
    }
    const documentData = getData(document);
    const issuerFragments = documentData.issuers.map(
      (issuer) => storeToData(registry, issuer.documentStore || issuer.certificateStore || "")
    );
    const invalidIssuer = issuerFragments.find(OpencertsRegistryVerificationInvalidData.guard);
    if (OpencertsRegistryVerificationInvalidDataArray.guard(issuerFragments) && OpencertsRegistryVerificationInvalidData.guard(invalidIssuer)) {
      return {
        type,
        name,
        status: "SKIPPED",
        reason: invalidIssuer.reason
      };
    } else if (OpencertsRegistryVerificationValidDataArray.guard(issuerFragments)) {
      return {
        type,
        name,
        status: "VALID",
        data: issuerFragments
      };
    }
    throw new CodedError(
      "Unable to retrieve the reason of the failure",
      2 /* UNEXPECTED_ERROR */,
      "UNEXPECTED_ERROR"
    );
  }, "verify")
};
const isValidOpenCert = /* @__PURE__ */ __name((verificationFragments, types = ["DOCUMENT_STATUS", "DOCUMENT_INTEGRITY", "ISSUER_IDENTITY"]) => {
  if (verificationFragments.length < 1) {
    throw new Error("Please provide at least one verification fragment to check");
  }
  if (types.length < 1) {
    throw new Error("Please provide at least one type to check");
  }
  return types.every((currentType) => {
    const verificationFragmentsForType = verificationFragments.filter(
      (fragment) => fragment.type === currentType
    );
    const defaultCheck = isValid(verificationFragments, [currentType]);
    if (currentType === "DOCUMENT_STATUS" || currentType === "DOCUMENT_INTEGRITY" || defaultCheck) {
      return defaultCheck;
    }
    const issuerIdentityFragments = verificationFragmentsForType.filter(
      (fragment) => fragment.type === "ISSUER_IDENTITY"
    );
    return issuerIdentityFragments.some((fragment) => fragment.status === "VALID");
  });
}, "isValidOpenCert");
const getOpencertsRegistryVerifierFragment = utils.getFragmentByName(name);
const verifyOpenCertSignature = /* @__PURE__ */ __name((builderOptions) => verificationBuilder([...openAttestationVerifiers, registryVerifier], builderOptions), "verifyOpenCertSignature");

export { OpencertsRegistryCode, OpencertsRegistryVerificationInvalidData, OpencertsRegistryVerificationInvalidDataArray, OpencertsRegistryVerificationValidData, OpencertsRegistryVerificationValidDataArray, RegistryEntry, getOpencertsRegistryVerifierFragment, isValidOpenCert, name, registryVerifier, type, verifyOpenCertSignature };
