// @tradetrust-tt/token-registry-v5/dist/contracts/contracts/TitleEscrowFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../common";

export interface TitleEscrowFactoryInterface extends Interface {
  getFunction(
    nameOrSignature: "create" | "getEscrowAddress" | "implementation"
  ): FunctionFragment;

  getEvent(nameOrSignatureOrTopic: "TitleEscrowCreated"): EventFragment;

  encodeFunctionData(
    functionFragment: "create",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getEscrowAddress",
    values: [AddressLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "implementation",
    values?: undefined
  ): string;

  decodeFunctionResult(functionFragment: "create", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "getEscrowAddress",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "implementation",
    data: BytesLike
  ): Result;
}

export namespace TitleEscrowCreatedEvent {
  export type InputTuple = [
    titleEscrow: AddressLike,
    tokenRegistry: AddressLike,
    tokenId: BigNumberish
  ];
  export type OutputTuple = [
    titleEscrow: string,
    tokenRegistry: string,
    tokenId: bigint
  ];
  export interface OutputObject {
    titleEscrow: string;
    tokenRegistry: string;
    tokenId: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface TitleEscrowFactory extends BaseContract {
  connect(runner?: ContractRunner | null): TitleEscrowFactory;
  waitForDeployment(): Promise<this>;

  interface: TitleEscrowFactoryInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  create: TypedContractMethod<[tokenId: BigNumberish], [string], "nonpayable">;

  getEscrowAddress: TypedContractMethod<
    [tokenRegistry: AddressLike, tokenId: BigNumberish],
    [string],
    "view"
  >;

  implementation: TypedContractMethod<[], [string], "view">;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "create"
  ): TypedContractMethod<[tokenId: BigNumberish], [string], "nonpayable">;
  getFunction(
    nameOrSignature: "getEscrowAddress"
  ): TypedContractMethod<
    [tokenRegistry: AddressLike, tokenId: BigNumberish],
    [string],
    "view"
  >;
  getFunction(
    nameOrSignature: "implementation"
  ): TypedContractMethod<[], [string], "view">;

  getEvent(
    key: "TitleEscrowCreated"
  ): TypedContractEvent<
    TitleEscrowCreatedEvent.InputTuple,
    TitleEscrowCreatedEvent.OutputTuple,
    TitleEscrowCreatedEvent.OutputObject
  >;

  filters: {
    "TitleEscrowCreated(address,address,uint256)": TypedContractEvent<
      TitleEscrowCreatedEvent.InputTuple,
      TitleEscrowCreatedEvent.OutputTuple,
      TitleEscrowCreatedEvent.OutputObject
    >;
    TitleEscrowCreated: TypedContractEvent<
      TitleEscrowCreatedEvent.InputTuple,
      TitleEscrowCreatedEvent.OutputTuple,
      TitleEscrowCreatedEvent.OutputObject
    >;
  };
}
