// @trustvc/trustvc/dist/cjs/utils/documents/index.js
'use strict';

var openAttestation = require('../../open-attestation');
var fragments = require('../../verify/fragments');
var vc = require('../../w3c/vc');
var supportedChains = require('../supportedChains');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const getTransferableRecordsCredentialStatus = /* @__PURE__ */ __name((document) => {
  return [
    document?.credentialStatus
  ].flat()?.[0];
}, "getTransferableRecordsCredentialStatus");
const isTransferableRecord = /* @__PURE__ */ __name((document) => {
  let isTransferableAssetVal = false;
  if (vc.isSignedDocument(document)) {
    const credentialStatus = getTransferableRecordsCredentialStatus(document);
    isTransferableAssetVal = credentialStatus?.type === fragments.TRANSFERABLE_RECORDS_TYPE;
  } else {
    isTransferableAssetVal = openAttestation.isTransferableAsset(document);
  }
  return isTransferableAssetVal;
}, "isTransferableRecord");
const getTokenRegistryAddress = /* @__PURE__ */ __name((document) => {
  let issuerAddress = "";
  if (vc.isSignedDocument(document)) {
    const credentialStatus = getTransferableRecordsCredentialStatus(document);
    issuerAddress = credentialStatus?.tokenRegistry;
  } else {
    issuerAddress = openAttestation.getIssuerAddress(document);
  }
  return issuerAddress instanceof Array ? issuerAddress[0] : issuerAddress;
}, "getTokenRegistryAddress");
const getTokenId = /* @__PURE__ */ __name((document) => {
  let tokenId = "";
  if (vc.isSignedDocument(document)) {
    const credentialStatus = getTransferableRecordsCredentialStatus(document);
    tokenId = credentialStatus?.tokenId;
  } else {
    tokenId = openAttestation.getAssetId(document);
  }
  return tokenId && `0x${tokenId}`;
}, "getTokenId");
function processOAChainId(document) {
  if (document.network?.chainId) {
    const chainId = parseInt(document.network.chainId, 10);
    if (Object.values(supportedChains.CHAIN_ID).map(Number).includes(chainId)) {
      return chainId;
    }
    throw new Error(`Chain ID ${chainId} is not supported`);
  }
  console.warn(
    "You are using an older version of Open-Attestation Document, to use the auto network feature, please use an updated version. Otherwise, please make sure that you select the correct network."
  );
  return void 0;
}
__name(processOAChainId, "processOAChainId");
const getChainId = /* @__PURE__ */ __name((document) => {
  if (vc.isSignedDocument(document)) {
    const credentialStatus = getTransferableRecordsCredentialStatus(document);
    return credentialStatus?.tokenNetwork?.chainId;
  } else if (openAttestation.isWrappedV2Document(document)) {
    const documentData = openAttestation.getDataV2(document);
    const identityProofType = documentData?.issuers?.[0]?.identityProof?.type;
    if (identityProofType === "DNS-DID" || identityProofType === "DID") return void 0;
    return processOAChainId(documentData);
  } else if (openAttestation.isWrappedV3Document(document)) {
    const identityProofType = document?.openAttestationMetadata?.identityProof?.type;
    if (identityProofType === "DNS-DID" || identityProofType === "DID") return void 0;
    return processOAChainId(document);
  } else {
    return void 0;
  }
}, "getChainId");
const isObfuscated = /* @__PURE__ */ __name(async (document) => {
  if (openAttestation.isWrappedV3Document(document)) {
    return !!document.proof.privacy?.obfuscated?.length;
  }
  if (openAttestation.isWrappedV2Document(document)) {
    return !!document.privacy?.obfuscatedData?.length;
  }
  if (vc.isSignedDocument(document)) {
    return await vc.isDerived(document);
  }
  throw new Error(
    "Unsupported document type: Can only check if there are obfuscated data from wrapped OpenAttestation v2, v3 documents and signed verifiable credentials."
  );
}, "isObfuscated");
const getObfuscatedData = /* @__PURE__ */ __name((document) => {
  if (openAttestation.isWrappedV3Document(document)) {
    return document.proof.privacy?.obfuscated;
  }
  if (openAttestation.isWrappedV2Document(document)) {
    return document.privacy?.obfuscatedData || [];
  }
  throw new Error(
    "Unsupported document type: Can only retrieve obfuscated data from wrapped OpenAttestation v2 & v3 documents."
  );
}, "getObfuscatedData");

exports.getChainId = getChainId;
exports.getObfuscatedData = getObfuscatedData;
exports.getTokenId = getTokenId;
exports.getTokenRegistryAddress = getTokenRegistryAddress;
exports.getTransferableRecordsCredentialStatus = getTransferableRecordsCredentialStatus;
exports.isObfuscated = isObfuscated;
exports.isTransferableRecord = isTransferableRecord;
