// @tradetrust-tt/tt-verify/dist/cjs/verifiers/documentStatus/documentStore/ethereumDocumentStoreStatus.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.openAttestationEthereumDocumentStoreStatus = exports.isIssuedOnDocumentStore = exports.getIssuersDocumentStores = void 0;
const tradetrust_1 = require("@tradetrust-tt/tradetrust");
const document_store_1 = require("@tradetrust-tt/document-store");
const error_1 = require("../../../types/error");
const error_2 = require("../../../common/error");
const errorHandler_1 = require("../../../common/errorHandler");
const utils_1 = require("../utils");
const revocation_types_1 = require("../revocation.types");
const ethereumDocumentStoreStatus_type_1 = require("./ethereumDocumentStoreStatus.type");
const name = "OpenAttestationEthereumDocumentStoreStatus";
const type = "DOCUMENT_STATUS";
// Returns list of all document stores, throws when not all issuers are using document store
const getIssuersDocumentStores = (document) => {
    const data = (0, tradetrust_1.getData)(document);
    return data.issuers.map((issuer) => {
        const documentStoreAddress = issuer.documentStore || issuer.certificateStore;
        if (!documentStoreAddress)
            throw new error_2.CodedError(`Document store address not found in issuer ${issuer.name}`, error_1.OpenAttestationEthereumDocumentStoreStatusCode.INVALID_ISSUERS, error_1.OpenAttestationEthereumDocumentStoreStatusCode[error_1.OpenAttestationEthereumDocumentStoreStatusCode.INVALID_ISSUERS]);
        return documentStoreAddress;
    });
};
exports.getIssuersDocumentStores = getIssuersDocumentStores;
const isIssuedOnDocumentStore = ({ documentStore, merkleRoot, provider, }) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const documentStoreContract = yield document_store_1.DocumentStoreFactory.connect(documentStore, provider);
        const issued = yield documentStoreContract.isIssued(merkleRoot);
        return issued
            ? {
                issued: true,
                address: documentStore,
            }
            : {
                issued: false,
                address: documentStore,
                reason: {
                    message: `Document ${merkleRoot} has not been issued under contract ${documentStore}`,
                    code: error_1.OpenAttestationEthereumDocumentStoreStatusCode.DOCUMENT_NOT_ISSUED,
                    codeString: error_1.OpenAttestationEthereumDocumentStoreStatusCode[error_1.OpenAttestationEthereumDocumentStoreStatusCode.DOCUMENT_NOT_ISSUED],
                },
            };
    }
    catch (error) {
        // If error can be decoded and it's because of document is not issued, we return false
        // Else allow error to continue to bubble up
        return {
            issued: false,
            address: documentStore,
            reason: {
                message: (0, utils_1.decodeError)(error),
                code: error_1.OpenAttestationEthereumDocumentStoreStatusCode.DOCUMENT_NOT_ISSUED,
                codeString: error_1.OpenAttestationEthereumDocumentStoreStatusCode[error_1.OpenAttestationEthereumDocumentStoreStatusCode.DOCUMENT_NOT_ISSUED],
            },
        };
    }
});
exports.isIssuedOnDocumentStore = isIssuedOnDocumentStore;
const skip = () => __awaiter(void 0, void 0, void 0, function* () {
    return {
        status: "SKIPPED",
        type,
        name,
        reason: {
            code: error_1.OpenAttestationEthereumDocumentStoreStatusCode.SKIPPED,
            codeString: error_1.OpenAttestationEthereumDocumentStoreStatusCode[error_1.OpenAttestationEthereumDocumentStoreStatusCode.SKIPPED],
            message: `Document issuers doesn't have "documentStore" or "certificateStore" property or ${tradetrust_1.v3.Method.DocumentStore} method`,
        },
    };
});
const test = (document) => {
    if (tradetrust_1.utils.isWrappedV2Document(document)) {
        const documentData = (0, tradetrust_1.getData)(document);
        return documentData.issuers.some((issuer) => "documentStore" in issuer || "certificateStore" in issuer);
    }
    else if (tradetrust_1.utils.isWrappedV3Document(document)) {
        return document.openAttestationMetadata.proof.method === tradetrust_1.v3.Method.DocumentStore;
    }
    return false;
};
const verifyV2 = (document, options) => __awaiter(void 0, void 0, void 0, function* () {
    const documentStores = (0, exports.getIssuersDocumentStores)(document);
    const merkleRoot = `0x${document.signature.merkleRoot}`;
    const { targetHash } = document.signature;
    const proofs = document.signature.proof || [];
    const issuanceStatuses = yield Promise.all(documentStores.map((documentStore) => (0, exports.isIssuedOnDocumentStore)({ documentStore, merkleRoot, provider: options.provider })));
    const notIssued = issuanceStatuses.find(ethereumDocumentStoreStatus_type_1.InvalidDocumentStoreIssuanceStatus.guard);
    if (ethereumDocumentStoreStatus_type_1.InvalidDocumentStoreIssuanceStatus.guard(notIssued)) {
        return {
            name,
            type,
            data: {
                issuedOnAll: false,
                details: { issuance: issuanceStatuses },
            },
            reason: notIssued.reason,
            status: "INVALID",
        };
    }
    const revocationStatuses = yield Promise.all(documentStores.map((documentStore) => (0, utils_1.isRevokedOnDocumentStore)({
        documentStore,
        merkleRoot,
        targetHash,
        proofs,
        provider: options.provider,
    })));
    const revoked = revocationStatuses.find(revocation_types_1.InvalidRevocationStatus.guard);
    if (revocation_types_1.InvalidRevocationStatus.guard(revoked)) {
        return {
            name,
            type,
            data: {
                issuedOnAll: true,
                revokedOnAny: true,
                details: { issuance: issuanceStatuses, revocation: revocationStatuses },
            },
            reason: revoked.reason,
            status: "INVALID",
        };
    }
    if (ethereumDocumentStoreStatus_type_1.ValidDocumentStoreIssuanceStatusArray.guard(issuanceStatuses) &&
        revocation_types_1.ValidRevocationStatusArray.guard(revocationStatuses)) {
        return {
            name,
            type,
            data: {
                issuedOnAll: true,
                revokedOnAny: false,
                details: { issuance: issuanceStatuses, revocation: revocationStatuses },
            },
            status: "VALID",
        };
    }
    throw new error_2.CodedError("Reached an unexpected state when verifying v2 document", error_1.OpenAttestationEthereumDocumentStoreStatusCode.UNEXPECTED_ERROR, "UNEXPECTED_ERROR");
});
const verifyV3 = (document, options) => __awaiter(void 0, void 0, void 0, function* () {
    const { merkleRoot: merkleRootRaw, targetHash, proofs } = document.proof;
    const merkleRoot = `0x${merkleRootRaw}`;
    const { value: documentStore } = document.openAttestationMetadata.proof;
    const issuance = yield (0, exports.isIssuedOnDocumentStore)({ documentStore, merkleRoot, provider: options.provider });
    const revocation = yield (0, utils_1.isRevokedOnDocumentStore)({
        documentStore,
        merkleRoot,
        targetHash,
        proofs,
        provider: options.provider,
    });
    const data = {
        issuedOnAll: issuance.issued,
        revokedOnAny: revocation.revoked,
        details: {
            issuance,
            revocation,
        },
    };
    if (ethereumDocumentStoreStatus_type_1.ValidDocumentStoreDataV3.guard(data)) {
        return {
            name,
            type,
            data,
            status: "VALID",
        };
    }
    let reason;
    if (revocation_types_1.InvalidRevocationStatus.guard(revocation)) {
        reason = revocation.reason;
    }
    else if (ethereumDocumentStoreStatus_type_1.InvalidDocumentStoreIssuanceStatus.guard(issuance)) {
        reason = issuance.reason;
    }
    if (!reason) {
        throw new error_2.CodedError("Unable to retrieve the reason of the failure", error_1.OpenAttestationEthereumDocumentStoreStatusCode.UNEXPECTED_ERROR, "UNEXPECTED_ERROR");
    }
    return {
        name,
        type,
        data,
        status: "INVALID",
        reason,
    };
});
const verify = (document, options) => __awaiter(void 0, void 0, void 0, function* () {
    if (tradetrust_1.utils.isWrappedV2Document(document))
        return verifyV2(document, options);
    else if (tradetrust_1.utils.isWrappedV3Document(document))
        return verifyV3(document, options);
    throw new error_2.CodedError(`Document does not match either v2 or v3 formats. Consider using \`utils.diagnose\` from open-attestation to find out more.`, error_1.OpenAttestationEthereumDocumentStoreStatusCode.UNRECOGNIZED_DOCUMENT, error_1.OpenAttestationEthereumDocumentStoreStatusCode[error_1.OpenAttestationEthereumDocumentStoreStatusCode.UNRECOGNIZED_DOCUMENT]);
});
exports.openAttestationEthereumDocumentStoreStatus = {
    skip,
    test,
    verify: (0, errorHandler_1.withCodedErrorHandler)(verify, {
        name,
        type,
        unexpectedErrorCode: error_1.OpenAttestationEthereumDocumentStoreStatusCode.UNEXPECTED_ERROR,
        unexpectedErrorString: error_1.OpenAttestationEthereumDocumentStoreStatusCode[error_1.OpenAttestationEthereumDocumentStoreStatusCode.UNEXPECTED_ERROR],
    }),
};
