// @tradetrust-tt/token-registry/dist/contracts/contracts/interfaces/ITitleEscrowSignable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TokenReceivedEvent = exports.ShredEvent = exports.ReturnToIssuerEvent = exports.RejectTransferOwnersEvent = exports.RejectTransferHolderEvent = exports.RejectTransferBeneficiaryEvent = exports.NominationEvent = exports.HolderTransferEvent = exports.CancelBeneficiaryTransferEndorsementEvent = exports.BeneficiaryTransferEvent = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck
let BeneficiaryTransferEvent = exports.BeneficiaryTransferEvent = void 0;
let CancelBeneficiaryTransferEndorsementEvent = exports.CancelBeneficiaryTransferEndorsementEvent = void 0;
let HolderTransferEvent = exports.HolderTransferEvent = void 0;
let NominationEvent = exports.NominationEvent = void 0;
let RejectTransferBeneficiaryEvent = exports.RejectTransferBeneficiaryEvent = void 0;
let RejectTransferHolderEvent = exports.RejectTransferHolderEvent = void 0;
let RejectTransferOwnersEvent = exports.RejectTransferOwnersEvent = void 0;
let ReturnToIssuerEvent = exports.ReturnToIssuerEvent = void 0;
let ShredEvent = exports.ShredEvent = void 0;
let TokenReceivedEvent = exports.TokenReceivedEvent = void 0;
