// @trustvc/trustvc/dist/cjs/index.js
'use strict';

var tokenRegistryV4 = require('./token-registry-v4');
var tokenRegistryV5 = require('./token-registry-v5');
var documentStore = require('./document-store');
var transaction = require('./transaction');
var tokenRegistryFunctions = require('./token-registry-functions');
var core = require('./core');
var openAttestation = require('./open-attestation');
var verify = require('./verify');
var openCert = require('./open-cert');
var w3c = require('./w3c');
var utils = require('./utils');
var dnsprove = require('./dnsprove');



Object.defineProperty(exports, "v4ComputeInterfaceId", {
  enumerable: true,
  get: function () { return tokenRegistryV4.v4ComputeInterfaceId; }
});
Object.defineProperty(exports, "v4ComputeTitleEscrowAddress", {
  enumerable: true,
  get: function () { return tokenRegistryV4.v4ComputeTitleEscrowAddress; }
});
Object.defineProperty(exports, "v4ContractAddress", {
  enumerable: true,
  get: function () { return tokenRegistryV4.v4ContractAddress; }
});
Object.defineProperty(exports, "v4Contracts", {
  enumerable: true,
  get: function () { return tokenRegistryV4.v4Contracts; }
});
Object.defineProperty(exports, "v4EncodeInitParams", {
  enumerable: true,
  get: function () { return tokenRegistryV4.v4EncodeInitParams; }
});
Object.defineProperty(exports, "v4GetEventFromReceipt", {
  enumerable: true,
  get: function () { return tokenRegistryV4.v4GetEventFromReceipt; }
});
Object.defineProperty(exports, "v4RoleHash", {
  enumerable: true,
  get: function () { return tokenRegistryV4.v4RoleHash; }
});
Object.defineProperty(exports, "v4SupportInterfaceIds", {
  enumerable: true,
  get: function () { return tokenRegistryV4.v4SupportInterfaceIds; }
});
Object.defineProperty(exports, "v4Utils", {
  enumerable: true,
  get: function () { return tokenRegistryV4.v4Utils; }
});
Object.defineProperty(exports, "v5ComputeInterfaceId", {
  enumerable: true,
  get: function () { return tokenRegistryV5.v5ComputeInterfaceId; }
});
Object.defineProperty(exports, "v5ContractAddress", {
  enumerable: true,
  get: function () { return tokenRegistryV5.v5ContractAddress; }
});
Object.defineProperty(exports, "v5Contracts", {
  enumerable: true,
  get: function () { return tokenRegistryV5.v5Contracts; }
});
Object.defineProperty(exports, "v5EncodeInitParams", {
  enumerable: true,
  get: function () { return tokenRegistryV5.v5EncodeInitParams; }
});
Object.defineProperty(exports, "v5GetEventFromReceipt", {
  enumerable: true,
  get: function () { return tokenRegistryV5.v5GetEventFromReceipt; }
});
Object.defineProperty(exports, "v5RoleHash", {
  enumerable: true,
  get: function () { return tokenRegistryV5.v5RoleHash; }
});
Object.defineProperty(exports, "v5SupportInterfaceIds", {
  enumerable: true,
  get: function () { return tokenRegistryV5.v5SupportInterfaceIds; }
});
Object.defineProperty(exports, "v5Utils", {
  enumerable: true,
  get: function () { return tokenRegistryV5.v5Utils; }
});
Object.defineProperty(exports, "DocumentStore__factory", {
  enumerable: true,
  get: function () { return documentStore.DocumentStore__factory; }
});
Object.defineProperty(exports, "TransferableDocumentStore__factory", {
  enumerable: true,
  get: function () { return documentStore.TransferableDocumentStore__factory; }
});
Object.defineProperty(exports, "deployDocumentStore", {
  enumerable: true,
  get: function () { return documentStore.deployDocumentStore; }
});
Object.defineProperty(exports, "documentStoreGrantRole", {
  enumerable: true,
  get: function () { return documentStore.documentStoreGrantRole; }
});
Object.defineProperty(exports, "documentStoreIssue", {
  enumerable: true,
  get: function () { return documentStore.documentStoreIssue; }
});
Object.defineProperty(exports, "documentStoreRevoke", {
  enumerable: true,
  get: function () { return documentStore.documentStoreRevoke; }
});
Object.defineProperty(exports, "documentStoreRevokeRole", {
  enumerable: true,
  get: function () { return documentStore.documentStoreRevokeRole; }
});
Object.defineProperty(exports, "documentStoreTransferOwnership", {
  enumerable: true,
  get: function () { return documentStore.documentStoreTransferOwnership; }
});
Object.defineProperty(exports, "getRoleString", {
  enumerable: true,
  get: function () { return documentStore.getRoleString; }
});
Object.defineProperty(exports, "cancelTransaction", {
  enumerable: true,
  get: function () { return transaction.cancelTransaction; }
});
Object.keys(transaction).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return transaction[k]; }
  });
});
Object.keys(tokenRegistryFunctions).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return tokenRegistryFunctions[k]; }
  });
});
Object.keys(core).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return core[k]; }
  });
});
Object.keys(openAttestation).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return openAttestation[k]; }
  });
});
Object.keys(verify).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return verify[k]; }
  });
});
Object.keys(openCert).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return openCert[k]; }
  });
});
Object.keys(w3c).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return w3c[k]; }
  });
});
Object.keys(utils).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return utils[k]; }
  });
});
Object.keys(dnsprove).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return dnsprove[k]; }
  });
});
