// @tradetrust-tt/tt-verify/dist/cjs/did/verifier.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifySignature = exports.verifySecp256k1VerificationKey2018 = exports.DidVerificationStatusArray = exports.DidVerificationStatus = exports.InvalidDidVerificationStatus = exports.ValidDidVerificationStatusArray = exports.ValidDidVerificationStatus = void 0;
const ethers_1 = require("ethers");
const runtypes_1 = require("runtypes");
const resolver_1 = require("./resolver");
const error_1 = require("../types/error");
const error_2 = require("../common/error");
exports.ValidDidVerificationStatus = (0, runtypes_1.Record)({
    verified: (0, runtypes_1.Literal)(true),
    did: runtypes_1.String,
});
exports.ValidDidVerificationStatusArray = (0, runtypes_1.Array)(exports.ValidDidVerificationStatus).withConstraint((elements) => elements.length > 0 || "Expect at least one valid element");
exports.InvalidDidVerificationStatus = (0, runtypes_1.Record)({
    verified: (0, runtypes_1.Literal)(false),
    did: runtypes_1.String,
    reason: error_1.Reason,
});
exports.DidVerificationStatus = (0, runtypes_1.Union)(exports.ValidDidVerificationStatus, exports.InvalidDidVerificationStatus);
exports.DidVerificationStatusArray = (0, runtypes_1.Array)(exports.DidVerificationStatus);
const verifySecp256k1VerificationKey2018 = ({ did, verificationMethod, merkleRoot, signature, }) => {
    const messageBytes = ethers_1.utils.arrayify(merkleRoot);
    const { blockchainAccountId } = verificationMethod;
    if (!blockchainAccountId) {
        return {
            did,
            verified: false,
            reason: {
                code: error_1.OpenAttestationSignatureCode.KEY_MISSING,
                codeString: error_1.OpenAttestationSignatureCode[error_1.OpenAttestationSignatureCode.KEY_MISSING],
                message: `ethereumAddress not found on public key ${JSON.stringify(verificationMethod)}`,
            },
        };
    }
    // blockchainAccountId looks like 0x0cE1854a3836daF9130028Cf90D6d35B1Ae46457@eip155:3, let's get rid of the part after @, @ included
    const ethereumAddress = blockchainAccountId.split("@")[0];
    const merkleRootSigned = ethers_1.utils.verifyMessage(messageBytes, signature).toLowerCase() === ethereumAddress.toLowerCase();
    if (!merkleRootSigned) {
        return {
            did,
            verified: false,
            reason: {
                code: error_1.OpenAttestationSignatureCode.WRONG_SIGNATURE,
                codeString: error_1.OpenAttestationSignatureCode[error_1.OpenAttestationSignatureCode.WRONG_SIGNATURE],
                message: `merkle root is not signed correctly by ${ethereumAddress}`,
            },
        };
    }
    return {
        did,
        verified: true,
    };
};
exports.verifySecp256k1VerificationKey2018 = verifySecp256k1VerificationKey2018;
const verifySignature = ({ key, merkleRoot, signature, did, resolver, }) => __awaiter(void 0, void 0, void 0, function* () {
    const verificationMethod = yield (0, resolver_1.getVerificationMethod)(did, key, resolver);
    if (!verificationMethod)
        throw new error_2.CodedError(`No public key found on DID document for the DID ${did} and key ${key}`, error_1.OpenAttestationSignatureCode.KEY_NOT_IN_DID, "KEY_NOT_IN_DID");
    switch (verificationMethod.type) {
        case "EcdsaSecp256k1RecoveryMethod2020":
            return (0, exports.verifySecp256k1VerificationKey2018)({
                did,
                verificationMethod,
                merkleRoot,
                signature,
            });
        default:
            throw new error_2.CodedError(`Signature type ${verificationMethod.type} is currently not support`, error_1.OpenAttestationSignatureCode.UNSUPPORTED_KEY_TYPE, "UNSUPPORTED_KEY_TYPE");
    }
});
exports.verifySignature = verifySignature;
