// @tradetrust-tt/token-registry-v4/dist/contracts/contracts/mocks/SigHelperMock.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../common";

export declare namespace SigHelper {
  export type SigStruct = {
    r: PromiseOrValue<BytesLike>;
    s: PromiseOrValue<BytesLike>;
    v: PromiseOrValue<BigNumberish>;
  };

  export type SigStructOutput = [string, string, number] & {
    r: string;
    s: string;
    v: number;
  };
}

export interface SigHelperMockInterface extends utils.Interface {
  functions: {
    "DOMAIN_SEPARATOR()": FunctionFragment;
    "__SigHelper_initInternal(string,string)": FunctionFragment;
    "cancelHashInternal(bytes32)": FunctionFragment;
    "cancelled(bytes32)": FunctionFragment;
    "nonces(address)": FunctionFragment;
    "validateSigInternal(bytes32,address,(bytes32,bytes32,uint8))": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "DOMAIN_SEPARATOR"
      | "DOMAIN_SEPARATOR()"
      | "__SigHelper_initInternal"
      | "__SigHelper_initInternal(string,string)"
      | "cancelHashInternal"
      | "cancelHashInternal(bytes32)"
      | "cancelled"
      | "cancelled(bytes32)"
      | "nonces"
      | "nonces(address)"
      | "validateSigInternal"
      | "validateSigInternal(bytes32,address,(bytes32,bytes32,uint8))"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "DOMAIN_SEPARATOR",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "DOMAIN_SEPARATOR()",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "__SigHelper_initInternal",
    values: [PromiseOrValue<string>, PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "__SigHelper_initInternal(string,string)",
    values: [PromiseOrValue<string>, PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "cancelHashInternal",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(
    functionFragment: "cancelHashInternal(bytes32)",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(
    functionFragment: "cancelled",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(
    functionFragment: "cancelled(bytes32)",
    values: [PromiseOrValue<BytesLike>]
  ): string;
  encodeFunctionData(
    functionFragment: "nonces",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "nonces(address)",
    values: [PromiseOrValue<string>]
  ): string;
  encodeFunctionData(
    functionFragment: "validateSigInternal",
    values: [
      PromiseOrValue<BytesLike>,
      PromiseOrValue<string>,
      SigHelper.SigStruct
    ]
  ): string;
  encodeFunctionData(
    functionFragment: "validateSigInternal(bytes32,address,(bytes32,bytes32,uint8))",
    values: [
      PromiseOrValue<BytesLike>,
      PromiseOrValue<string>,
      SigHelper.SigStruct
    ]
  ): string;

  decodeFunctionResult(
    functionFragment: "DOMAIN_SEPARATOR",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "DOMAIN_SEPARATOR()",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "__SigHelper_initInternal",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "__SigHelper_initInternal(string,string)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "cancelHashInternal",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "cancelHashInternal(bytes32)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "cancelled", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "cancelled(bytes32)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "nonces", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "nonces(address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "validateSigInternal",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "validateSigInternal(bytes32,address,(bytes32,bytes32,uint8))",
    data: BytesLike
  ): Result;

  events: {};
}

export interface SigHelperMock extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: SigHelperMockInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    DOMAIN_SEPARATOR(overrides?: CallOverrides): Promise<[string]>;

    "DOMAIN_SEPARATOR()"(overrides?: CallOverrides): Promise<[string]>;

    __SigHelper_initInternal(
      name: PromiseOrValue<string>,
      version: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    "__SigHelper_initInternal(string,string)"(
      name: PromiseOrValue<string>,
      version: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    cancelHashInternal(
      hash: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    "cancelHashInternal(bytes32)"(
      hash: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    cancelled(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    "cancelled(bytes32)"(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    nonces(
      arg0: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    "nonces(address)"(
      arg0: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    validateSigInternal(
      hash: PromiseOrValue<BytesLike>,
      signer: PromiseOrValue<string>,
      sig: SigHelper.SigStruct,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    "validateSigInternal(bytes32,address,(bytes32,bytes32,uint8))"(
      hash: PromiseOrValue<BytesLike>,
      signer: PromiseOrValue<string>,
      sig: SigHelper.SigStruct,
      overrides?: CallOverrides
    ): Promise<[boolean]>;
  };

  DOMAIN_SEPARATOR(overrides?: CallOverrides): Promise<string>;

  "DOMAIN_SEPARATOR()"(overrides?: CallOverrides): Promise<string>;

  __SigHelper_initInternal(
    name: PromiseOrValue<string>,
    version: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  "__SigHelper_initInternal(string,string)"(
    name: PromiseOrValue<string>,
    version: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  cancelHashInternal(
    hash: PromiseOrValue<BytesLike>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  "cancelHashInternal(bytes32)"(
    hash: PromiseOrValue<BytesLike>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  cancelled(
    arg0: PromiseOrValue<BytesLike>,
    overrides?: CallOverrides
  ): Promise<boolean>;

  "cancelled(bytes32)"(
    arg0: PromiseOrValue<BytesLike>,
    overrides?: CallOverrides
  ): Promise<boolean>;

  nonces(
    arg0: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  "nonces(address)"(
    arg0: PromiseOrValue<string>,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  validateSigInternal(
    hash: PromiseOrValue<BytesLike>,
    signer: PromiseOrValue<string>,
    sig: SigHelper.SigStruct,
    overrides?: CallOverrides
  ): Promise<boolean>;

  "validateSigInternal(bytes32,address,(bytes32,bytes32,uint8))"(
    hash: PromiseOrValue<BytesLike>,
    signer: PromiseOrValue<string>,
    sig: SigHelper.SigStruct,
    overrides?: CallOverrides
  ): Promise<boolean>;

  callStatic: {
    DOMAIN_SEPARATOR(overrides?: CallOverrides): Promise<string>;

    "DOMAIN_SEPARATOR()"(overrides?: CallOverrides): Promise<string>;

    __SigHelper_initInternal(
      name: PromiseOrValue<string>,
      version: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;

    "__SigHelper_initInternal(string,string)"(
      name: PromiseOrValue<string>,
      version: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;

    cancelHashInternal(
      hash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<void>;

    "cancelHashInternal(bytes32)"(
      hash: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<void>;

    cancelled(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<boolean>;

    "cancelled(bytes32)"(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<boolean>;

    nonces(
      arg0: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    "nonces(address)"(
      arg0: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    validateSigInternal(
      hash: PromiseOrValue<BytesLike>,
      signer: PromiseOrValue<string>,
      sig: SigHelper.SigStruct,
      overrides?: CallOverrides
    ): Promise<boolean>;

    "validateSigInternal(bytes32,address,(bytes32,bytes32,uint8))"(
      hash: PromiseOrValue<BytesLike>,
      signer: PromiseOrValue<string>,
      sig: SigHelper.SigStruct,
      overrides?: CallOverrides
    ): Promise<boolean>;
  };

  filters: {};

  estimateGas: {
    DOMAIN_SEPARATOR(overrides?: CallOverrides): Promise<BigNumber>;

    "DOMAIN_SEPARATOR()"(overrides?: CallOverrides): Promise<BigNumber>;

    __SigHelper_initInternal(
      name: PromiseOrValue<string>,
      version: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    "__SigHelper_initInternal(string,string)"(
      name: PromiseOrValue<string>,
      version: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    cancelHashInternal(
      hash: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    "cancelHashInternal(bytes32)"(
      hash: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    cancelled(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    "cancelled(bytes32)"(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    nonces(
      arg0: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    "nonces(address)"(
      arg0: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    validateSigInternal(
      hash: PromiseOrValue<BytesLike>,
      signer: PromiseOrValue<string>,
      sig: SigHelper.SigStruct,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    "validateSigInternal(bytes32,address,(bytes32,bytes32,uint8))"(
      hash: PromiseOrValue<BytesLike>,
      signer: PromiseOrValue<string>,
      sig: SigHelper.SigStruct,
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    DOMAIN_SEPARATOR(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    "DOMAIN_SEPARATOR()"(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    __SigHelper_initInternal(
      name: PromiseOrValue<string>,
      version: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    "__SigHelper_initInternal(string,string)"(
      name: PromiseOrValue<string>,
      version: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    cancelHashInternal(
      hash: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    "cancelHashInternal(bytes32)"(
      hash: PromiseOrValue<BytesLike>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    cancelled(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    "cancelled(bytes32)"(
      arg0: PromiseOrValue<BytesLike>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    nonces(
      arg0: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    "nonces(address)"(
      arg0: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    validateSigInternal(
      hash: PromiseOrValue<BytesLike>,
      signer: PromiseOrValue<string>,
      sig: SigHelper.SigStruct,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    "validateSigInternal(bytes32,address,(bytes32,bytes32,uint8))"(
      hash: PromiseOrValue<BytesLike>,
      signer: PromiseOrValue<string>,
      sig: SigHelper.SigStruct,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
