// @tradetrust-tt/token-registry/dist/contracts/factories/contracts/interfaces/ITradeTrustTokenBurnable__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  ITradeTrustTokenBurnable,
  ITradeTrustTokenBurnableInterface,
} from "../../../contracts/interfaces/ITradeTrustTokenBurnable";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256",
        name: "tokenId",
        type: "uint256",
      },
      {
        internalType: "bytes",
        name: "remark",
        type: "bytes",
      },
    ],
    name: "burn",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class ITradeTrustTokenBurnable__factory {
  static readonly abi = _abi;
  static createInterface(): ITradeTrustTokenBurnableInterface {
    return new Interface(_abi) as ITradeTrustTokenBurnableInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): ITradeTrustTokenBurnable {
    return new Contract(
      address,
      _abi,
      runner
    ) as unknown as ITradeTrustTokenBurnable;
  }
}
