// @tradetrust-tt/token-registry/dist/contracts/factories/contracts/interfaces/TradeTrustTokenErrors__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  TradeTrustTokenErrors,
  TradeTrustTokenErrorsInterface,
} from "../../../contracts/interfaces/TradeTrustTokenErrors";

const _abi = [
  {
    inputs: [],
    name: "InvalidTokenId",
    type: "error",
  },
  {
    inputs: [],
    name: "RemarkLengthExceeded",
    type: "error",
  },
  {
    inputs: [],
    name: "TokenExists",
    type: "error",
  },
  {
    inputs: [],
    name: "TokenNotReturnedToIssuer",
    type: "error",
  },
  {
    inputs: [],
    name: "TransferFailure",
    type: "error",
  },
] as const;

export class TradeTrustTokenErrors__factory {
  static readonly abi = _abi;
  static createInterface(): TradeTrustTokenErrorsInterface {
    return new Interface(_abi) as TradeTrustTokenErrorsInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): TradeTrustTokenErrors {
    return new Contract(
      address,
      _abi,
      runner
    ) as unknown as TradeTrustTokenErrors;
  }
}
