// @trustvc/w3c-vc/dist/esm/lib/w3c-vc.js
import { Bls12381G2KeyPair, BbsBlsSignature2020, deriveProof, BbsBlsSignatureProof2020 } from '@mattrglobal/jsonld-signatures-bbs';
import { getDocumentLoader } from '@trustvc/w3c-context';
import jsonldSignatures from 'jsonld-signatures';
import * as jsonld from 'jsonld';
import { _checkCredential, _checkKeyPair, prefilCredentialId } from './helper';
import { proofTypeMapping } from './types';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const isRawDocument = /* @__PURE__ */ __name((document) => {
  try {
    _checkCredential(document, void 0, "sign");
  } catch (err) {
    return false;
  }
  return typeof document === "object";
}, "isRawDocument");
const isSignedDocument = /* @__PURE__ */ __name((document) => {
  try {
    _checkCredential(document, void 0, "verify");
  } catch (err) {
    return false;
  }
  return typeof document === "object" && "proof" in document;
}, "isSignedDocument");
const signCredential = /* @__PURE__ */ __name(async (credential, keyPair, cryptoSuite = "BbsBlsSignature2020", options) => {
  try {
    if (cryptoSuite === "BbsBlsSignature2020") {
      _checkKeyPair(keyPair);
      _checkCredential(credential, void 0, "sign");
      const documentLoader = options?.documentLoader ?? await getDocumentLoader();
      const key = new Bls12381G2KeyPair(keyPair);
      credential = prefilCredentialId(credential);
      const signed = await jsonldSignatures.sign(credential, {
        suite: new BbsBlsSignature2020({
          key
        }),
        purpose: new jsonldSignatures.purposes.AssertionProofPurpose(),
        documentLoader
      });
      return {
        signed
      };
    } else {
      return {
        error: `"${cryptoSuite}" is not supported.`
      };
    }
  } catch (err) {
    if (!(err instanceof Error)) {
      return {
        error: "An error occurred while signing the credential."
      };
    }
    const errorMessage = err?.message;
    if (errorMessage.includes("Dereferencing a URL did not result in a valid JSON-LD object")) {
      const errorUrl = err?.details?.url;
      return {
        error: `Dereferencing a URL did not result in a valid JSON-LD object: "${errorUrl || "Unknown URL"}"`
      };
    }
    return {
      error: errorMessage
    };
  }
}, "signCredential");
const verifyCredential = /* @__PURE__ */ __name(async (credential, options) => {
  try {
    _checkCredential(credential);
    const documentLoader = options?.documentLoader ?? await getDocumentLoader();
    let verificationResult;
    const proofType = jsonld.getValues(credential, "proof")[0].type;
    const SuiteClass = proofTypeMapping[proofType];
    if (SuiteClass) verificationResult = await jsonldSignatures.verify(credential, {
      suite: new SuiteClass(),
      purpose: new jsonldSignatures.purposes.AssertionProofPurpose(),
      documentLoader
    });
    if (verificationResult.verified) {
      return {
        verified: true
      };
    } else if (verificationResult.error && verificationResult.error.errors) {
      const errorMessage = verificationResult.error.errors[0].message;
      if (errorMessage.includes("Unexpected token N in JSON at position 0") || errorMessage.includes("Cannot read properties of undefined (reading 'length')")) {
        return {
          verified: false,
          error: "Verification method could not be found."
        };
      }
      if (errorMessage.includes("Dereferencing a URL did not result in a valid JSON-LD object")) {
        const errorUrl = verificationResult.error.errors[0].details?.url;
        return {
          verified: false,
          error: `Dereferencing a URL did not result in a valid JSON-LD object: "${errorUrl || "Unknown URL"}"`
        };
      }
      return {
        verified: false,
        error: errorMessage
      };
    }
    return {
      verified: false,
      error: "Verification error."
    };
  } catch (err) {
    if (!(err instanceof Error)) {
      return {
        verified: false,
        error: "An error occurred while verifying the credential."
      };
    }
    return {
      verified: false,
      error: err.message
    };
  }
}, "verifyCredential");
const deriveCredential = /* @__PURE__ */ __name(async (credential, revealedAttributes, options) => {
  try {
    _checkCredential(credential);
    const documentLoader = options?.documentLoader ?? await getDocumentLoader();
    const derivedProof = await deriveProof(credential, revealedAttributes, {
      suite: new BbsBlsSignatureProof2020(),
      documentLoader
    });
    return {
      derived: derivedProof
    };
  } catch (err) {
    if (!(err instanceof Error)) {
      return {
        error: "An error occurred while deriving the proof."
      };
    }
    return {
      error: err.message
    };
  }
}, "deriveCredential");

export { deriveCredential, isRawDocument, isSignedDocument, signCredential, verifyCredential };
