// @trustvc/trustvc/dist/cjs/dnsprove/index.js
'use strict';

var dnsprove = require('@tradetrust-tt/dnsprove');



Object.defineProperty(exports, "defaultDnsResolvers", {
  enumerable: true,
  get: function () { return dnsprove.defaultDnsResolvers; }
});
Object.defineProperty(exports, "getDnsDidRecords", {
  enumerable: true,
  get: function () { return dnsprove.getDnsDidRecords; }
});
Object.defineProperty(exports, "getDocumentStoreRecords", {
  enumerable: true,
  get: function () { return dnsprove.getDocumentStoreRecords; }
});
Object.defineProperty(exports, "parseDnsDidResults", {
  enumerable: true,
  get: function () { return dnsprove.parseDnsDidResults; }
});
Object.defineProperty(exports, "parseDocumentStoreResults", {
  enumerable: true,
  get: function () { return dnsprove.parseDocumentStoreResults; }
});
Object.defineProperty(exports, "parseOpenAttestationRecord", {
  enumerable: true,
  get: function () { return dnsprove.parseOpenAttestationRecord; }
});
Object.defineProperty(exports, "queryDns", {
  enumerable: true,
  get: function () { return dnsprove.queryDns; }
});
