// @tradetrust-tt/token-registry-v5/dist/contracts/factories/contracts/interfaces/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "IERC721MetadataUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _IERC721MetadataUpgradeable__factory.IERC721MetadataUpgradeable__factory;
  }
});
Object.defineProperty(exports, "ISBTUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _ISBTUpgradeable__factory.ISBTUpgradeable__factory;
  }
});
Object.defineProperty(exports, "ITitleEscrowFactory__factory", {
  enumerable: true,
  get: function get() {
    return _ITitleEscrowFactory__factory.ITitleEscrowFactory__factory;
  }
});
Object.defineProperty(exports, "ITitleEscrowSignable__factory", {
  enumerable: true,
  get: function get() {
    return _ITitleEscrowSignable__factory.ITitleEscrowSignable__factory;
  }
});
Object.defineProperty(exports, "ITitleEscrow__factory", {
  enumerable: true,
  get: function get() {
    return _ITitleEscrow__factory.ITitleEscrow__factory;
  }
});
Object.defineProperty(exports, "ITradeTrustSBT__factory", {
  enumerable: true,
  get: function get() {
    return _ITradeTrustSBT__factory.ITradeTrustSBT__factory;
  }
});
Object.defineProperty(exports, "ITradeTrustTokenBurnable__factory", {
  enumerable: true,
  get: function get() {
    return _ITradeTrustTokenBurnable__factory.ITradeTrustTokenBurnable__factory;
  }
});
Object.defineProperty(exports, "ITradeTrustTokenMintable__factory", {
  enumerable: true,
  get: function get() {
    return _ITradeTrustTokenMintable__factory.ITradeTrustTokenMintable__factory;
  }
});
Object.defineProperty(exports, "ITradeTrustTokenRestorable__factory", {
  enumerable: true,
  get: function get() {
    return _ITradeTrustTokenRestorable__factory.ITradeTrustTokenRestorable__factory;
  }
});
Object.defineProperty(exports, "ITradeTrustToken__factory", {
  enumerable: true,
  get: function get() {
    return _ITradeTrustToken__factory.ITradeTrustToken__factory;
  }
});
Object.defineProperty(exports, "RegistryAccessErrors__factory", {
  enumerable: true,
  get: function get() {
    return _RegistryAccessErrors__factory.RegistryAccessErrors__factory;
  }
});
Object.defineProperty(exports, "SigHelperErrors__factory", {
  enumerable: true,
  get: function get() {
    return _SigHelperErrors__factory.SigHelperErrors__factory;
  }
});
Object.defineProperty(exports, "TDocDeployerErrors__factory", {
  enumerable: true,
  get: function get() {
    return _TDocDeployerErrors__factory.TDocDeployerErrors__factory;
  }
});
Object.defineProperty(exports, "TitleEscrowErrors__factory", {
  enumerable: true,
  get: function get() {
    return _TitleEscrowErrors__factory.TitleEscrowErrors__factory;
  }
});
Object.defineProperty(exports, "TitleEscrowFactoryErrors__factory", {
  enumerable: true,
  get: function get() {
    return _TitleEscrowFactoryErrors__factory.TitleEscrowFactoryErrors__factory;
  }
});
Object.defineProperty(exports, "TitleEscrowSignableErrors__factory", {
  enumerable: true,
  get: function get() {
    return _TitleEscrowSignableErrors__factory.TitleEscrowSignableErrors__factory;
  }
});
Object.defineProperty(exports, "TradeTrustTokenErrors__factory", {
  enumerable: true,
  get: function get() {
    return _TradeTrustTokenErrors__factory.TradeTrustTokenErrors__factory;
  }
});
var _IERC721MetadataUpgradeable__factory = require("./IERC721MetadataUpgradeable__factory");
var _ISBTUpgradeable__factory = require("./ISBTUpgradeable__factory");
var _ITitleEscrow__factory = require("./ITitleEscrow__factory");
var _ITitleEscrowFactory__factory = require("./ITitleEscrowFactory__factory");
var _ITitleEscrowSignable__factory = require("./ITitleEscrowSignable__factory");
var _ITradeTrustSBT__factory = require("./ITradeTrustSBT__factory");
var _ITradeTrustToken__factory = require("./ITradeTrustToken__factory");
var _ITradeTrustTokenBurnable__factory = require("./ITradeTrustTokenBurnable__factory");
var _ITradeTrustTokenMintable__factory = require("./ITradeTrustTokenMintable__factory");
var _ITradeTrustTokenRestorable__factory = require("./ITradeTrustTokenRestorable__factory");
var _RegistryAccessErrors__factory = require("./RegistryAccessErrors__factory");
var _SigHelperErrors__factory = require("./SigHelperErrors__factory");
var _TDocDeployerErrors__factory = require("./TDocDeployerErrors__factory");
var _TitleEscrowErrors__factory = require("./TitleEscrowErrors__factory");
var _TitleEscrowFactoryErrors__factory = require("./TitleEscrowFactoryErrors__factory");
var _TitleEscrowSignableErrors__factory = require("./TitleEscrowSignableErrors__factory");
var _TradeTrustTokenErrors__factory = require("./TradeTrustTokenErrors__factory");
