// @tradetrust-tt/token-registry-v5/dist/contracts/factories/contracts/interfaces/ITradeTrustTokenRestorable__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  ITradeTrustTokenRestorable,
  ITradeTrustTokenRestorableInterface,
} from "../../../contracts/interfaces/ITradeTrustTokenRestorable";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256",
        name: "tokenId",
        type: "uint256",
      },
      {
        internalType: "bytes",
        name: "remark",
        type: "bytes",
      },
    ],
    name: "restore",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class ITradeTrustTokenRestorable__factory {
  static readonly abi = _abi;
  static createInterface(): ITradeTrustTokenRestorableInterface {
    return new Interface(_abi) as ITradeTrustTokenRestorableInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): ITradeTrustTokenRestorable {
    return new Contract(
      address,
      _abi,
      runner
    ) as unknown as ITradeTrustTokenRestorable;
  }
}
