// @tradetrust-tt/token-registry-v5/dist/contracts/factories/contracts/interfaces/ITradeTrustTokenMintable__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  ITradeTrustTokenMintable,
  ITradeTrustTokenMintableInterface,
} from "../../../contracts/interfaces/ITradeTrustTokenMintable";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "beneficiary",
        type: "address",
      },
      {
        internalType: "address",
        name: "holder",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "tokenId",
        type: "uint256",
      },
      {
        internalType: "bytes",
        name: "remark",
        type: "bytes",
      },
    ],
    name: "mint",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class ITradeTrustTokenMintable__factory {
  static readonly abi = _abi;
  static createInterface(): ITradeTrustTokenMintableInterface {
    return new Interface(_abi) as ITradeTrustTokenMintableInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): ITradeTrustTokenMintable {
    return new Contract(
      address,
      _abi,
      runner
    ) as unknown as ITradeTrustTokenMintable;
  }
}
