// @tradetrust-tt/token-registry-v5/dist/contracts/factories/contracts/interfaces/TDocDeployerErrors__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  TDocDeployerErrors,
  TDocDeployerErrorsInterface,
} from "../../../contracts/interfaces/TDocDeployerErrors";

const _abi = [
  {
    inputs: [],
    name: "ImplementationAlreadyAdded",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "bytes",
        name: "payload",
        type: "bytes",
      },
    ],
    name: "ImplementationInitializationFailure",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidImplementation",
    type: "error",
  },
  {
    inputs: [],
    name: "UnsupportedImplementationContractAddress",
    type: "error",
  },
] as const;

export class TDocDeployerErrors__factory {
  static readonly abi = _abi;
  static createInterface(): TDocDeployerErrorsInterface {
    return new Interface(_abi) as TDocDeployerErrorsInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): TDocDeployerErrors {
    return new Contract(address, _abi, runner) as unknown as TDocDeployerErrors;
  }
}
