// @tradetrust-tt/token-registry-v5/dist/contracts/factories/contracts/interfaces/TitleEscrowSignableErrors__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  TitleEscrowSignableErrors,
  TitleEscrowSignableErrorsInterface,
} from "../../../contracts/interfaces/TitleEscrowSignableErrors";

const _abi = [
  {
    inputs: [],
    name: "CallerNotEndorser",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidEndorsement",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidSignature",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "endorsedBeneficiary",
        type: "address",
      },
      {
        internalType: "address",
        name: "currentBeneficiary",
        type: "address",
      },
    ],
    name: "MismatchedEndorsedBeneficiaryAndCurrentBeneficiary",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "endorsedNominee",
        type: "address",
      },
      {
        internalType: "address",
        name: "onChainNominee",
        type: "address",
      },
    ],
    name: "MismatchedEndorsedNomineeAndOnChainNominee",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "currentTimestamp",
        type: "uint256",
      },
    ],
    name: "SignatureExpired",
    type: "error",
  },
] as const;

export class TitleEscrowSignableErrors__factory {
  static readonly abi = _abi;
  static createInterface(): TitleEscrowSignableErrorsInterface {
    return new Interface(_abi) as TitleEscrowSignableErrorsInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): TitleEscrowSignableErrors {
    return new Contract(
      address,
      _abi,
      runner
    ) as unknown as TitleEscrowSignableErrors;
  }
}
