// @tradetrust-tt/tt-verify/dist/cjs/verifiers/documentStatus/didSigned/didSignedDocumentStatus.type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidDidSignedDataV3 = exports.ValidDidSignedDataV3 = exports.InvalidDidSignedDataV2 = exports.ValidDidSignedDataV2 = exports.ValidOcspResponseRevoked = exports.ValidOcspResponse = exports.ValidOcspReasonCode = exports.DidSignedIssuanceStatusArray = exports.DidSignedIssuanceStatus = exports.InvalidDidSignedIssuanceStatus = exports.ValidDidSignedIssuanceStatusArray = exports.ValidDidSignedIssuanceStatus = void 0;
const runtypes_1 = require("runtypes");
const error_1 = require("../../../types/error");
const revocation_types_1 = require("../revocation.types");
/**
 * DID signed issuance status
 */
exports.ValidDidSignedIssuanceStatus = (0, runtypes_1.Record)({
    did: runtypes_1.String,
    issued: (0, runtypes_1.Literal)(true),
});
exports.ValidDidSignedIssuanceStatusArray = (0, runtypes_1.Array)(exports.ValidDidSignedIssuanceStatus);
exports.InvalidDidSignedIssuanceStatus = (0, runtypes_1.Record)({
    did: runtypes_1.String,
    issued: (0, runtypes_1.Literal)(false),
    reason: error_1.Reason,
});
exports.DidSignedIssuanceStatus = (0, runtypes_1.Union)(exports.ValidDidSignedIssuanceStatus, exports.InvalidDidSignedIssuanceStatus);
exports.DidSignedIssuanceStatusArray = (0, runtypes_1.Array)(exports.DidSignedIssuanceStatus);
/**
 * OCSP response
 */
exports.ValidOcspReasonCode = runtypes_1.Number.withConstraint((n) => Object.values(revocation_types_1.OcspResponderRevocationReason).includes(n));
exports.ValidOcspResponse = (0, runtypes_1.Record)({
    revoked: (0, runtypes_1.Literal)(false),
    documentHash: runtypes_1.String,
});
exports.ValidOcspResponseRevoked = (0, runtypes_1.Record)({
    revoked: (0, runtypes_1.Literal)(true),
    documentHash: runtypes_1.String,
    reasonCode: exports.ValidOcspReasonCode,
});
/**
 * Data for v2 Fragments
 */
exports.ValidDidSignedDataV2 = (0, runtypes_1.Record)({
    issuedOnAll: (0, runtypes_1.Literal)(true),
    revokedOnAny: (0, runtypes_1.Literal)(false),
    details: (0, runtypes_1.Record)({
        issuance: exports.ValidDidSignedIssuanceStatusArray,
        revocation: revocation_types_1.ValidRevocationStatusArray,
    }),
});
exports.InvalidDidSignedDataV2 = (0, runtypes_1.Record)({
    issuedOnAll: runtypes_1.Boolean,
    revokedOnAny: runtypes_1.Boolean,
    details: (0, runtypes_1.Record)({
        issuance: exports.DidSignedIssuanceStatusArray,
        revocation: revocation_types_1.RevocationStatusArray,
    }),
});
/**
 * Data for v3 Fragments
 */
exports.ValidDidSignedDataV3 = (0, runtypes_1.Record)({
    issuedOnAll: (0, runtypes_1.Literal)(true),
    revokedOnAny: (0, runtypes_1.Literal)(false),
    details: (0, runtypes_1.Record)({
        issuance: exports.ValidDidSignedIssuanceStatus,
        revocation: revocation_types_1.ValidRevocationStatus,
    }),
});
exports.InvalidDidSignedDataV3 = (0, runtypes_1.Record)({
    issuedOnAll: runtypes_1.Boolean,
    revokedOnAny: runtypes_1.Boolean,
    details: (0, runtypes_1.Record)({
        issuance: exports.DidSignedIssuanceStatus,
        revocation: revocation_types_1.RevocationStatus,
    }),
});
