// @tradetrust-tt/tt-verify/dist/cjs/verifiers/documentStatus/didSigned/didSignedDocumentStatus.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.openAttestationDidSignedDocumentStatus = void 0;
const tradetrust_1 = require("@tradetrust-tt/tradetrust");
const error_1 = require("../../../types/error");
const verifier_1 = require("../../../did/verifier");
const error_2 = require("../../../common/error");
const errorHandler_1 = require("../../../common/errorHandler");
const utils_1 = require("../utils");
const revocation_types_1 = require("../revocation.types");
const didSignedDocumentStatus_type_1 = require("./didSignedDocumentStatus.type");
const name = "OpenAttestationDidSignedDocumentStatus";
const type = "DOCUMENT_STATUS";
const skip = () => __awaiter(void 0, void 0, void 0, function* () {
    return {
        status: "SKIPPED",
        type,
        name,
        reason: {
            code: error_1.OpenAttestationDidSignedDocumentStatusCode.SKIPPED,
            codeString: error_1.OpenAttestationDidSignedDocumentStatusCode[error_1.OpenAttestationDidSignedDocumentStatusCode.SKIPPED],
            message: `Document was not signed by DID directly`,
        },
    };
});
const test = (document) => {
    if (tradetrust_1.utils.isSignedWrappedV2Document(document)) {
        return document.proof.some((proof) => proof.type === "OpenAttestationSignature2018");
    }
    else if (tradetrust_1.utils.isSignedWrappedV3Document(document)) {
        return document.proof.type === "OpenAttestationMerkleProofSignature2018";
    }
    return false;
};
const transformToDidSignedIssuanceStatus = (status) => {
    return verifier_1.ValidDidVerificationStatus.guard(status)
        ? {
            issued: true,
            did: status.did,
        }
        : {
            issued: false,
            did: status.did,
            reason: status.reason,
        };
};
const verifyV2 = (document, options) => __awaiter(void 0, void 0, void 0, function* () {
    const documentData = (0, tradetrust_1.getData)(document);
    const merkleRoot = `0x${document.signature.merkleRoot}`;
    const { targetHash, proof: proofs } = document.signature;
    documentData.issuers.forEach((issuer) => {
        var _a, _b;
        if (!(((_a = issuer.identityProof) === null || _a === void 0 ? void 0 : _a.type) === "DID" || ((_b = issuer.identityProof) === null || _b === void 0 ? void 0 : _b.type) === "DNS-DID"))
            throw new error_2.CodedError("All issuers must use DID or DNS-DID identityProof type.", error_1.OpenAttestationDidSignedDocumentStatusCode.INVALID_ISSUERS, error_1.OpenAttestationDidSignedDocumentStatusCode[error_1.OpenAttestationDidSignedDocumentStatusCode.INVALID_ISSUERS]);
    });
    const { issuers } = documentData;
    const revocation = issuers.map((issuer) => issuer.revocation);
    if (revocation.some((r) => typeof (r === null || r === void 0 ? void 0 : r.type) === "undefined"))
        throw new error_2.CodedError("revocation block not found for an issuer", error_1.OpenAttestationDidSignedDocumentStatusCode.MISSING_REVOCATION, "MISSING_REVOCATION");
    const revocationStatusCallback = (revocationItem) => {
        switch (revocationItem.type) {
            case tradetrust_1.v2.RevocationType.RevocationStore:
                if (typeof revocationItem.location === "string") {
                    return (0, utils_1.isRevokedOnDocumentStore)({
                        documentStore: revocationItem.location,
                        merkleRoot,
                        provider: options.provider,
                        targetHash,
                        proofs,
                    });
                }
                throw new error_2.CodedError("missing revocation location for an issuer", error_1.OpenAttestationDidSignedDocumentStatusCode.REVOCATION_LOCATION_MISSING, "REVOCATION_LOCATION_MISSING");
            case tradetrust_1.v2.RevocationType.OcspResponder:
                const { location } = revocationItem;
                if (typeof location === "string") {
                    return (0, utils_1.isRevokedByOcspResponder)({
                        merkleRoot,
                        targetHash,
                        proofs,
                        location,
                    });
                }
                throw new error_2.CodedError("missing revocation location for an issuer", error_1.OpenAttestationDidSignedDocumentStatusCode.REVOCATION_LOCATION_MISSING, "REVOCATION_LOCATION_MISSING");
            case tradetrust_1.v2.RevocationType.None:
                return Promise.resolve({ revoked: false });
            default:
                throw new error_2.CodedError("unrecognized revocation type for an issuer", error_1.OpenAttestationDidSignedDocumentStatusCode.UNRECOGNIZED_REVOCATION_TYPE, "UNRECOGNIZED_REVOCATION_TYPE");
        }
    };
    const revocationStatuses = yield Promise.all(revocation.map(revocationStatusCallback));
    // Check that all the issuers have signed on the document
    if (!document.proof)
        throw new error_2.CodedError("Document is not signed. Proofs are missing.", error_1.OpenAttestationDidSignedDocumentStatusCode.UNSIGNED, "UNSIGNED");
    const signatureVerificationDeferred = issuers.map((issuer) => __awaiter(void 0, void 0, void 0, function* () {
        var _a;
        const key = (_a = issuer.identityProof) === null || _a === void 0 ? void 0 : _a.key;
        const did = issuer.id;
        if (!did)
            throw new error_2.CodedError("id is missing in issuer", error_1.OpenAttestationDidSignedDocumentStatusCode.DID_MISSING, "DID_MISSING");
        if (!key)
            throw new error_2.CodedError("Key is not present", error_1.OpenAttestationDidSignedDocumentStatusCode.MALFORMED_IDENTITY_PROOF, "MALFORMED_IDENTITY_PROOF");
        const correspondingProof = document.proof.find((p) => p.verificationMethod.toLowerCase() === key.toLowerCase());
        if (!correspondingProof)
            throw new error_2.CodedError(`Proof not found for ${key}`, error_1.OpenAttestationDidSignedDocumentStatusCode.CORRESPONDING_PROOF_MISSING, "CORRESPONDING_PROOF_MISSING");
        return (0, verifier_1.verifySignature)({
            merkleRoot,
            key,
            signature: correspondingProof.signature,
            did,
            resolver: options.resolver,
        });
    }));
    const issuance = yield (yield Promise.all(signatureVerificationDeferred)).map(transformToDidSignedIssuanceStatus);
    const notIssued = issuance.find(didSignedDocumentStatus_type_1.InvalidDidSignedIssuanceStatus.guard);
    const revoked = revocationStatuses.find(revocation_types_1.InvalidRevocationStatus.guard);
    const data = {
        issuedOnAll: !notIssued,
        revokedOnAny: !!revoked,
        details: {
            issuance,
            revocation: revocationStatuses,
        },
    };
    if (didSignedDocumentStatus_type_1.ValidDidSignedDataV2.guard(data)) {
        return {
            name,
            type,
            data,
            status: "VALID",
        };
    }
    let reason;
    if (didSignedDocumentStatus_type_1.InvalidDidSignedIssuanceStatus.guard(notIssued)) {
        reason = notIssued.reason;
    }
    else if (revocation_types_1.InvalidRevocationStatus.guard(revoked)) {
        reason = revoked.reason;
    }
    if (!reason) {
        throw new error_2.CodedError("Unable to retrieve the reason of the failure", error_1.OpenAttestationDidSignedDocumentStatusCode.UNEXPECTED_ERROR, "UNEXPECTED_ERROR");
    }
    return {
        name,
        type,
        data,
        status: "INVALID",
        reason,
    };
});
const verifyV3 = (document, options) => __awaiter(void 0, void 0, void 0, function* () {
    var _b;
    const { merkleRoot: merkleRootRaw, targetHash, proofs } = document.proof;
    const merkleRoot = `0x${merkleRootRaw}`;
    const metaData = document.openAttestationMetadata;
    const verificationResult = transformToDidSignedIssuanceStatus(yield (0, verifier_1.verifySignature)({
        key: document.proof.key,
        did: metaData.proof.value,
        merkleRoot,
        signature: document.proof.signature,
        resolver: options.resolver,
    }));
    if (!((_b = metaData.proof.revocation) === null || _b === void 0 ? void 0 : _b.type)) {
        throw new error_2.CodedError("revocation block not found for an issuer", error_1.OpenAttestationDidSignedDocumentStatusCode.MISSING_REVOCATION, "MISSING_REVOCATION");
    }
    const issuedOnAll = verificationResult.issued;
    const getRevocationStatus = (docType, location) => __awaiter(void 0, void 0, void 0, function* () {
        switch (docType) {
            case tradetrust_1.v3.RevocationType.RevocationStore:
                if (typeof location === "string") {
                    return (0, utils_1.isRevokedOnDocumentStore)({
                        documentStore: location,
                        merkleRoot,
                        targetHash,
                        proofs,
                        provider: options.provider,
                    });
                }
                throw new error_2.CodedError("missing revocation location for an issuer", error_1.OpenAttestationDidSignedDocumentStatusCode.REVOCATION_LOCATION_MISSING, "REVOCATION_LOCATION_MISSING");
            case tradetrust_1.v3.RevocationType.OcspResponder:
                if (typeof location === "string") {
                    return (0, utils_1.isRevokedByOcspResponder)({
                        merkleRoot,
                        targetHash,
                        proofs,
                        location,
                    });
                }
                throw new error_2.CodedError("missing revocation location for an issuer", error_1.OpenAttestationDidSignedDocumentStatusCode.REVOCATION_LOCATION_MISSING, "REVOCATION_LOCATION_MISSING");
            case tradetrust_1.v3.RevocationType.None:
                return { revoked: false };
            default:
                throw new error_2.CodedError("revocation type not found for an issuer", error_1.OpenAttestationDidSignedDocumentStatusCode.UNRECOGNIZED_REVOCATION_TYPE, "UNRECOGNIZED_REVOCATION_TYPE");
        }
    });
    const revocationStatus = yield getRevocationStatus(metaData.proof.revocation.type, metaData.proof.revocation.location);
    const revokedOnAny = revocationStatus.revoked;
    if (didSignedDocumentStatus_type_1.ValidDidSignedIssuanceStatus.guard(verificationResult) && revocation_types_1.ValidRevocationStatus.guard(revocationStatus)) {
        return {
            name,
            type,
            data: {
                issuedOnAll: true,
                revokedOnAny: false,
                details: {
                    issuance: verificationResult,
                    revocation: revocationStatus,
                },
            },
            status: "VALID",
        };
    }
    // eslint-disable-next-line no-nested-ternary
    const reason = didSignedDocumentStatus_type_1.InvalidDidSignedIssuanceStatus.guard(verificationResult)
        ? verificationResult.reason
        : revocation_types_1.InvalidRevocationStatus.guard(revocationStatus)
            ? revocationStatus.reason
            : undefined;
    if (!reason) {
        throw new error_2.CodedError("Unable to retrieve the reason of the failure", error_1.OpenAttestationDidSignedDocumentStatusCode.UNEXPECTED_ERROR, "UNEXPECTED_ERROR");
    }
    return {
        name,
        type,
        data: {
            issuedOnAll,
            revokedOnAny,
            details: {
                issuance: verificationResult,
                revocation: revocationStatus,
            },
        },
        status: "INVALID",
        reason,
    };
});
const verify = (document, options) => __awaiter(void 0, void 0, void 0, function* () {
    if (tradetrust_1.utils.isSignedWrappedV2Document(document)) {
        return verifyV2(document, options);
    }
    else if (tradetrust_1.utils.isSignedWrappedV3Document(document)) {
        return verifyV3(document, options);
    }
    throw new error_2.CodedError(`Document does not match either v2 or v3 formats. Consider using \`utils.diagnose\` from open-attestation to find out more.`, error_1.OpenAttestationDidSignedDocumentStatusCode.UNRECOGNIZED_DOCUMENT, error_1.OpenAttestationDidSignedDocumentStatusCode[error_1.OpenAttestationDidSignedDocumentStatusCode.UNRECOGNIZED_DOCUMENT]);
});
exports.openAttestationDidSignedDocumentStatus = {
    skip,
    test,
    verify: (0, errorHandler_1.withCodedErrorHandler)(verify, {
        name,
        type,
        unexpectedErrorCode: error_1.OpenAttestationDidSignedDocumentStatusCode.UNEXPECTED_ERROR,
        unexpectedErrorString: error_1.OpenAttestationDidSignedDocumentStatusCode[error_1.OpenAttestationDidSignedDocumentStatusCode.UNEXPECTED_ERROR],
    }),
};
