// @trustvc/w3c-credential-status/dist/esm/lib/index.js
import { CredentialContextVersion, BBS_V1_URL, STATUS_LIST_2021_CREDENTIAL_URL } from '@trustvc/w3c-context';
import { _checkCredentialSubjectForStatusList2021Credential } from './BitstringStatusList/assertions';
import { getValidFromDateFromCredentialStatusVC, assertCredentialStatusStatusListType } from './utils';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const VCCredentialStatusTypeToVCCredentialSubjectType = {
  StatusList2021Credential: "StatusList2021",
  BitstringStatusListCredential: "BitstringStatusList"
};
const createCredentialStatusPayload = /* @__PURE__ */ __name(async (options, keyPair, type = "StatusList2021Credential", cryptoSuite = "BbsBlsSignature2020") => {
  try {
    const { id, credentialSubject } = options;
    switch (type) {
      case "StatusList2021Credential":
        _checkCredentialSubjectForStatusList2021Credential(credentialSubject);
        break;
      default:
        throw new Error(`Unsupported type: ${type}`);
    }
    if (!credentialSubject.id && id) {
      credentialSubject.id = `${id}#list`;
    }
    const context = [
      CredentialContextVersion.v1
    ];
    if (cryptoSuite === "BbsBlsSignature2020") {
      context.push(BBS_V1_URL);
    }
    if (type === "StatusList2021Credential") {
      context.push(STATUS_LIST_2021_CREDENTIAL_URL);
    }
    const validFrom = await getValidFromDateFromCredentialStatusVC(id);
    const vc = {
      "@context": context,
      type: [
        "VerifiableCredential",
        type
      ],
      issuer: keyPair.controller,
      issuanceDate: (/* @__PURE__ */ new Date()).toISOString(),
      validFrom,
      credentialSubject
    };
    return vc;
  } catch (err) {
    if (!(err instanceof Error)) {
      throw new Error("An error occurred while signing the credential status VC.");
    }
    throw err;
  }
}, "createCredentialStatusPayload");
const isCredentialStatusStatusList = /* @__PURE__ */ __name((credentialStatus) => {
  try {
    assertCredentialStatusStatusListType(credentialStatus?.type);
    return true;
  } catch (err) {
    return false;
  }
}, "isCredentialStatusStatusList");

export { VCCredentialStatusTypeToVCCredentialSubjectType, createCredentialStatusPayload, isCredentialStatusStatusList };
