// @tradetrust-tt/token-registry/dist/contracts/factories/contracts/utils/SigHelper__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  SigHelper,
  SigHelperInterface,
} from "../../../contracts/utils/SigHelper";

const _abi = [
  {
    inputs: [],
    name: "SignatureAlreadyCancelled",
    type: "error",
  },
  {
    inputs: [],
    name: "DOMAIN_SEPARATOR",
    outputs: [
      {
        internalType: "bytes32",
        name: "",
        type: "bytes32",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "",
        type: "bytes32",
      },
    ],
    name: "cancelled",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    name: "nonces",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
] as const;

export class SigHelper__factory {
  static readonly abi = _abi;
  static createInterface(): SigHelperInterface {
    return new Interface(_abi) as SigHelperInterface;
  }
  static connect(address: string, runner?: ContractRunner | null): SigHelper {
    return new Contract(address, _abi, runner) as unknown as SigHelper;
  }
}
