// @tradetrust-tt/token-registry/dist/contracts/contracts/mocks/TitleEscrowFactoryGetterMock.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../../common";

export interface TitleEscrowFactoryGetterMockInterface extends Interface {
  getFunction(
    nameOrSignature: "callCreate" | "getEscrowAddress" | "setAddress"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "callCreate",
    values: [AddressLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getEscrowAddress",
    values: [AddressLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setAddress",
    values: [AddressLike]
  ): string;

  decodeFunctionResult(functionFragment: "callCreate", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "getEscrowAddress",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "setAddress", data: BytesLike): Result;
}

export interface TitleEscrowFactoryGetterMock extends BaseContract {
  connect(runner?: ContractRunner | null): TitleEscrowFactoryGetterMock;
  waitForDeployment(): Promise<this>;

  interface: TitleEscrowFactoryGetterMockInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  callCreate: TypedContractMethod<
    [titleEscrowFactory: AddressLike, tokenId: BigNumberish],
    [void],
    "nonpayable"
  >;

  getEscrowAddress: TypedContractMethod<
    [tokenRegistry: AddressLike, tokenId: BigNumberish],
    [string],
    "view"
  >;

  setAddress: TypedContractMethod<
    [_titleEscrowAddress: AddressLike],
    [void],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "callCreate"
  ): TypedContractMethod<
    [titleEscrowFactory: AddressLike, tokenId: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "getEscrowAddress"
  ): TypedContractMethod<
    [tokenRegistry: AddressLike, tokenId: BigNumberish],
    [string],
    "view"
  >;
  getFunction(
    nameOrSignature: "setAddress"
  ): TypedContractMethod<
    [_titleEscrowAddress: AddressLike],
    [void],
    "nonpayable"
  >;

  filters: {};
}
