// @trustvc/w3c-vc/dist/esm/lib/__fixtures__/test-scenarios.js
import { bbs2020KeyPair, ecdsa2023KeyPair, bbs2023KeyPair } from './key-pairs';
import { modernCredentialV1_1, modernCredentialV2_0 } from './modern-credentials';
import { bbs2020CredentialV1_1, bbs2020DerivedCredentialV1_1, bbs2020CredentialV2_0, bbs2020DerivedCredentialV2_0 } from './bbs2020-credentials';

const bbs2020TestScenarios = [
  {
    cryptosuite: "BbsBlsSignature2020",
    keyPair: bbs2020KeyPair,
    version: "v1.1",
    credential: bbs2020CredentialV1_1,
    derivedCredential: bbs2020DerivedCredentialV1_1
  },
  {
    cryptosuite: "BbsBlsSignature2020",
    keyPair: bbs2020KeyPair,
    version: "v2.0",
    credential: bbs2020CredentialV2_0,
    derivedCredential: bbs2020DerivedCredentialV2_0
  }
];
const modernCryptosuiteTestScenarios = [
  {
    cryptosuite: "ecdsa-sd-2023",
    keyPair: ecdsa2023KeyPair,
    version: "v1.1",
    credential: modernCredentialV1_1,
    dateField: "issuanceDate",
    dateValue: "2024-04-01T12:19:52Z"
  },
  {
    cryptosuite: "ecdsa-sd-2023",
    keyPair: ecdsa2023KeyPair,
    version: "v2.0",
    credential: modernCredentialV2_0,
    dateField: "validFrom",
    dateValue: "2024-04-01T12:19:52Z"
  },
  {
    cryptosuite: "bbs-2023",
    keyPair: bbs2023KeyPair,
    version: "v1.1",
    credential: modernCredentialV1_1,
    dateField: "issuanceDate",
    dateValue: "2024-04-01T12:19:52Z"
  },
  {
    cryptosuite: "bbs-2023",
    keyPair: bbs2023KeyPair,
    version: "v2.0",
    credential: modernCredentialV2_0,
    dateField: "validFrom",
    dateValue: "2024-04-01T12:19:52Z"
  }
];

export { bbs2020TestScenarios, modernCryptosuiteTestScenarios };
