// @trustvc/trustvc/dist/cjs/verify/fragments/document-status/transferableRecords/transferableRecordVerifier.js
'use strict';

var ttVerify = require('@tradetrust-tt/tt-verify');
var w3cVC = require('@trustvc/w3c-vc');
var utils = require('./utils');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var w3cVC__namespace = /*#__PURE__*/_interopNamespace(w3cVC);

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const TRANSFERABLE_RECORDS_TYPE = "TransferableRecords";
const type = "DOCUMENT_STATUS";
const name = TRANSFERABLE_RECORDS_TYPE;
const verify = /* @__PURE__ */ __name(async (document, options) => {
  const signedDocument = document;
  const credentialStatuses = Array.isArray(signedDocument?.credentialStatus) ? signedDocument?.credentialStatus : [signedDocument?.credentialStatus];
  const { provider } = options;
  const verificationResult = await Promise.all(
    credentialStatuses.map(async (credentialStatus) => {
      const tokenId = "0x" + credentialStatus.tokenId;
      if (!credentialStatus?.tokenRegistry) {
        throw new ttVerify.CodedError(
          "Document's credentialStatus does not have tokenRegistry",
          ttVerify.OpenAttestationEthereumTokenRegistryStatusCode.UNRECOGNIZED_DOCUMENT,
          ttVerify.OpenAttestationEthereumTokenRegistryStatusCode[ttVerify.OpenAttestationEthereumTokenRegistryStatusCode.UNRECOGNIZED_DOCUMENT]
        );
      }
      if (!credentialStatus?.tokenNetwork || !credentialStatus?.tokenNetwork?.chainId) {
        throw new ttVerify.CodedError(
          "Document's credentialStatus does not have tokenNetwork.chainId",
          ttVerify.OpenAttestationEthereumTokenRegistryStatusCode.UNRECOGNIZED_DOCUMENT,
          ttVerify.OpenAttestationEthereumTokenRegistryStatusCode[ttVerify.OpenAttestationEthereumTokenRegistryStatusCode.UNRECOGNIZED_DOCUMENT]
        );
      }
      const mintStatus = await utils.isTokenMintedOnRegistry({
        tokenRegistryAddress: credentialStatus?.tokenRegistry,
        tokenId,
        provider
      });
      return mintStatus;
    })
  );
  const result = {
    name,
    type,
    status: "INVALID",
    data: {
      tokenRegistry: credentialStatuses?.[0]?.tokenRegistry
    }
  };
  if (verificationResult.every(ttVerify.ValidTokenRegistryStatus.guard)) {
    result.status = "VALID";
  } else {
    result.reason = verificationResult?.[0]?.reason;
  }
  return result;
}, "verify");
const skip = /* @__PURE__ */ __name(async () => {
  return {
    status: "SKIPPED",
    type,
    name,
    reason: {
      code: ttVerify.OpenAttestationEthereumTokenRegistryStatusCode.SKIPPED,
      codeString: ttVerify.OpenAttestationEthereumTokenRegistryStatusCode[ttVerify.OpenAttestationEthereumTokenRegistryStatusCode.SKIPPED],
      message: `Document does not have TransferableRecords status`
    }
  };
}, "skip");
const test = /* @__PURE__ */ __name((document) => {
  const doc = document;
  const credentialStatuses = Array.isArray(doc?.credentialStatus) ? doc?.credentialStatus : [doc?.credentialStatus];
  if (w3cVC__namespace.isSignedDocument(document) && credentialStatuses.every((cs) => cs?.type === TRANSFERABLE_RECORDS_TYPE)) {
    return true;
  }
  return false;
}, "test");
const credentialStatusTransferableRecordVerifier = {
  skip,
  test,
  verify: /* @__PURE__ */ __name(async (...args) => {
    try {
      return await verify(...args);
    } catch (e) {
      if (e instanceof ttVerify.CodedError) {
        const err = {
          name,
          type,
          status: "ERROR",
          reason: {
            code: e.code,
            codeString: e.codeString,
            message: e.message
          }
        };
        return err;
      }
      return {
        name,
        type,
        status: "ERROR",
        reason: {
          code: ttVerify.OpenAttestationEthereumTokenRegistryStatusCode.UNEXPECTED_ERROR,
          codeString: ttVerify.OpenAttestationEthereumTokenRegistryStatusCode[ttVerify.OpenAttestationEthereumTokenRegistryStatusCode.UNEXPECTED_ERROR],
          message: e instanceof Error ? e.message : "An unexpected error occurred"
        }
      };
    }
  }, "verify")
};

exports.TRANSFERABLE_RECORDS_TYPE = TRANSFERABLE_RECORDS_TYPE;
exports.credentialStatusTransferableRecordVerifier = credentialStatusTransferableRecordVerifier;
