// @tradetrust-tt/token-registry-v5/dist/contracts/contracts/mocks/SigHelperMock.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../../common";

export declare namespace SigHelper {
  export type SigStruct = { r: BytesLike; s: BytesLike; v: BigNumberish };

  export type SigStructOutput = [r: string, s: string, v: bigint] & {
    r: string;
    s: string;
    v: bigint;
  };
}

export interface SigHelperMockInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "DOMAIN_SEPARATOR"
      | "__SigHelper_initInternal"
      | "cancelHashInternal"
      | "cancelled"
      | "nonces"
      | "validateSigInternal"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "DOMAIN_SEPARATOR",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "__SigHelper_initInternal",
    values: [string, string]
  ): string;
  encodeFunctionData(
    functionFragment: "cancelHashInternal",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "cancelled",
    values: [BytesLike]
  ): string;
  encodeFunctionData(functionFragment: "nonces", values: [AddressLike]): string;
  encodeFunctionData(
    functionFragment: "validateSigInternal",
    values: [BytesLike, AddressLike, SigHelper.SigStruct]
  ): string;

  decodeFunctionResult(
    functionFragment: "DOMAIN_SEPARATOR",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "__SigHelper_initInternal",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "cancelHashInternal",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "cancelled", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "nonces", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "validateSigInternal",
    data: BytesLike
  ): Result;
}

export interface SigHelperMock extends BaseContract {
  connect(runner?: ContractRunner | null): SigHelperMock;
  waitForDeployment(): Promise<this>;

  interface: SigHelperMockInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  DOMAIN_SEPARATOR: TypedContractMethod<[], [string], "view">;

  __SigHelper_initInternal: TypedContractMethod<
    [name: string, version: string],
    [void],
    "nonpayable"
  >;

  cancelHashInternal: TypedContractMethod<
    [hash: BytesLike],
    [void],
    "nonpayable"
  >;

  cancelled: TypedContractMethod<[arg0: BytesLike], [boolean], "view">;

  nonces: TypedContractMethod<[arg0: AddressLike], [bigint], "view">;

  validateSigInternal: TypedContractMethod<
    [hash: BytesLike, signer: AddressLike, sig: SigHelper.SigStruct],
    [boolean],
    "view"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "DOMAIN_SEPARATOR"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "__SigHelper_initInternal"
  ): TypedContractMethod<[name: string, version: string], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "cancelHashInternal"
  ): TypedContractMethod<[hash: BytesLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "cancelled"
  ): TypedContractMethod<[arg0: BytesLike], [boolean], "view">;
  getFunction(
    nameOrSignature: "nonces"
  ): TypedContractMethod<[arg0: AddressLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "validateSigInternal"
  ): TypedContractMethod<
    [hash: BytesLike, signer: AddressLike, sig: SigHelper.SigStruct],
    [boolean],
    "view"
  >;

  filters: {};
}
