// @trustvc/trustvc/dist/cjs/deploy/token-registry.js
'use strict';

var contracts = require('@tradetrust-tt/token-registry/contracts');
var utils = require('../token-registry-functions/utils');
var utils$1 = require('../token-registry-v5/utils');
var ethers = require('../utils/ethers');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const deployTokenRegistry = /* @__PURE__ */ __name(async (registryName, registrySymbol, signer, options = {}) => {
  const { maxFeePerGas, maxPriorityFeePerGas } = options;
  let { chainId, standalone, factoryAddress, tokenRegistryImplAddress, deployerContractAddress } = options;
  const deployerAddress = await signer.getAddress();
  if (!chainId) {
    chainId = await utils.getChainIdSafe(signer);
  }
  const {
    TitleEscrowFactory: defaultTitleEscrowFactoryAddress,
    TokenImplementation: defaultTokenImplementationContractAddress,
    Deployer: defaultDeployerContractAddress
  } = utils.getDefaultContractAddress(chainId);
  if (!utils.isValidAddress(deployerContractAddress)) {
    deployerContractAddress = defaultDeployerContractAddress;
  }
  if (!utils.isValidAddress(tokenRegistryImplAddress)) {
    tokenRegistryImplAddress = defaultTokenImplementationContractAddress;
  }
  if (standalone !== false && (!deployerContractAddress || !tokenRegistryImplAddress)) {
    console.error(
      `Network ${chainId} does not support "quick-start" mode. Defaulting to standalone mode.`
    );
    standalone = true;
  }
  if (!standalone) {
    if (!deployerContractAddress || !tokenRegistryImplAddress) {
      throw new Error(`Network ${chainId} currently is not supported. Use --standalone instead.`);
    }
    const Contract = ethers.getEthersContractFromProvider(signer.provider);
    const deployerContract = new Contract(
      deployerContractAddress,
      contracts.TDocDeployer__factory.abi,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      signer
      // Type assertion needed for v5/v6 compatibility
    );
    const initParam = utils$1.encodeInitParams({
      name: registryName,
      symbol: registrySymbol,
      deployer: deployerAddress
    });
    const txOptions = await utils.getTxOptions(signer, chainId, maxFeePerGas, maxPriorityFeePerGas);
    return await deployerContract.deploy(tokenRegistryImplAddress, initParam, txOptions);
  } else {
    if (!factoryAddress || !utils.isValidAddress(factoryAddress)) {
      factoryAddress = defaultTitleEscrowFactoryAddress;
      if (!factoryAddress) {
        throw new Error(`Network ${chainId} currently is not supported. Supply a factory address.`);
      }
    }
    const supportedTitleEscrowFactory = await utils.isSupportedTitleEscrowFactory(
      factoryAddress,
      signer.provider
    );
    if (!supportedTitleEscrowFactory) {
      throw new Error(`Title Escrow Factory ${factoryAddress} is not supported.`);
    }
    const Contract = ethers.getEthersContractFactoryFromProvider(signer.provider);
    const tokenFactory = new Contract(
      contracts.TradeTrustToken__factory.abi,
      contracts.TradeTrustToken__factory.bytecode,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      signer
      // Type assertion needed for v5/v6 compatibility
    );
    const isV6 = ethers.isV6EthersProvider(signer.provider);
    const txOptions = await utils.getTxOptions(signer, chainId, maxFeePerGas, maxPriorityFeePerGas);
    if (isV6) {
      const contract = await tokenFactory.deploy(
        registryName,
        registrySymbol,
        factoryAddress,
        txOptions
      );
      return await contract.deploymentTransaction().wait();
    } else {
      const contract = await tokenFactory.deploy(
        registryName,
        registrySymbol,
        factoryAddress,
        txOptions
      );
      return await contract.deployTransaction.wait();
    }
  }
}, "deployTokenRegistry");

exports.deployTokenRegistry = deployTokenRegistry;
