// @tradetrust-tt/token-registry-v5/dist/contracts/contracts/presets/TitleEscrowSignable.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../../common";

export type BeneficiaryTransferEndorsementStruct = {
  beneficiary: AddressLike;
  holder: AddressLike;
  nominee: AddressLike;
  registry: AddressLike;
  tokenId: BigNumberish;
  deadline: BigNumberish;
  nonce: BigNumberish;
};

export type BeneficiaryTransferEndorsementStructOutput = [
  beneficiary: string,
  holder: string,
  nominee: string,
  registry: string,
  tokenId: bigint,
  deadline: bigint,
  nonce: bigint
] & {
  beneficiary: string;
  holder: string;
  nominee: string;
  registry: string;
  tokenId: bigint;
  deadline: bigint;
  nonce: bigint;
};

export declare namespace SigHelper {
  export type SigStruct = { r: BytesLike; s: BytesLike; v: BigNumberish };

  export type SigStructOutput = [r: string, s: string, v: bigint] & {
    r: string;
    s: string;
    v: bigint;
  };
}

export interface TitleEscrowSignableInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "BENEFICIARY_TRANSFER_TYPEHASH"
      | "DOMAIN_SEPARATOR"
      | "active"
      | "beneficiary"
      | "cancelBeneficiaryTransfer"
      | "cancelled"
      | "holder"
      | "initialize"
      | "isHoldingToken"
      | "name"
      | "nominate"
      | "nominee"
      | "nonces"
      | "onERC721Received"
      | "prevBeneficiary"
      | "prevHolder"
      | "registry"
      | "rejectTransferBeneficiary"
      | "rejectTransferHolder"
      | "rejectTransferOwners"
      | "remark"
      | "returnToIssuer"
      | "shred"
      | "supportsInterface"
      | "tokenId"
      | "transferBeneficiary"
      | "transferBeneficiaryWithSig"
      | "transferHolder"
      | "transferOwners"
  ): FunctionFragment;

  getEvent(
    nameOrSignatureOrTopic:
      | "BeneficiaryTransfer"
      | "CancelBeneficiaryTransferEndorsement"
      | "HolderTransfer"
      | "Initialized"
      | "Nomination"
      | "RejectTransferBeneficiary"
      | "RejectTransferHolder"
      | "RejectTransferOwners"
      | "ReturnToIssuer"
      | "Shred"
      | "TokenReceived"
  ): EventFragment;

  encodeFunctionData(
    functionFragment: "BENEFICIARY_TRANSFER_TYPEHASH",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "DOMAIN_SEPARATOR",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "active", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "beneficiary",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "cancelBeneficiaryTransfer",
    values: [BeneficiaryTransferEndorsementStruct]
  ): string;
  encodeFunctionData(
    functionFragment: "cancelled",
    values: [BytesLike]
  ): string;
  encodeFunctionData(functionFragment: "holder", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "initialize",
    values: [AddressLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "isHoldingToken",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "name", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "nominate",
    values: [AddressLike, BytesLike]
  ): string;
  encodeFunctionData(functionFragment: "nominee", values?: undefined): string;
  encodeFunctionData(functionFragment: "nonces", values: [AddressLike]): string;
  encodeFunctionData(
    functionFragment: "onERC721Received",
    values: [AddressLike, AddressLike, BigNumberish, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "prevBeneficiary",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "prevHolder",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "registry", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "rejectTransferBeneficiary",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "rejectTransferHolder",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "rejectTransferOwners",
    values: [BytesLike]
  ): string;
  encodeFunctionData(functionFragment: "remark", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "returnToIssuer",
    values: [BytesLike]
  ): string;
  encodeFunctionData(functionFragment: "shred", values: [BytesLike]): string;
  encodeFunctionData(
    functionFragment: "supportsInterface",
    values: [BytesLike]
  ): string;
  encodeFunctionData(functionFragment: "tokenId", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "transferBeneficiary",
    values: [AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "transferBeneficiaryWithSig",
    values: [BeneficiaryTransferEndorsementStruct, SigHelper.SigStruct]
  ): string;
  encodeFunctionData(
    functionFragment: "transferHolder",
    values: [AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "transferOwners",
    values: [AddressLike, AddressLike, BytesLike]
  ): string;

  decodeFunctionResult(
    functionFragment: "BENEFICIARY_TRANSFER_TYPEHASH",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "DOMAIN_SEPARATOR",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "active", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "beneficiary",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "cancelBeneficiaryTransfer",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "cancelled", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "holder", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "initialize", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "isHoldingToken",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "name", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "nominate", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "nominee", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "nonces", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "onERC721Received",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "prevBeneficiary",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "prevHolder", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "registry", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "rejectTransferBeneficiary",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "rejectTransferHolder",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "rejectTransferOwners",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "remark", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "returnToIssuer",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "shred", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "supportsInterface",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "tokenId", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "transferBeneficiary",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferBeneficiaryWithSig",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferHolder",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferOwners",
    data: BytesLike
  ): Result;
}

export namespace BeneficiaryTransferEvent {
  export type InputTuple = [
    fromBeneficiary: AddressLike,
    toBeneficiary: AddressLike,
    registry: AddressLike,
    tokenId: BigNumberish,
    remark: BytesLike
  ];
  export type OutputTuple = [
    fromBeneficiary: string,
    toBeneficiary: string,
    registry: string,
    tokenId: bigint,
    remark: string
  ];
  export interface OutputObject {
    fromBeneficiary: string;
    toBeneficiary: string;
    registry: string;
    tokenId: bigint;
    remark: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace CancelBeneficiaryTransferEndorsementEvent {
  export type InputTuple = [
    hash: BytesLike,
    endorser: AddressLike,
    tokenId: BigNumberish
  ];
  export type OutputTuple = [hash: string, endorser: string, tokenId: bigint];
  export interface OutputObject {
    hash: string;
    endorser: string;
    tokenId: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace HolderTransferEvent {
  export type InputTuple = [
    fromHolder: AddressLike,
    toHolder: AddressLike,
    registry: AddressLike,
    tokenId: BigNumberish,
    remark: BytesLike
  ];
  export type OutputTuple = [
    fromHolder: string,
    toHolder: string,
    registry: string,
    tokenId: bigint,
    remark: string
  ];
  export interface OutputObject {
    fromHolder: string;
    toHolder: string;
    registry: string;
    tokenId: bigint;
    remark: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace InitializedEvent {
  export type InputTuple = [version: BigNumberish];
  export type OutputTuple = [version: bigint];
  export interface OutputObject {
    version: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace NominationEvent {
  export type InputTuple = [
    prevNominee: AddressLike,
    nominee: AddressLike,
    registry: AddressLike,
    tokenId: BigNumberish,
    remark: BytesLike
  ];
  export type OutputTuple = [
    prevNominee: string,
    nominee: string,
    registry: string,
    tokenId: bigint,
    remark: string
  ];
  export interface OutputObject {
    prevNominee: string;
    nominee: string;
    registry: string;
    tokenId: bigint;
    remark: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RejectTransferBeneficiaryEvent {
  export type InputTuple = [
    fromBeneficiary: AddressLike,
    toBeneficiary: AddressLike,
    registry: AddressLike,
    tokenId: BigNumberish,
    remark: BytesLike
  ];
  export type OutputTuple = [
    fromBeneficiary: string,
    toBeneficiary: string,
    registry: string,
    tokenId: bigint,
    remark: string
  ];
  export interface OutputObject {
    fromBeneficiary: string;
    toBeneficiary: string;
    registry: string;
    tokenId: bigint;
    remark: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RejectTransferHolderEvent {
  export type InputTuple = [
    fromHolder: AddressLike,
    toHolder: AddressLike,
    registry: AddressLike,
    tokenId: BigNumberish,
    remark: BytesLike
  ];
  export type OutputTuple = [
    fromHolder: string,
    toHolder: string,
    registry: string,
    tokenId: bigint,
    remark: string
  ];
  export interface OutputObject {
    fromHolder: string;
    toHolder: string;
    registry: string;
    tokenId: bigint;
    remark: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RejectTransferOwnersEvent {
  export type InputTuple = [
    fromBeneficiary: AddressLike,
    toBeneficiary: AddressLike,
    fromHolder: AddressLike,
    toHolder: AddressLike,
    registry: AddressLike,
    tokenId: BigNumberish,
    remark: BytesLike
  ];
  export type OutputTuple = [
    fromBeneficiary: string,
    toBeneficiary: string,
    fromHolder: string,
    toHolder: string,
    registry: string,
    tokenId: bigint,
    remark: string
  ];
  export interface OutputObject {
    fromBeneficiary: string;
    toBeneficiary: string;
    fromHolder: string;
    toHolder: string;
    registry: string;
    tokenId: bigint;
    remark: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace ReturnToIssuerEvent {
  export type InputTuple = [
    caller: AddressLike,
    registry: AddressLike,
    tokenId: BigNumberish,
    remark: BytesLike
  ];
  export type OutputTuple = [
    caller: string,
    registry: string,
    tokenId: bigint,
    remark: string
  ];
  export interface OutputObject {
    caller: string;
    registry: string;
    tokenId: bigint;
    remark: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace ShredEvent {
  export type InputTuple = [
    registry: AddressLike,
    tokenId: BigNumberish,
    remark: BytesLike
  ];
  export type OutputTuple = [registry: string, tokenId: bigint, remark: string];
  export interface OutputObject {
    registry: string;
    tokenId: bigint;
    remark: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace TokenReceivedEvent {
  export type InputTuple = [
    beneficiary: AddressLike,
    holder: AddressLike,
    isMinting: boolean,
    registry: AddressLike,
    tokenId: BigNumberish,
    remark: BytesLike
  ];
  export type OutputTuple = [
    beneficiary: string,
    holder: string,
    isMinting: boolean,
    registry: string,
    tokenId: bigint,
    remark: string
  ];
  export interface OutputObject {
    beneficiary: string;
    holder: string;
    isMinting: boolean;
    registry: string;
    tokenId: bigint;
    remark: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface TitleEscrowSignable extends BaseContract {
  connect(runner?: ContractRunner | null): TitleEscrowSignable;
  waitForDeployment(): Promise<this>;

  interface: TitleEscrowSignableInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  BENEFICIARY_TRANSFER_TYPEHASH: TypedContractMethod<[], [string], "view">;

  DOMAIN_SEPARATOR: TypedContractMethod<[], [string], "view">;

  active: TypedContractMethod<[], [boolean], "view">;

  beneficiary: TypedContractMethod<[], [string], "view">;

  cancelBeneficiaryTransfer: TypedContractMethod<
    [endorsement: BeneficiaryTransferEndorsementStruct],
    [void],
    "nonpayable"
  >;

  cancelled: TypedContractMethod<[arg0: BytesLike], [boolean], "view">;

  holder: TypedContractMethod<[], [string], "view">;

  initialize: TypedContractMethod<
    [_registry: AddressLike, _tokenId: BigNumberish],
    [void],
    "nonpayable"
  >;

  isHoldingToken: TypedContractMethod<[], [boolean], "view">;

  name: TypedContractMethod<[], [string], "view">;

  nominate: TypedContractMethod<
    [_nominee: AddressLike, _remark: BytesLike],
    [void],
    "nonpayable"
  >;

  nominee: TypedContractMethod<[], [string], "view">;

  nonces: TypedContractMethod<[arg0: AddressLike], [bigint], "view">;

  onERC721Received: TypedContractMethod<
    [
      arg0: AddressLike,
      arg1: AddressLike,
      _tokenId: BigNumberish,
      data: BytesLike
    ],
    [string],
    "nonpayable"
  >;

  prevBeneficiary: TypedContractMethod<[], [string], "view">;

  prevHolder: TypedContractMethod<[], [string], "view">;

  registry: TypedContractMethod<[], [string], "view">;

  rejectTransferBeneficiary: TypedContractMethod<
    [_remark: BytesLike],
    [void],
    "nonpayable"
  >;

  rejectTransferHolder: TypedContractMethod<
    [_remark: BytesLike],
    [void],
    "nonpayable"
  >;

  rejectTransferOwners: TypedContractMethod<
    [_remark: BytesLike],
    [void],
    "nonpayable"
  >;

  remark: TypedContractMethod<[], [string], "view">;

  returnToIssuer: TypedContractMethod<
    [_remark: BytesLike],
    [void],
    "nonpayable"
  >;

  shred: TypedContractMethod<[_remark: BytesLike], [void], "nonpayable">;

  supportsInterface: TypedContractMethod<
    [interfaceId: BytesLike],
    [boolean],
    "view"
  >;

  tokenId: TypedContractMethod<[], [bigint], "view">;

  transferBeneficiary: TypedContractMethod<
    [_nominee: AddressLike, _remark: BytesLike],
    [void],
    "nonpayable"
  >;

  transferBeneficiaryWithSig: TypedContractMethod<
    [
      endorsement: BeneficiaryTransferEndorsementStruct,
      sig: SigHelper.SigStruct
    ],
    [void],
    "nonpayable"
  >;

  transferHolder: TypedContractMethod<
    [newHolder: AddressLike, _remark: BytesLike],
    [void],
    "nonpayable"
  >;

  transferOwners: TypedContractMethod<
    [_nominee: AddressLike, newHolder: AddressLike, _remark: BytesLike],
    [void],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "BENEFICIARY_TRANSFER_TYPEHASH"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "DOMAIN_SEPARATOR"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "active"
  ): TypedContractMethod<[], [boolean], "view">;
  getFunction(
    nameOrSignature: "beneficiary"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "cancelBeneficiaryTransfer"
  ): TypedContractMethod<
    [endorsement: BeneficiaryTransferEndorsementStruct],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "cancelled"
  ): TypedContractMethod<[arg0: BytesLike], [boolean], "view">;
  getFunction(
    nameOrSignature: "holder"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "initialize"
  ): TypedContractMethod<
    [_registry: AddressLike, _tokenId: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "isHoldingToken"
  ): TypedContractMethod<[], [boolean], "view">;
  getFunction(
    nameOrSignature: "name"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "nominate"
  ): TypedContractMethod<
    [_nominee: AddressLike, _remark: BytesLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "nominee"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "nonces"
  ): TypedContractMethod<[arg0: AddressLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "onERC721Received"
  ): TypedContractMethod<
    [
      arg0: AddressLike,
      arg1: AddressLike,
      _tokenId: BigNumberish,
      data: BytesLike
    ],
    [string],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "prevBeneficiary"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "prevHolder"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "registry"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "rejectTransferBeneficiary"
  ): TypedContractMethod<[_remark: BytesLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "rejectTransferHolder"
  ): TypedContractMethod<[_remark: BytesLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "rejectTransferOwners"
  ): TypedContractMethod<[_remark: BytesLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "remark"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "returnToIssuer"
  ): TypedContractMethod<[_remark: BytesLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "shred"
  ): TypedContractMethod<[_remark: BytesLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "supportsInterface"
  ): TypedContractMethod<[interfaceId: BytesLike], [boolean], "view">;
  getFunction(
    nameOrSignature: "tokenId"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "transferBeneficiary"
  ): TypedContractMethod<
    [_nominee: AddressLike, _remark: BytesLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "transferBeneficiaryWithSig"
  ): TypedContractMethod<
    [
      endorsement: BeneficiaryTransferEndorsementStruct,
      sig: SigHelper.SigStruct
    ],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "transferHolder"
  ): TypedContractMethod<
    [newHolder: AddressLike, _remark: BytesLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "transferOwners"
  ): TypedContractMethod<
    [_nominee: AddressLike, newHolder: AddressLike, _remark: BytesLike],
    [void],
    "nonpayable"
  >;

  getEvent(
    key: "BeneficiaryTransfer"
  ): TypedContractEvent<
    BeneficiaryTransferEvent.InputTuple,
    BeneficiaryTransferEvent.OutputTuple,
    BeneficiaryTransferEvent.OutputObject
  >;
  getEvent(
    key: "CancelBeneficiaryTransferEndorsement"
  ): TypedContractEvent<
    CancelBeneficiaryTransferEndorsementEvent.InputTuple,
    CancelBeneficiaryTransferEndorsementEvent.OutputTuple,
    CancelBeneficiaryTransferEndorsementEvent.OutputObject
  >;
  getEvent(
    key: "HolderTransfer"
  ): TypedContractEvent<
    HolderTransferEvent.InputTuple,
    HolderTransferEvent.OutputTuple,
    HolderTransferEvent.OutputObject
  >;
  getEvent(
    key: "Initialized"
  ): TypedContractEvent<
    InitializedEvent.InputTuple,
    InitializedEvent.OutputTuple,
    InitializedEvent.OutputObject
  >;
  getEvent(
    key: "Nomination"
  ): TypedContractEvent<
    NominationEvent.InputTuple,
    NominationEvent.OutputTuple,
    NominationEvent.OutputObject
  >;
  getEvent(
    key: "RejectTransferBeneficiary"
  ): TypedContractEvent<
    RejectTransferBeneficiaryEvent.InputTuple,
    RejectTransferBeneficiaryEvent.OutputTuple,
    RejectTransferBeneficiaryEvent.OutputObject
  >;
  getEvent(
    key: "RejectTransferHolder"
  ): TypedContractEvent<
    RejectTransferHolderEvent.InputTuple,
    RejectTransferHolderEvent.OutputTuple,
    RejectTransferHolderEvent.OutputObject
  >;
  getEvent(
    key: "RejectTransferOwners"
  ): TypedContractEvent<
    RejectTransferOwnersEvent.InputTuple,
    RejectTransferOwnersEvent.OutputTuple,
    RejectTransferOwnersEvent.OutputObject
  >;
  getEvent(
    key: "ReturnToIssuer"
  ): TypedContractEvent<
    ReturnToIssuerEvent.InputTuple,
    ReturnToIssuerEvent.OutputTuple,
    ReturnToIssuerEvent.OutputObject
  >;
  getEvent(
    key: "Shred"
  ): TypedContractEvent<
    ShredEvent.InputTuple,
    ShredEvent.OutputTuple,
    ShredEvent.OutputObject
  >;
  getEvent(
    key: "TokenReceived"
  ): TypedContractEvent<
    TokenReceivedEvent.InputTuple,
    TokenReceivedEvent.OutputTuple,
    TokenReceivedEvent.OutputObject
  >;

  filters: {
    "BeneficiaryTransfer(address,address,address,uint256,bytes)": TypedContractEvent<
      BeneficiaryTransferEvent.InputTuple,
      BeneficiaryTransferEvent.OutputTuple,
      BeneficiaryTransferEvent.OutputObject
    >;
    BeneficiaryTransfer: TypedContractEvent<
      BeneficiaryTransferEvent.InputTuple,
      BeneficiaryTransferEvent.OutputTuple,
      BeneficiaryTransferEvent.OutputObject
    >;

    "CancelBeneficiaryTransferEndorsement(bytes32,address,uint256)": TypedContractEvent<
      CancelBeneficiaryTransferEndorsementEvent.InputTuple,
      CancelBeneficiaryTransferEndorsementEvent.OutputTuple,
      CancelBeneficiaryTransferEndorsementEvent.OutputObject
    >;
    CancelBeneficiaryTransferEndorsement: TypedContractEvent<
      CancelBeneficiaryTransferEndorsementEvent.InputTuple,
      CancelBeneficiaryTransferEndorsementEvent.OutputTuple,
      CancelBeneficiaryTransferEndorsementEvent.OutputObject
    >;

    "HolderTransfer(address,address,address,uint256,bytes)": TypedContractEvent<
      HolderTransferEvent.InputTuple,
      HolderTransferEvent.OutputTuple,
      HolderTransferEvent.OutputObject
    >;
    HolderTransfer: TypedContractEvent<
      HolderTransferEvent.InputTuple,
      HolderTransferEvent.OutputTuple,
      HolderTransferEvent.OutputObject
    >;

    "Initialized(uint64)": TypedContractEvent<
      InitializedEvent.InputTuple,
      InitializedEvent.OutputTuple,
      InitializedEvent.OutputObject
    >;
    Initialized: TypedContractEvent<
      InitializedEvent.InputTuple,
      InitializedEvent.OutputTuple,
      InitializedEvent.OutputObject
    >;

    "Nomination(address,address,address,uint256,bytes)": TypedContractEvent<
      NominationEvent.InputTuple,
      NominationEvent.OutputTuple,
      NominationEvent.OutputObject
    >;
    Nomination: TypedContractEvent<
      NominationEvent.InputTuple,
      NominationEvent.OutputTuple,
      NominationEvent.OutputObject
    >;

    "RejectTransferBeneficiary(address,address,address,uint256,bytes)": TypedContractEvent<
      RejectTransferBeneficiaryEvent.InputTuple,
      RejectTransferBeneficiaryEvent.OutputTuple,
      RejectTransferBeneficiaryEvent.OutputObject
    >;
    RejectTransferBeneficiary: TypedContractEvent<
      RejectTransferBeneficiaryEvent.InputTuple,
      RejectTransferBeneficiaryEvent.OutputTuple,
      RejectTransferBeneficiaryEvent.OutputObject
    >;

    "RejectTransferHolder(address,address,address,uint256,bytes)": TypedContractEvent<
      RejectTransferHolderEvent.InputTuple,
      RejectTransferHolderEvent.OutputTuple,
      RejectTransferHolderEvent.OutputObject
    >;
    RejectTransferHolder: TypedContractEvent<
      RejectTransferHolderEvent.InputTuple,
      RejectTransferHolderEvent.OutputTuple,
      RejectTransferHolderEvent.OutputObject
    >;

    "RejectTransferOwners(address,address,address,address,address,uint256,bytes)": TypedContractEvent<
      RejectTransferOwnersEvent.InputTuple,
      RejectTransferOwnersEvent.OutputTuple,
      RejectTransferOwnersEvent.OutputObject
    >;
    RejectTransferOwners: TypedContractEvent<
      RejectTransferOwnersEvent.InputTuple,
      RejectTransferOwnersEvent.OutputTuple,
      RejectTransferOwnersEvent.OutputObject
    >;

    "ReturnToIssuer(address,address,uint256,bytes)": TypedContractEvent<
      ReturnToIssuerEvent.InputTuple,
      ReturnToIssuerEvent.OutputTuple,
      ReturnToIssuerEvent.OutputObject
    >;
    ReturnToIssuer: TypedContractEvent<
      ReturnToIssuerEvent.InputTuple,
      ReturnToIssuerEvent.OutputTuple,
      ReturnToIssuerEvent.OutputObject
    >;

    "Shred(address,uint256,bytes)": TypedContractEvent<
      ShredEvent.InputTuple,
      ShredEvent.OutputTuple,
      ShredEvent.OutputObject
    >;
    Shred: TypedContractEvent<
      ShredEvent.InputTuple,
      ShredEvent.OutputTuple,
      ShredEvent.OutputObject
    >;

    "TokenReceived(address,address,bool,address,uint256,bytes)": TypedContractEvent<
      TokenReceivedEvent.InputTuple,
      TokenReceivedEvent.OutputTuple,
      TokenReceivedEvent.OutputObject
    >;
    TokenReceived: TypedContractEvent<
      TokenReceivedEvent.InputTuple,
      TokenReceivedEvent.OutputTuple,
      TokenReceivedEvent.OutputObject
    >;
  };
}
