// @trustvc/trustvc/dist/cjs/token-registry-v5/index.js
'use strict';

var contractAddress = require('./contractAddress');
var roleHash = require('./roleHash');
var supportInterfaceIds = require('./supportInterfaceIds');
var contracts = require('./contracts');
var utils = require('./utils');
var tokenRegistryV5 = require('@tradetrust-tt/token-registry-v5');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var contracts__namespace = /*#__PURE__*/_interopNamespace(contracts);



Object.defineProperty(exports, "v5ContractAddress", {
  enumerable: true,
  get: function () { return contractAddress.contractAddress; }
});
Object.defineProperty(exports, "v5RoleHash", {
  enumerable: true,
  get: function () { return roleHash.roleHash; }
});
Object.defineProperty(exports, "v5SupportInterfaceIds", {
  enumerable: true,
  get: function () { return supportInterfaceIds.supportInterfaceIds; }
});
exports.v5Contracts = contracts__namespace;
Object.defineProperty(exports, "v5ComputeInterfaceId", {
  enumerable: true,
  get: function () { return utils.computeInterfaceId; }
});
Object.defineProperty(exports, "v5EncodeInitParams", {
  enumerable: true,
  get: function () { return utils.encodeInitParams; }
});
Object.defineProperty(exports, "v5GetEventFromReceipt", {
  enumerable: true,
  get: function () { return utils.getEventFromReceipt; }
});
Object.defineProperty(exports, "constants", {
  enumerable: true,
  get: function () { return tokenRegistryV5.constants; }
});
Object.defineProperty(exports, "utils", {
  enumerable: true,
  get: function () { return tokenRegistryV5.utils; }
});
Object.defineProperty(exports, "v5Utils", {
  enumerable: true,
  get: function () { return tokenRegistryV5.utils; }
});
