// @trustvc/w3c-vc/dist/lib/__fixtures__/key-pairs.js
'use strict';

var w3cIssuer = require('@trustvc/w3c-issuer');

const bbs2020KeyPair = {
  id: "did:web:trustvc.github.io:did:1#keys-1",
  controller: "did:web:trustvc.github.io:did:1",
  type: w3cIssuer.VerificationType.Bls12381G2Key2020,
  publicKeyBase58: "oRfEeWFresvhRtXCkihZbxyoi2JER7gHTJ5psXhHsdCoU1MttRMi3Yp9b9fpjmKh7bMgfWKLESiK2YovRd8KGzJsGuamoAXfqDDVhckxuc9nmsJ84skCSTijKeU4pfAcxeJ",
  privateKeyBase58: "4LDU56PUhA9ZEutnR1qCWQnUhtLtpLu2EHSq4h1o7vtF"
};
const ecdsa2023KeyPair = {
  "@context": "https://w3id.org/security/multikey/v1",
  id: "did:web:trustvc.github.io:did:1#multikey-1",
  type: w3cIssuer.VerificationType.Multikey,
  controller: "did:web:trustvc.github.io:did:1",
  publicKeyMultibase: "zDnaemDNwi4G5eTzGfRooFFu5Kns3be6yfyVNtiaMhWkZbwtc",
  secretKeyMultibase: "z42tmUXTVn3n9BihE6NhdMpvVBTnFTgmb6fw18o5Ud6puhRW"
};
const bbs2023KeyPair = {
  "@context": "https://w3id.org/security/multikey/v1",
  id: "did:web:trustvc.github.io:did:1#multikey-2",
  type: w3cIssuer.VerificationType.Multikey,
  controller: "did:web:trustvc.github.io:did:1",
  publicKeyMultibase: "zUC7HnpncVAkTjtL6B8prX6bQM2WA5sJ7rXFeCqyrvPnrzoFBjYsVUTNwzhhPUazja73tWwPeEBWCUgq5qBSrtrXiYhVvBCgZPTCiWANj7TSiZJ6SnyC3pkt94GiuChhAvmRRbt",
  secretKeyMultibase: "z488ur1KSFDd3Y1L6pXcPrZRjE18PNBhgzwJvMeoSxKPNysj"
};

exports.bbs2020KeyPair = bbs2020KeyPair;
exports.bbs2023KeyPair = bbs2023KeyPair;
exports.ecdsa2023KeyPair = ecdsa2023KeyPair;
