// @trustvc/w3c-vc/dist/lib/__fixtures__/test-scenarios.js
'use strict';

var keyPairs = require('./key-pairs');
var modernCredentials = require('./modern-credentials');
var bbs2020Credentials = require('./bbs2020-credentials');

const bbs2020TestScenarios = [
  {
    cryptosuite: "BbsBlsSignature2020",
    keyPair: keyPairs.bbs2020KeyPair,
    version: "v1.1",
    credential: bbs2020Credentials.bbs2020CredentialV1_1,
    derivedCredential: bbs2020Credentials.bbs2020DerivedCredentialV1_1
  },
  {
    cryptosuite: "BbsBlsSignature2020",
    keyPair: keyPairs.bbs2020KeyPair,
    version: "v2.0",
    credential: bbs2020Credentials.bbs2020CredentialV2_0,
    derivedCredential: bbs2020Credentials.bbs2020DerivedCredentialV2_0
  }
];
const modernCryptosuiteTestScenarios = [
  {
    cryptosuite: "ecdsa-sd-2023",
    keyPair: keyPairs.ecdsa2023KeyPair,
    version: "v1.1",
    credential: modernCredentials.modernCredentialV1_1,
    dateField: "issuanceDate",
    dateValue: "2024-04-01T12:19:52Z"
  },
  {
    cryptosuite: "ecdsa-sd-2023",
    keyPair: keyPairs.ecdsa2023KeyPair,
    version: "v2.0",
    credential: modernCredentials.modernCredentialV2_0,
    dateField: "validFrom",
    dateValue: "2024-04-01T12:19:52Z"
  },
  {
    cryptosuite: "bbs-2023",
    keyPair: keyPairs.bbs2023KeyPair,
    version: "v1.1",
    credential: modernCredentials.modernCredentialV1_1,
    dateField: "issuanceDate",
    dateValue: "2024-04-01T12:19:52Z"
  },
  {
    cryptosuite: "bbs-2023",
    keyPair: keyPairs.bbs2023KeyPair,
    version: "v2.0",
    credential: modernCredentials.modernCredentialV2_0,
    dateField: "validFrom",
    dateValue: "2024-04-01T12:19:52Z"
  }
];

exports.bbs2020TestScenarios = bbs2020TestScenarios;
exports.modernCryptosuiteTestScenarios = modernCryptosuiteTestScenarios;
