// @tradetrust-tt/tt-verify/dist/cjs/verifiers/documentStatus/tokenRegistry/ethereumTokenRegistryStatus.type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidTokenRegistryDataV3 = exports.ValidTokenRegistryDataV3 = exports.InvalidTokenRegistryDataV2 = exports.ValidTokenRegistryDataV2 = exports.InvalidTokenRegistryStatus = exports.ValidTokenRegistryStatus = void 0;
const runtypes_1 = require("runtypes");
const error_1 = require("../../../types/error");
/**
 * Token registry mint status
 */
exports.ValidTokenRegistryStatus = (0, runtypes_1.Record)({
    minted: (0, runtypes_1.Literal)(true),
    address: runtypes_1.String,
});
exports.InvalidTokenRegistryStatus = (0, runtypes_1.Record)({
    minted: (0, runtypes_1.Literal)(false),
    address: runtypes_1.String,
    reason: error_1.Reason,
});
/**
 * Data for v2 Fragments
 */
exports.ValidTokenRegistryDataV2 = (0, runtypes_1.Record)({
    mintedOnAll: (0, runtypes_1.Literal)(true),
    details: (0, runtypes_1.Array)(exports.ValidTokenRegistryStatus),
});
exports.InvalidTokenRegistryDataV2 = (0, runtypes_1.Record)({
    mintedOnAll: (0, runtypes_1.Literal)(false),
    details: (0, runtypes_1.Array)((0, runtypes_1.Union)(exports.ValidTokenRegistryStatus, exports.InvalidTokenRegistryStatus)),
});
/**
 * Data for v3 Fragments
 */
exports.ValidTokenRegistryDataV3 = (0, runtypes_1.Record)({
    mintedOnAll: (0, runtypes_1.Literal)(true),
    details: exports.ValidTokenRegistryStatus,
});
exports.InvalidTokenRegistryDataV3 = (0, runtypes_1.Record)({
    mintedOnAll: (0, runtypes_1.Literal)(false),
    details: exports.InvalidTokenRegistryStatus,
});
