// @tradetrust-tt/tt-verify/dist/cjs/verifiers/documentStatus/tokenRegistry/ethereumTokenRegistryStatus.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.openAttestationEthereumTokenRegistryStatus = exports.isTokenMintedOnRegistry = exports.getTokenRegistry = void 0;
const tradetrust_1 = require("@tradetrust-tt/tradetrust");
const contracts_1 = require("@tradetrust-tt/token-registry/contracts");
const ethers_1 = require("ethers");
const error_1 = require("../../../types/error");
const error_2 = require("../../../common/error");
const errorHandler_1 = require("../../../common/errorHandler");
const ethereumTokenRegistryStatus_type_1 = require("./ethereumTokenRegistryStatus.type");
const name = "OpenAttestationEthereumTokenRegistryStatus";
const type = "DOCUMENT_STATUS";
const getTokenRegistry = (document) => {
    if (tradetrust_1.utils.isWrappedV2Document(document)) {
        const { issuers } = (0, tradetrust_1.getData)(document);
        if (issuers.length !== 1)
            throw new error_2.CodedError("Only one issuer is allowed for tokens", error_1.OpenAttestationEthereumTokenRegistryStatusCode.INVALID_ISSUERS, error_1.OpenAttestationEthereumTokenRegistryStatusCode[error_1.OpenAttestationEthereumTokenRegistryStatusCode.INVALID_ISSUERS]);
        if (!issuers[0].tokenRegistry)
            throw new error_2.CodedError("Token registry is undefined", error_1.OpenAttestationEthereumTokenRegistryStatusCode.UNDEFINED_TOKEN_REGISTRY, error_1.OpenAttestationEthereumTokenRegistryStatusCode[error_1.OpenAttestationEthereumTokenRegistryStatusCode.UNDEFINED_TOKEN_REGISTRY]);
        return issuers[0].tokenRegistry;
    }
    if (tradetrust_1.utils.isWrappedV3Document(document)) {
        if (!document.openAttestationMetadata.proof.value)
            throw new error_2.CodedError("Token registry is undefined", error_1.OpenAttestationEthereumTokenRegistryStatusCode.UNDEFINED_TOKEN_REGISTRY, error_1.OpenAttestationEthereumTokenRegistryStatusCode[error_1.OpenAttestationEthereumTokenRegistryStatusCode.UNDEFINED_TOKEN_REGISTRY]);
        return document.openAttestationMetadata.proof.value;
    }
    throw new error_2.CodedError(`Document does not match either v2 or v3 formats. Consider using \`utils.diagnose\` from open-attestation to find out more.`, error_1.OpenAttestationEthereumTokenRegistryStatusCode.UNRECOGNIZED_DOCUMENT, error_1.OpenAttestationEthereumTokenRegistryStatusCode[error_1.OpenAttestationEthereumTokenRegistryStatusCode.UNRECOGNIZED_DOCUMENT]);
};
exports.getTokenRegistry = getTokenRegistry;
const getMerkleRoot = (document) => {
    if (tradetrust_1.utils.isWrappedV2Document(document))
        return `0x${document.signature.merkleRoot}`;
    else if (tradetrust_1.utils.isWrappedV3Document(document))
        return `0x${document.proof.merkleRoot}`;
    throw new error_2.CodedError(`Document does not match either v2 or v3 formats. Consider using \`utils.diagnose\` from open-attestation to find out more.`, error_1.OpenAttestationEthereumTokenRegistryStatusCode.UNRECOGNIZED_DOCUMENT, error_1.OpenAttestationEthereumTokenRegistryStatusCode[error_1.OpenAttestationEthereumTokenRegistryStatusCode.UNRECOGNIZED_DOCUMENT]);
};
const isNonExistentToken = (error) => {
    const message = error.message;
    if (!message)
        return false;
    return message.includes("owner query for nonexistent token");
};
const isMissingTokenRegistry = (error) => {
    var _a;
    return (!error.reason &&
        ((_a = error.method) === null || _a === void 0 ? void 0 : _a.toLowerCase()) === "ownerOf(uint256)".toLowerCase() &&
        error.code === ethers_1.errors.CALL_EXCEPTION);
};
const decodeError = (error) => {
    var _a;
    const reason = error.reason && Array.isArray(error.reason) ? error.reason[0] : (_a = error.reason) !== null && _a !== void 0 ? _a : "";
    switch (true) {
        case isNonExistentToken(error):
            return `Document has not been issued under token registry`;
        case isMissingTokenRegistry(error):
            return `Token registry is not found`;
        case reason.toLowerCase() === "ENS name not configured".toLowerCase() &&
            error.code === ethers_1.errors.UNSUPPORTED_OPERATION:
            return "ENS name is not configured";
        case reason.toLowerCase() === "invalid address".toLowerCase() && error.code === ethers_1.errors.INVALID_ARGUMENT:
            return `Invalid token registry address`;
        case error.code === ethers_1.errors.INVALID_ARGUMENT:
            return `Invalid contract arguments`;
        case error.code === ethers_1.errors.SERVER_ERROR:
            throw new error_2.CodedError("Unable to connect to the Ethereum network, please try again later", error_1.OpenAttestationEthereumTokenRegistryStatusCode.SERVER_ERROR, error_1.OpenAttestationEthereumTokenRegistryStatusCode[error_1.OpenAttestationEthereumTokenRegistryStatusCode.SERVER_ERROR]);
        default:
            throw error;
    }
};
const isTokenMintedOnRegistry = ({ tokenRegistry, merkleRoot, provider, }) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const tokenRegistryContract = yield contracts_1.TradeTrustToken__factory.connect(tokenRegistry, provider);
        const minted = yield tokenRegistryContract
            .ownerOf(merkleRoot)
            .then((owner) => !(owner === ethers_1.constants.AddressZero));
        return minted
            ? { minted, address: tokenRegistry }
            : {
                minted,
                address: tokenRegistry,
                reason: {
                    code: error_1.OpenAttestationEthereumTokenRegistryStatusCode.DOCUMENT_NOT_MINTED,
                    codeString: error_1.OpenAttestationEthereumTokenRegistryStatusCode[error_1.OpenAttestationEthereumTokenRegistryStatusCode.DOCUMENT_NOT_MINTED],
                    message: `Document ${merkleRoot} has not been issued under contract ${tokenRegistry}`,
                },
            };
    }
    catch (error) {
        return {
            minted: false,
            address: tokenRegistry,
            reason: {
                message: decodeError(error),
                code: error_1.OpenAttestationEthereumTokenRegistryStatusCode.DOCUMENT_NOT_MINTED,
                codeString: error_1.OpenAttestationEthereumTokenRegistryStatusCode[error_1.OpenAttestationEthereumTokenRegistryStatusCode.DOCUMENT_NOT_MINTED],
            },
        };
    }
});
exports.isTokenMintedOnRegistry = isTokenMintedOnRegistry;
const skip = () => __awaiter(void 0, void 0, void 0, function* () {
    return {
        status: "SKIPPED",
        type,
        name,
        reason: {
            code: error_1.OpenAttestationEthereumTokenRegistryStatusCode.SKIPPED,
            codeString: error_1.OpenAttestationEthereumTokenRegistryStatusCode[error_1.OpenAttestationEthereumTokenRegistryStatusCode.SKIPPED],
            message: `Document issuers doesn't have "tokenRegistry" property or ${tradetrust_1.v3.Method.TokenRegistry} method`,
        },
    };
});
const test = (document) => {
    if (tradetrust_1.utils.isWrappedV2Document(document)) {
        const documentData = (0, tradetrust_1.getData)(document);
        return documentData.issuers.some((issuer) => "tokenRegistry" in issuer);
    }
    else if (tradetrust_1.utils.isWrappedV3Document(document)) {
        return document.openAttestationMetadata.proof.method === tradetrust_1.v3.Method.TokenRegistry;
    }
    return false;
};
// TODO split
const verify = (document, options) => __awaiter(void 0, void 0, void 0, function* () {
    if (!tradetrust_1.utils.isWrappedV3Document(document) && !tradetrust_1.utils.isWrappedV2Document(document))
        throw new error_2.CodedError(`Document does not match either v2 or v3 formats. Consider using \`utils.diagnose\` from open-attestation to find out more.`, error_1.OpenAttestationEthereumTokenRegistryStatusCode.UNRECOGNIZED_DOCUMENT, error_1.OpenAttestationEthereumTokenRegistryStatusCode[error_1.OpenAttestationEthereumTokenRegistryStatusCode.UNRECOGNIZED_DOCUMENT]);
    const tokenRegistry = (0, exports.getTokenRegistry)(document);
    const merkleRoot = getMerkleRoot(document);
    const mintStatus = yield (0, exports.isTokenMintedOnRegistry)({ tokenRegistry, merkleRoot, provider: options.provider });
    if (ethereumTokenRegistryStatus_type_1.ValidTokenRegistryStatus.guard(mintStatus)) {
        const fragment = {
            name,
            type,
            status: "VALID",
        };
        if (tradetrust_1.utils.isWrappedV3Document(document)) {
            return Object.assign(Object.assign({}, fragment), { data: { mintedOnAll: true, details: mintStatus } });
        }
        else {
            return Object.assign(Object.assign({}, fragment), { data: { mintedOnAll: true, details: [mintStatus] } });
        }
    }
    else {
        const fragment = {
            name,
            type,
            reason: mintStatus.reason,
            status: "INVALID",
        };
        if (tradetrust_1.utils.isWrappedV3Document(document)) {
            return Object.assign(Object.assign({}, fragment), { data: { mintedOnAll: false, details: mintStatus } });
        }
        else {
            return Object.assign(Object.assign({}, fragment), { data: { mintedOnAll: false, details: [mintStatus] } });
        }
    }
});
exports.openAttestationEthereumTokenRegistryStatus = {
    skip,
    test,
    verify: (0, errorHandler_1.withCodedErrorHandler)(verify, {
        name,
        type,
        unexpectedErrorCode: error_1.OpenAttestationEthereumTokenRegistryStatusCode.UNEXPECTED_ERROR,
        unexpectedErrorString: error_1.OpenAttestationEthereumTokenRegistryStatusCode[error_1.OpenAttestationEthereumTokenRegistryStatusCode.UNEXPECTED_ERROR],
    }),
};
