// @trustvc/trustvc/dist/cjs/verify/verify.js
'use strict';

var ttVerify = require('@tradetrust-tt/tt-verify');
var w3cSignatureIntegrity = require('./fragments/document-integrity/w3cSignatureIntegrity');
var ecdsaW3CSignatureIntegrity = require('./fragments/document-integrity/ecdsaW3CSignatureIntegrity');
var transferableRecordVerifier = require('./fragments/document-status/transferableRecords/transferableRecordVerifier');
var w3cCredentialStatus = require('./fragments/document-status/w3cCredentialStatus');
var w3cIssuerIdentity = require('./fragments/issuer-identity/w3cIssuerIdentity');
var fragments = require('./fragments');
var bbs2023W3CSignatureIntegrity = require('./fragments/document-integrity/bbs2023W3CSignatureIntegrity');
var openCert = require('../open-cert');

const verifiers = {
  documentIntegrity: {
    openAttestationHash: ttVerify.openAttestationHash,
    w3cSignatureIntegrity: w3cSignatureIntegrity.w3cSignatureIntegrity
  },
  documentStatus: {
    openAttestationDidSignedDocumentStatus: ttVerify.openAttestationDidSignedDocumentStatus,
    openAttestationEthereumDocumentStoreStatus: ttVerify.openAttestationEthereumDocumentStoreStatus,
    openAttestationEthereumTokenRegistryStatus: ttVerify.openAttestationEthereumTokenRegistryStatus,
    w3cCredentialStatus: w3cCredentialStatus.w3cCredentialStatus,
    w3cEmptyCredentialStatus: fragments.w3cEmptyCredentialStatus,
    credentialStatusTransferableRecordVerifier: transferableRecordVerifier.credentialStatusTransferableRecordVerifier
  },
  issuerIdentity: {
    openAttestationDnsDidIdentityProof: ttVerify.openAttestationDnsDidIdentityProof,
    openAttestationDnsTxtIdentityProof: ttVerify.openAttestationDnsTxtIdentityProof,
    w3cIssuerIdentity: w3cIssuerIdentity.w3cIssuerIdentity
  }
};
const openAttestationVerifiers = [...ttVerify.openAttestationVerifiers, openCert.registryVerifier];
const w3cVerifiers = [
  w3cSignatureIntegrity.w3cSignatureIntegrity,
  ecdsaW3CSignatureIntegrity.ecdsaW3CSignatureIntegrity,
  bbs2023W3CSignatureIntegrity.bbs2023W3CSignatureIntegrity,
  w3cCredentialStatus.w3cCredentialStatus,
  transferableRecordVerifier.credentialStatusTransferableRecordVerifier,
  fragments.w3cEmptyCredentialStatus,
  w3cIssuerIdentity.w3cIssuerIdentity
];

Object.defineProperty(exports, "createResolver", {
  enumerable: true,
  get: function () { return ttVerify.createResolver; }
});
Object.defineProperty(exports, "getIdentifier", {
  enumerable: true,
  get: function () { return ttVerify.getIdentifier; }
});
Object.defineProperty(exports, "isValid", {
  enumerable: true,
  get: function () { return ttVerify.isValid; }
});
Object.defineProperty(exports, "utils", {
  enumerable: true,
  get: function () { return ttVerify.utils; }
});
Object.defineProperty(exports, "verificationBuilder", {
  enumerable: true,
  get: function () { return ttVerify.verificationBuilder; }
});
Object.defineProperty(exports, "verify", {
  enumerable: true,
  get: function () { return ttVerify.verify; }
});
exports.openAttestationVerifiers = openAttestationVerifiers;
exports.verifiers = verifiers;
exports.w3cVerifiers = w3cVerifiers;
