// @trustvc/trustvc/dist/cjs/document-store/index.js
'use strict';

var issue = require('./issue');
var revoke = require('./revoke');
var revokeRole = require('./revoke-role');
var grantRole = require('./grant-role');
var transferOwnership = require('./transferOwnership');
var documentStore$1 = require('../deploy/document-store');
var documentStoreRoles = require('./document-store-roles');
var supportInterfaceIds = require('./supportInterfaceIds');
var documentStore = require('@trustvc/document-store');



Object.defineProperty(exports, "documentStoreIssue", {
  enumerable: true,
  get: function () { return issue.documentStoreIssue; }
});
Object.defineProperty(exports, "documentStoreRevoke", {
  enumerable: true,
  get: function () { return revoke.documentStoreRevoke; }
});
Object.defineProperty(exports, "documentStoreRevokeRole", {
  enumerable: true,
  get: function () { return revokeRole.documentStoreRevokeRole; }
});
Object.defineProperty(exports, "documentStoreGrantRole", {
  enumerable: true,
  get: function () { return grantRole.documentStoreGrantRole; }
});
Object.defineProperty(exports, "documentStoreTransferOwnership", {
  enumerable: true,
  get: function () { return transferOwnership.documentStoreTransferOwnership; }
});
Object.defineProperty(exports, "deployDocumentStore", {
  enumerable: true,
  get: function () { return documentStore$1.deployDocumentStore; }
});
Object.defineProperty(exports, "getRoleString", {
  enumerable: true,
  get: function () { return documentStoreRoles.getRoleString; }
});
Object.defineProperty(exports, "supportInterfaceIds", {
  enumerable: true,
  get: function () { return supportInterfaceIds.supportInterfaceIds; }
});
Object.defineProperty(exports, "DocumentStore__factory", {
  enumerable: true,
  get: function () { return documentStore.DocumentStore__factory; }
});
Object.defineProperty(exports, "TransferableDocumentStore__factory", {
  enumerable: true,
  get: function () { return documentStore.TransferableDocumentStore__factory; }
});
