// @tradetrust-tt/tt-verify/dist/cjs/verifiers/issuerIdentity/dnsDid/dnsDidProof.type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DnsDidVerificationStatusArray = exports.DnsDidVerificationStatus = exports.InvalidDnsDidVerificationStatus = exports.ValidDnsDidVerificationStatusArray = exports.ValidDnsDidVerificationStatus = void 0;
const runtypes_1 = require("runtypes");
/**
 * DNS-DID verification status
 */
exports.ValidDnsDidVerificationStatus = (0, runtypes_1.Record)({
    status: (0, runtypes_1.Literal)("VALID"),
    location: runtypes_1.String,
    key: runtypes_1.String,
});
exports.ValidDnsDidVerificationStatusArray = (0, runtypes_1.Array)(exports.ValidDnsDidVerificationStatus).withConstraint((elements) => elements.length > 0 || "Expect at least one valid element");
exports.InvalidDnsDidVerificationStatus = (0, runtypes_1.Record)({
    status: (0, runtypes_1.Literal)("INVALID"),
    location: runtypes_1.String,
    key: runtypes_1.String,
});
exports.DnsDidVerificationStatus = (0, runtypes_1.Union)(exports.ValidDnsDidVerificationStatus, exports.InvalidDnsDidVerificationStatus);
exports.DnsDidVerificationStatusArray = (0, runtypes_1.Array)(exports.DnsDidVerificationStatus);
