// @tradetrust-tt/tt-verify/dist/cjs/verifiers/issuerIdentity/dnsDid/dnsDidProof.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.openAttestationDnsDidIdentityProof = void 0;
const tradetrust_1 = require("@tradetrust-tt/tradetrust");
const dnsprove_1 = require("@tradetrust-tt/dnsprove");
const error_1 = require("../../../types/error");
const errorHandler_1 = require("../../../common/errorHandler");
const error_2 = require("../../../common/error");
const dnsDidProof_type_1 = require("./dnsDidProof.type");
const name = "OpenAttestationDnsDidIdentityProof";
const type = "ISSUER_IDENTITY";
const skip = () => __awaiter(void 0, void 0, void 0, function* () {
    return {
        status: "SKIPPED",
        type,
        name,
        reason: {
            code: error_1.OpenAttestationDnsDidCode.SKIPPED,
            codeString: error_1.OpenAttestationDnsDidCode[error_1.OpenAttestationDnsDidCode.SKIPPED],
            message: `Document was not issued using DNS-DID`,
        },
    };
});
const test = (document) => {
    if (tradetrust_1.utils.isSignedWrappedV2Document(document)) {
        const data = (0, tradetrust_1.getData)(document);
        return data.issuers.some((issuer) => { var _a; return ((_a = issuer.identityProof) === null || _a === void 0 ? void 0 : _a.type) === "DNS-DID"; });
    }
    else if (tradetrust_1.utils.isSignedWrappedV3Document(document)) {
        return document.openAttestationMetadata.identityProof.type === tradetrust_1.v3.IdentityProofType.DNSDid;
    }
    return false;
};
const verifyIssuerDnsDid = ({ key, location, }) => __awaiter(void 0, void 0, void 0, function* () {
    const records = yield (0, dnsprove_1.getDnsDidRecords)(location);
    return {
        location,
        key,
        status: records.some((record) => record.publicKey.toLowerCase() === key.toLowerCase()) ? "VALID" : "INVALID",
    };
});
const verifyV2 = (document) => __awaiter(void 0, void 0, void 0, function* () {
    const documentData = (0, tradetrust_1.getData)(document);
    const deferredVerificationStatus = documentData.issuers.map((issuer) => {
        const { identityProof } = issuer;
        if (!identityProof)
            throw new error_2.CodedError("Identity proof missing", error_1.OpenAttestationDnsDidCode.MALFORMED_IDENTITY_PROOF, error_1.OpenAttestationDnsDidCode[error_1.OpenAttestationDnsDidCode.MALFORMED_IDENTITY_PROOF]);
        const { key, location, type: identityProofType } = identityProof;
        if (identityProofType !== tradetrust_1.v2.IdentityProofType.DNSDid)
            throw new error_2.CodedError("Issuer is not using DID-DNS identityProof type", error_1.OpenAttestationDnsDidCode.INVALID_ISSUERS, error_1.OpenAttestationDnsDidCode[error_1.OpenAttestationDnsDidCode.INVALID_ISSUERS]);
        if (!location)
            throw new error_2.CodedError("location is not present in identity proof", error_1.OpenAttestationDnsDidCode.MALFORMED_IDENTITY_PROOF, error_1.OpenAttestationDnsDidCode[error_1.OpenAttestationDnsDidCode.MALFORMED_IDENTITY_PROOF]);
        if (!key)
            throw new error_2.CodedError("key is not present in identity proof", error_1.OpenAttestationDnsDidCode.MALFORMED_IDENTITY_PROOF, error_1.OpenAttestationDnsDidCode[error_1.OpenAttestationDnsDidCode.MALFORMED_IDENTITY_PROOF]);
        return verifyIssuerDnsDid({ key, location });
    });
    const verificationStatus = yield Promise.all(deferredVerificationStatus);
    if (dnsDidProof_type_1.ValidDnsDidVerificationStatusArray.guard(verificationStatus)) {
        return {
            name,
            type,
            data: verificationStatus,
            status: "VALID",
        };
    }
    return {
        name,
        type,
        data: verificationStatus,
        reason: {
            message: "Could not find identity at location",
            code: error_1.OpenAttestationDnsDidCode.INVALID_IDENTITY,
            codeString: "INVALID_IDENTITY",
        },
        status: "INVALID",
    };
});
const verifyV3 = (document) => __awaiter(void 0, void 0, void 0, function* () {
    if (!tradetrust_1.utils.isSignedWrappedV3Document(document))
        throw new error_2.CodedError("document is not signed", error_1.OpenAttestationDnsDidCode.UNSIGNED, error_1.OpenAttestationDnsDidCode[error_1.OpenAttestationDnsDidCode.UNSIGNED]);
    const location = document.openAttestationMetadata.identityProof.identifier;
    const { key } = document.proof;
    const verificationStatus = yield verifyIssuerDnsDid({ key, location });
    if (dnsDidProof_type_1.ValidDnsDidVerificationStatus.guard(verificationStatus)) {
        return {
            name,
            type,
            data: verificationStatus,
            status: "VALID",
        };
    }
    return {
        name,
        type,
        data: verificationStatus,
        status: "INVALID",
        reason: {
            message: "Could not find identity at location",
            code: error_1.OpenAttestationDnsDidCode.INVALID_IDENTITY,
            codeString: "INVALID_IDENTITY",
        },
    };
});
const verify = (document) => __awaiter(void 0, void 0, void 0, function* () {
    if (tradetrust_1.utils.isSignedWrappedV2Document(document))
        return verifyV2(document);
    else if (tradetrust_1.utils.isSignedWrappedV3Document(document))
        return verifyV3(document);
    throw new error_2.CodedError("Document does not match either v2 or v3 formats. Consider using `utils.diagnose` from open-attestation to find out more.", error_1.OpenAttestationDnsDidCode.UNRECOGNIZED_DOCUMENT, error_1.OpenAttestationDnsDidCode[error_1.OpenAttestationDnsDidCode.UNRECOGNIZED_DOCUMENT]);
});
exports.openAttestationDnsDidIdentityProof = {
    skip,
    test,
    verify: (0, errorHandler_1.withCodedErrorHandler)(verify, {
        name,
        type,
        unexpectedErrorCode: error_1.OpenAttestationDnsDidCode.UNEXPECTED_ERROR,
        unexpectedErrorString: error_1.OpenAttestationDnsDidCode[error_1.OpenAttestationDnsDidCode.UNEXPECTED_ERROR],
    }),
};
