// @trustvc/w3c-issuer/dist/did-web/keyPair/bbs2023.js
'use strict';

var Bls12381Multikey = require('@digitalbazaar/bls12-381-multikey');
var base58 = require('multiformats/bases/base58');
var types = require('../../lib/types');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var Bls12381Multikey__namespace = /*#__PURE__*/_interopNamespace(Bls12381Multikey);

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const generateBbs2023KeyPair = /* @__PURE__ */ __name(async ({ seed }) => {
  if (!seed) {
    throw new Error("Invalid seed");
  }
  const bbsKeyPair = await Bls12381Multikey__namespace.generateBbsKeyPair({
    algorithm: "BBS-BLS12-381-SHA-256",
    seed
  });
  const exportedKeys = await bbsKeyPair.export({
    publicKey: true,
    secretKey: true
  });
  const keyPair = {
    type: types.VerificationType.Multikey,
    seedBase58: base58.base58btc.encode(seed).slice(1),
    secretKeyMultibase: exportedKeys.secretKeyMultibase,
    publicKeyMultibase: exportedKeys.publicKeyMultibase
  };
  return keyPair;
}, "generateBbs2023KeyPair");

exports.generateBbs2023KeyPair = generateBbs2023KeyPair;
