// @trustvc/w3c-issuer/dist/did-web/keyPair/ecdsaSd2023.js
'use strict';

var EcdsaMultikey = require('@digitalbazaar/ecdsa-multikey');
var types = require('../../lib/types');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var EcdsaMultikey__namespace = /*#__PURE__*/_interopNamespace(EcdsaMultikey);

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const generateEcdsaSd2023KeyPair = /* @__PURE__ */ __name(async () => {
  const ecdsaKeyPair = await EcdsaMultikey__namespace.generate({
    curve: "P-256"
  });
  const exportedKeys = await ecdsaKeyPair.export({
    publicKey: true,
    secretKey: true
  });
  const keyPair = {
    type: types.VerificationType.Multikey,
    secretKeyMultibase: exportedKeys.secretKeyMultibase,
    publicKeyMultibase: exportedKeys.publicKeyMultibase
  };
  return keyPair;
}, "generateEcdsaSd2023KeyPair");

exports.generateEcdsaSd2023KeyPair = generateEcdsaSd2023KeyPair;
