// @tradetrust-tt/tt-verify/dist/cjs/verifiers/issuerIdentity/did/didIdentityProof.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.openAttestationDidIdentityProof = void 0;
const tradetrust_1 = require("@tradetrust-tt/tradetrust");
const error_1 = require("../../../types/error");
const verifier_1 = require("../../../did/verifier");
const errorHandler_1 = require("../../../common/errorHandler");
const error_2 = require("../../../common/error");
const name = "OpenAttestationDidIdentityProof";
const type = "ISSUER_IDENTITY";
const skip = () => __awaiter(void 0, void 0, void 0, function* () {
    return {
        status: "SKIPPED",
        type,
        name,
        reason: {
            code: error_1.OpenAttestationDidCode.SKIPPED,
            codeString: error_1.OpenAttestationDidCode[error_1.OpenAttestationDidCode.SKIPPED],
            message: `Document is not using DID as top level identifier or has not been wrapped`,
        },
    };
});
const test = (document) => {
    if (tradetrust_1.utils.isWrappedV2Document(document)) {
        const { issuers } = (0, tradetrust_1.getData)(document);
        return issuers.some((issuer) => { var _a; return ((_a = issuer.identityProof) === null || _a === void 0 ? void 0 : _a.type) === tradetrust_1.v2.IdentityProofType.Did; });
    }
    else if (tradetrust_1.utils.isWrappedV3Document(document)) {
        return document.openAttestationMetadata.identityProof.type === tradetrust_1.v3.IdentityProofType.Did;
    }
    return false;
};
const verifyV2 = (document, options) => __awaiter(void 0, void 0, void 0, function* () {
    if (!tradetrust_1.utils.isSignedWrappedV2Document(document))
        throw new error_2.CodedError("Document is not signed", error_1.OpenAttestationDidCode.UNSIGNED, "UNSIGNED");
    const data = (0, tradetrust_1.getData)(document);
    const merkleRoot = `0x${document.signature.merkleRoot}`;
    const signatureVerificationDeferred = data.issuers.map((issuer) => __awaiter(void 0, void 0, void 0, function* () {
        var _a, _b;
        if (((_a = issuer.identityProof) === null || _a === void 0 ? void 0 : _a.type) === "DID") {
            const did = issuer.id;
            if (!did)
                throw new error_2.CodedError("id is missing in issuer", error_1.OpenAttestationDidCode.DID_MISSING, "DID_MISSING");
            const key = (_b = issuer.identityProof) === null || _b === void 0 ? void 0 : _b.key;
            if (!key)
                throw new error_2.CodedError("Key is not present", error_1.OpenAttestationDidCode.MALFORMED_IDENTITY_PROOF, "MALFORMED_IDENTITY_PROOF");
            const correspondingProof = document.proof.find((p) => p.verificationMethod.toLowerCase() === key.toLowerCase());
            if (!correspondingProof)
                throw new error_2.CodedError(`No proof for ${key}`, error_1.OpenAttestationDidCode.MALFORMED_IDENTITY_PROOF, "MALFORMED_IDENTITY_PROOF");
            return (0, verifier_1.verifySignature)({
                merkleRoot,
                key,
                signature: correspondingProof.signature,
                did,
                resolver: options.resolver,
            });
        }
        throw new error_2.CodedError("Issuer is not using DID identityProof type", error_1.OpenAttestationDidCode.INVALID_ISSUERS, error_1.OpenAttestationDidCode[error_1.OpenAttestationDidCode.INVALID_ISSUERS]);
    }));
    const signatureVerifications = yield Promise.all(signatureVerificationDeferred);
    if (verifier_1.ValidDidVerificationStatusArray.guard(signatureVerifications)) {
        return {
            name,
            type,
            data: signatureVerifications,
            status: "VALID",
        };
    }
    const invalidSignature = signatureVerifications.find(verifier_1.InvalidDidVerificationStatus.guard);
    if (verifier_1.InvalidDidVerificationStatus.guard(invalidSignature)) {
        return {
            name,
            type,
            data: signatureVerifications,
            reason: invalidSignature.reason,
            status: "INVALID",
        };
    }
    throw new error_2.CodedError("Unable to retrieve the reason of the failure", error_1.OpenAttestationDidCode.UNEXPECTED_ERROR, "UNEXPECTED_ERROR");
});
const verifyV3 = (document, options) => __awaiter(void 0, void 0, void 0, function* () {
    if (!tradetrust_1.utils.isSignedWrappedV3Document(document))
        throw new error_2.CodedError("Document is not signed", error_1.OpenAttestationDidCode.UNSIGNED, "UNSIGNED");
    const merkleRoot = `0x${document.proof.merkleRoot}`;
    const { key, signature } = document.proof;
    const did = document.openAttestationMetadata.identityProof.identifier;
    const verificationStatus = yield (0, verifier_1.verifySignature)({
        did,
        merkleRoot,
        key,
        signature,
        resolver: options.resolver,
    });
    if (verifier_1.ValidDidVerificationStatus.guard(verificationStatus)) {
        return {
            name,
            type,
            data: verificationStatus,
            status: "VALID",
        };
    }
    return {
        name,
        type,
        data: verificationStatus,
        reason: verificationStatus.reason,
        status: "INVALID",
    };
});
const verify = (document, options) => __awaiter(void 0, void 0, void 0, function* () {
    if (tradetrust_1.utils.isWrappedV2Document(document))
        return verifyV2(document, options);
    else if (tradetrust_1.utils.isWrappedV3Document(document))
        return verifyV3(document, options);
    throw new error_2.CodedError("Document does not match either v2 or v3 formats. Consider using `utils.diagnose` from open-attestation to find out more.", error_1.OpenAttestationDidCode.UNRECOGNIZED_DOCUMENT, error_1.OpenAttestationDidCode[error_1.OpenAttestationDidCode.UNRECOGNIZED_DOCUMENT]);
});
exports.openAttestationDidIdentityProof = {
    skip,
    test,
    verify: (0, errorHandler_1.withCodedErrorHandler)(verify, {
        name,
        type,
        unexpectedErrorCode: error_1.OpenAttestationDidCode.UNEXPECTED_ERROR,
        unexpectedErrorString: error_1.OpenAttestationDidCode[error_1.OpenAttestationDidCode.UNEXPECTED_ERROR],
    }),
};
