// @tradetrust-tt/token-registry-v4/dist/contracts/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AccessControlUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _AccessControlUpgradeable__factory.AccessControlUpgradeable__factory;
  }
});
Object.defineProperty(exports, "ERC165Upgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _ERC165Upgradeable__factory.ERC165Upgradeable__factory;
  }
});
Object.defineProperty(exports, "ERC1967UpgradeUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _ERC1967UpgradeUpgradeable__factory.ERC1967UpgradeUpgradeable__factory;
  }
});
Object.defineProperty(exports, "ERC721ReceiverMock__factory", {
  enumerable: true,
  get: function get() {
    return _ERC721ReceiverMock__factory.ERC721ReceiverMock__factory;
  }
});
Object.defineProperty(exports, "IAccessControlUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _IAccessControlUpgradeable__factory.IAccessControlUpgradeable__factory;
  }
});
Object.defineProperty(exports, "IBeaconUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _IBeaconUpgradeable__factory.IBeaconUpgradeable__factory;
  }
});
Object.defineProperty(exports, "IERC165Upgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _IERC165Upgradeable__factory.IERC165Upgradeable__factory;
  }
});
Object.defineProperty(exports, "IERC165__factory", {
  enumerable: true,
  get: function get() {
    return _IERC165__factory.IERC165__factory;
  }
});
Object.defineProperty(exports, "IERC1822ProxiableUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _IERC1822ProxiableUpgradeable__factory.IERC1822ProxiableUpgradeable__factory;
  }
});
Object.defineProperty(exports, "IERC721MetadataUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _IERC721MetadataUpgradeable__factory.IERC721MetadataUpgradeable__factory;
  }
});
Object.defineProperty(exports, "IERC721ReceiverUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _IERC721ReceiverUpgradeable__factory.IERC721ReceiverUpgradeable__factory;
  }
});
Object.defineProperty(exports, "IERC721Receiver__factory", {
  enumerable: true,
  get: function get() {
    return _IERC721Receiver__factory.IERC721Receiver__factory;
  }
});
Object.defineProperty(exports, "ISBTUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _ISBTUpgradeable__factory.ISBTUpgradeable__factory;
  }
});
Object.defineProperty(exports, "ITitleEscrowFactory__factory", {
  enumerable: true,
  get: function get() {
    return _ITitleEscrowFactory__factory.ITitleEscrowFactory__factory;
  }
});
Object.defineProperty(exports, "ITitleEscrowSignable__factory", {
  enumerable: true,
  get: function get() {
    return _ITitleEscrowSignable__factory.ITitleEscrowSignable__factory;
  }
});
Object.defineProperty(exports, "ITitleEscrow__factory", {
  enumerable: true,
  get: function get() {
    return _ITitleEscrow__factory.ITitleEscrow__factory;
  }
});
Object.defineProperty(exports, "ITradeTrustSBT__factory", {
  enumerable: true,
  get: function get() {
    return _ITradeTrustSBT__factory.ITradeTrustSBT__factory;
  }
});
Object.defineProperty(exports, "ITradeTrustTokenBurnable__factory", {
  enumerable: true,
  get: function get() {
    return _ITradeTrustTokenBurnable__factory.ITradeTrustTokenBurnable__factory;
  }
});
Object.defineProperty(exports, "ITradeTrustTokenMintable__factory", {
  enumerable: true,
  get: function get() {
    return _ITradeTrustTokenMintable__factory.ITradeTrustTokenMintable__factory;
  }
});
Object.defineProperty(exports, "ITradeTrustTokenRestorable__factory", {
  enumerable: true,
  get: function get() {
    return _ITradeTrustTokenRestorable__factory.ITradeTrustTokenRestorable__factory;
  }
});
Object.defineProperty(exports, "ITradeTrustToken__factory", {
  enumerable: true,
  get: function get() {
    return _ITradeTrustToken__factory.ITradeTrustToken__factory;
  }
});
Object.defineProperty(exports, "OwnableUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _OwnableUpgradeable__factory.OwnableUpgradeable__factory;
  }
});
Object.defineProperty(exports, "PausableUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _PausableUpgradeable__factory.PausableUpgradeable__factory;
  }
});
Object.defineProperty(exports, "Pausable__factory", {
  enumerable: true,
  get: function get() {
    return _Pausable__factory.Pausable__factory;
  }
});
Object.defineProperty(exports, "RegistryAccessErrors__factory", {
  enumerable: true,
  get: function get() {
    return _RegistryAccessErrors__factory.RegistryAccessErrors__factory;
  }
});
Object.defineProperty(exports, "RegistryAccess__factory", {
  enumerable: true,
  get: function get() {
    return _RegistryAccess__factory.RegistryAccess__factory;
  }
});
Object.defineProperty(exports, "SBTUpgradeableMock__factory", {
  enumerable: true,
  get: function get() {
    return _SBTUpgradeableMock__factory.SBTUpgradeableMock__factory;
  }
});
Object.defineProperty(exports, "SBTUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _SBTUpgradeable__factory.SBTUpgradeable__factory;
  }
});
Object.defineProperty(exports, "SigHelperErrors__factory", {
  enumerable: true,
  get: function get() {
    return _SigHelperErrors__factory.SigHelperErrors__factory;
  }
});
Object.defineProperty(exports, "SigHelperMock__factory", {
  enumerable: true,
  get: function get() {
    return _SigHelperMock__factory.SigHelperMock__factory;
  }
});
Object.defineProperty(exports, "SigHelper__factory", {
  enumerable: true,
  get: function get() {
    return _SigHelper__factory.SigHelper__factory;
  }
});
Object.defineProperty(exports, "SimpleCaller__factory", {
  enumerable: true,
  get: function get() {
    return _SimpleCaller__factory.SimpleCaller__factory;
  }
});
Object.defineProperty(exports, "TDocDeployerErrors__factory", {
  enumerable: true,
  get: function get() {
    return _TDocDeployerErrors__factory.TDocDeployerErrors__factory;
  }
});
Object.defineProperty(exports, "TDocDeployer__factory", {
  enumerable: true,
  get: function get() {
    return _TDocDeployer__factory.TDocDeployer__factory;
  }
});
Object.defineProperty(exports, "TitleEscrowErrors__factory", {
  enumerable: true,
  get: function get() {
    return _TitleEscrowErrors__factory.TitleEscrowErrors__factory;
  }
});
Object.defineProperty(exports, "TitleEscrowFactoryCallerMock__factory", {
  enumerable: true,
  get: function get() {
    return _TitleEscrowFactoryCallerMock__factory.TitleEscrowFactoryCallerMock__factory;
  }
});
Object.defineProperty(exports, "TitleEscrowFactoryErrors__factory", {
  enumerable: true,
  get: function get() {
    return _TitleEscrowFactoryErrors__factory.TitleEscrowFactoryErrors__factory;
  }
});
Object.defineProperty(exports, "TitleEscrowFactoryGetterMock__factory", {
  enumerable: true,
  get: function get() {
    return _TitleEscrowFactoryGetterMock__factory.TitleEscrowFactoryGetterMock__factory;
  }
});
Object.defineProperty(exports, "TitleEscrowFactory__factory", {
  enumerable: true,
  get: function get() {
    return _TitleEscrowFactory__factory.TitleEscrowFactory__factory;
  }
});
Object.defineProperty(exports, "TitleEscrowMock__factory", {
  enumerable: true,
  get: function get() {
    return _TitleEscrowMock__factory.TitleEscrowMock__factory;
  }
});
Object.defineProperty(exports, "TitleEscrowSignableErrors__factory", {
  enumerable: true,
  get: function get() {
    return _TitleEscrowSignableErrors__factory.TitleEscrowSignableErrors__factory;
  }
});
Object.defineProperty(exports, "TitleEscrowSignable__factory", {
  enumerable: true,
  get: function get() {
    return _TitleEscrowSignable__factory.TitleEscrowSignable__factory;
  }
});
Object.defineProperty(exports, "TitleEscrow__factory", {
  enumerable: true,
  get: function get() {
    return _TitleEscrow__factory.TitleEscrow__factory;
  }
});
Object.defineProperty(exports, "TradeTrustSBT__factory", {
  enumerable: true,
  get: function get() {
    return _TradeTrustSBT__factory.TradeTrustSBT__factory;
  }
});
Object.defineProperty(exports, "TradeTrustTokenBaseURI__factory", {
  enumerable: true,
  get: function get() {
    return _TradeTrustTokenBaseURI__factory.TradeTrustTokenBaseURI__factory;
  }
});
Object.defineProperty(exports, "TradeTrustTokenBase__factory", {
  enumerable: true,
  get: function get() {
    return _TradeTrustTokenBase__factory.TradeTrustTokenBase__factory;
  }
});
Object.defineProperty(exports, "TradeTrustTokenBurnable__factory", {
  enumerable: true,
  get: function get() {
    return _TradeTrustTokenBurnable__factory.TradeTrustTokenBurnable__factory;
  }
});
Object.defineProperty(exports, "TradeTrustTokenErrors__factory", {
  enumerable: true,
  get: function get() {
    return _TradeTrustTokenErrors__factory.TradeTrustTokenErrors__factory;
  }
});
Object.defineProperty(exports, "TradeTrustTokenMintable__factory", {
  enumerable: true,
  get: function get() {
    return _TradeTrustTokenMintable__factory.TradeTrustTokenMintable__factory;
  }
});
Object.defineProperty(exports, "TradeTrustTokenMock__factory", {
  enumerable: true,
  get: function get() {
    return _TradeTrustTokenMock__factory.TradeTrustTokenMock__factory;
  }
});
Object.defineProperty(exports, "TradeTrustTokenRestorable__factory", {
  enumerable: true,
  get: function get() {
    return _TradeTrustTokenRestorable__factory.TradeTrustTokenRestorable__factory;
  }
});
Object.defineProperty(exports, "TradeTrustTokenStandard__factory", {
  enumerable: true,
  get: function get() {
    return _TradeTrustTokenStandard__factory.TradeTrustTokenStandard__factory;
  }
});
Object.defineProperty(exports, "TradeTrustToken__factory", {
  enumerable: true,
  get: function get() {
    return _TradeTrustToken__factory.TradeTrustToken__factory;
  }
});
Object.defineProperty(exports, "UUPSUpgradeable__factory", {
  enumerable: true,
  get: function get() {
    return _UUPSUpgradeable__factory.UUPSUpgradeable__factory;
  }
});
exports.factories = void 0;

var _factories = _interopRequireWildcard(require("./factories"));

exports.factories = _factories;

var _AccessControlUpgradeable__factory = require("./factories/@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable__factory");

var _IAccessControlUpgradeable__factory = require("./factories/@openzeppelin/contracts-upgradeable/access/IAccessControlUpgradeable__factory");

var _OwnableUpgradeable__factory = require("./factories/@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable__factory");

var _IERC1822ProxiableUpgradeable__factory = require("./factories/@openzeppelin/contracts-upgradeable/interfaces/draft-IERC1822Upgradeable.sol/IERC1822ProxiableUpgradeable__factory");

var _IBeaconUpgradeable__factory = require("./factories/@openzeppelin/contracts-upgradeable/proxy/beacon/IBeaconUpgradeable__factory");

var _ERC1967UpgradeUpgradeable__factory = require("./factories/@openzeppelin/contracts-upgradeable/proxy/ERC1967/ERC1967UpgradeUpgradeable__factory");

var _UUPSUpgradeable__factory = require("./factories/@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable__factory");

var _PausableUpgradeable__factory = require("./factories/@openzeppelin/contracts-upgradeable/security/PausableUpgradeable__factory");

var _IERC721ReceiverUpgradeable__factory = require("./factories/@openzeppelin/contracts-upgradeable/token/ERC721/IERC721ReceiverUpgradeable__factory");

var _ERC165Upgradeable__factory = require("./factories/@openzeppelin/contracts-upgradeable/utils/introspection/ERC165Upgradeable__factory");

var _IERC165Upgradeable__factory = require("./factories/@openzeppelin/contracts-upgradeable/utils/introspection/IERC165Upgradeable__factory");

var _Pausable__factory = require("./factories/@openzeppelin/contracts/security/Pausable__factory");

var _IERC721Receiver__factory = require("./factories/@openzeppelin/contracts/token/ERC721/IERC721Receiver__factory");

var _IERC165__factory = require("./factories/@openzeppelin/contracts/utils/introspection/IERC165__factory");

var _RegistryAccess__factory = require("./factories/contracts/base/RegistryAccess__factory");

var _SBTUpgradeable__factory = require("./factories/contracts/base/SBTUpgradeable__factory");

var _TradeTrustSBT__factory = require("./factories/contracts/base/TradeTrustSBT__factory");

var _TradeTrustTokenBase__factory = require("./factories/contracts/base/TradeTrustTokenBase__factory");

var _TradeTrustTokenBaseURI__factory = require("./factories/contracts/base/TradeTrustTokenBaseURI__factory");

var _TradeTrustTokenBurnable__factory = require("./factories/contracts/base/TradeTrustTokenBurnable__factory");

var _TradeTrustTokenMintable__factory = require("./factories/contracts/base/TradeTrustTokenMintable__factory");

var _TradeTrustTokenRestorable__factory = require("./factories/contracts/base/TradeTrustTokenRestorable__factory");

var _IERC721MetadataUpgradeable__factory = require("./factories/contracts/interfaces/IERC721MetadataUpgradeable__factory");

var _ISBTUpgradeable__factory = require("./factories/contracts/interfaces/ISBTUpgradeable__factory");

var _ITitleEscrow__factory = require("./factories/contracts/interfaces/ITitleEscrow__factory");

var _ITitleEscrowFactory__factory = require("./factories/contracts/interfaces/ITitleEscrowFactory__factory");

var _ITitleEscrowSignable__factory = require("./factories/contracts/interfaces/ITitleEscrowSignable__factory");

var _ITradeTrustSBT__factory = require("./factories/contracts/interfaces/ITradeTrustSBT__factory");

var _ITradeTrustToken__factory = require("./factories/contracts/interfaces/ITradeTrustToken__factory");

var _ITradeTrustTokenBurnable__factory = require("./factories/contracts/interfaces/ITradeTrustTokenBurnable__factory");

var _ITradeTrustTokenMintable__factory = require("./factories/contracts/interfaces/ITradeTrustTokenMintable__factory");

var _ITradeTrustTokenRestorable__factory = require("./factories/contracts/interfaces/ITradeTrustTokenRestorable__factory");

var _RegistryAccessErrors__factory = require("./factories/contracts/interfaces/RegistryAccessErrors__factory");

var _SigHelperErrors__factory = require("./factories/contracts/interfaces/SigHelperErrors__factory");

var _TDocDeployerErrors__factory = require("./factories/contracts/interfaces/TDocDeployerErrors__factory");

var _TitleEscrowErrors__factory = require("./factories/contracts/interfaces/TitleEscrowErrors__factory");

var _TitleEscrowFactoryErrors__factory = require("./factories/contracts/interfaces/TitleEscrowFactoryErrors__factory");

var _TitleEscrowSignableErrors__factory = require("./factories/contracts/interfaces/TitleEscrowSignableErrors__factory");

var _TradeTrustTokenErrors__factory = require("./factories/contracts/interfaces/TradeTrustTokenErrors__factory");

var _ERC721ReceiverMock__factory = require("./factories/contracts/mocks/ERC721ReceiverMock__factory");

var _SBTUpgradeableMock__factory = require("./factories/contracts/mocks/SBTUpgradeableMock__factory");

var _SigHelperMock__factory = require("./factories/contracts/mocks/SigHelperMock__factory");

var _SimpleCaller__factory = require("./factories/contracts/mocks/SimpleCaller__factory");

var _TitleEscrowFactoryCallerMock__factory = require("./factories/contracts/mocks/TitleEscrowFactoryCallerMock__factory");

var _TitleEscrowFactoryGetterMock__factory = require("./factories/contracts/mocks/TitleEscrowFactoryGetterMock__factory");

var _TitleEscrowMock__factory = require("./factories/contracts/mocks/TitleEscrowMock__factory");

var _TradeTrustTokenMock__factory = require("./factories/contracts/mocks/TradeTrustTokenMock__factory");

var _TitleEscrowSignable__factory = require("./factories/contracts/presets/TitleEscrowSignable__factory");

var _TradeTrustTokenStandard__factory = require("./factories/contracts/presets/TradeTrustTokenStandard__factory");

var _TitleEscrow__factory = require("./factories/contracts/TitleEscrow__factory");

var _TitleEscrowFactory__factory = require("./factories/contracts/TitleEscrowFactory__factory");

var _TradeTrustToken__factory = require("./factories/contracts/TradeTrustToken__factory");

var _SigHelper__factory = require("./factories/contracts/utils/SigHelper__factory");

var _TDocDeployer__factory = require("./factories/contracts/utils/TDocDeployer__factory");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
