// @trustvc/trustvc/dist/cjs/token-registry-v4/index.js
'use strict';

var roleHash = require('./roleHash');
var contractAddress = require('./contractAddress');
var supportInterfaceIds = require('./supportInterfaceIds');
var contracts = require('./contracts');
var utils = require('./utils');
var tokenRegistryV4 = require('@tradetrust-tt/token-registry-v4');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var contracts__namespace = /*#__PURE__*/_interopNamespace(contracts);



Object.defineProperty(exports, "v4RoleHash", {
  enumerable: true,
  get: function () { return roleHash.roleHash; }
});
Object.defineProperty(exports, "v4ContractAddress", {
  enumerable: true,
  get: function () { return contractAddress.contractAddress; }
});
Object.defineProperty(exports, "v4SupportInterfaceIds", {
  enumerable: true,
  get: function () { return supportInterfaceIds.supportInterfaceIds; }
});
exports.v4Contracts = contracts__namespace;
Object.defineProperty(exports, "v4ComputeInterfaceId", {
  enumerable: true,
  get: function () { return utils.computeInterfaceId; }
});
Object.defineProperty(exports, "v4ComputeTitleEscrowAddress", {
  enumerable: true,
  get: function () { return utils.computeTitleEscrowAddress; }
});
Object.defineProperty(exports, "v4EncodeInitParams", {
  enumerable: true,
  get: function () { return utils.encodeInitParams; }
});
Object.defineProperty(exports, "v4GetEventFromReceipt", {
  enumerable: true,
  get: function () { return utils.getEventFromReceipt; }
});
Object.defineProperty(exports, "constants", {
  enumerable: true,
  get: function () { return tokenRegistryV4.constants; }
});
Object.defineProperty(exports, "utils", {
  enumerable: true,
  get: function () { return tokenRegistryV4.utils; }
});
Object.defineProperty(exports, "v4Utils", {
  enumerable: true,
  get: function () { return tokenRegistryV4.utils; }
});
