// @tradetrust-tt/tt-verify/dist/cjs/verifiers/documentStatus/utils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRevokedByOcspResponder = exports.isRevokedOnDocumentStore = exports.isAnyHashRevoked = exports.decodeError = exports.getIntermediateHashes = void 0;
const tradetrust_1 = require("@tradetrust-tt/tradetrust");
const document_store_1 = require("@tradetrust-tt/document-store");
const ethers_1 = require("ethers");
const error_1 = require("../../types/error");
const error_2 = require("../../common/error");
const revocation_types_1 = require("./revocation.types");
const didSignedDocumentStatus_type_1 = require("./didSigned/didSignedDocumentStatus.type");
const getIntermediateHashes = (targetHash, proofs = []) => {
    const hashes = [`0x${targetHash}`];
    proofs.reduce((prev, curr) => {
        const next = tradetrust_1.utils.combineHashString(prev, curr);
        hashes.push(`0x${next}`);
        return next;
    }, targetHash);
    return hashes;
};
exports.getIntermediateHashes = getIntermediateHashes;
/**
 * Try to decode the error to see if we can deterministically tell if the document has NOT been issued or revoked.
 *
 * In case where we cannot tell, we throw an error
 * */
const decodeError = (error) => {
    var _a, _b, _c, _d;
    const reason = error.reason && Array.isArray(error.reason) ? error.reason[0] : (_a = error.reason) !== null && _a !== void 0 ? _a : "";
    switch (true) {
        case !error.reason &&
            (((_b = error.method) === null || _b === void 0 ? void 0 : _b.toLowerCase()) === "isRevoked(bytes32)".toLowerCase() ||
                ((_c = error.method) === null || _c === void 0 ? void 0 : _c.toLowerCase()) === "isIssued(bytes32)".toLowerCase()) &&
            error.code === ethers_1.errors.CALL_EXCEPTION:
            return "Contract is not found";
        case reason.toLowerCase() === "ENS name not configured".toLowerCase() &&
            error.code === ethers_1.errors.UNSUPPORTED_OPERATION:
            return "ENS name is not configured";
        case reason.toLowerCase() === "bad address checksum".toLowerCase() && error.code === ethers_1.errors.INVALID_ARGUMENT:
            return "Bad document store address checksum";
        case ((_d = error.message) === null || _d === void 0 ? void 0 : _d.toLowerCase()) === "name not found".toLowerCase():
            return "ENS name is not found";
        case reason.toLowerCase() === "invalid address".toLowerCase() && error.code === ethers_1.errors.INVALID_ARGUMENT:
            return "Invalid document store address";
        case error.code === ethers_1.errors.INVALID_ARGUMENT:
            return "Invalid call arguments";
        case error.code === ethers_1.errors.SERVER_ERROR:
            throw new error_2.CodedError("Unable to connect to the Ethereum network, please try again later", error_1.OpenAttestationEthereumDocumentStoreStatusCode.SERVER_ERROR, error_1.OpenAttestationEthereumDocumentStoreStatusCode[error_1.OpenAttestationEthereumDocumentStoreStatusCode.SERVER_ERROR]);
        default:
            throw error;
    }
};
exports.decodeError = decodeError;
/**
 * Given a list of hashes, check against one smart contract if any of the hash has been revoked
 * */
const isAnyHashRevoked = (smartContract, intermediateHashes) => __awaiter(void 0, void 0, void 0, function* () {
    const revokedStatusDeferred = intermediateHashes.map((hash) => smartContract.isRevoked(hash).then((status) => (status ? hash : undefined)));
    const revokedStatuses = yield Promise.all(revokedStatusDeferred);
    return revokedStatuses.find((hash) => hash);
});
exports.isAnyHashRevoked = isAnyHashRevoked;
const isRevokedOnDocumentStore = ({ documentStore, merkleRoot, provider, targetHash, proofs, }) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const documentStoreContract = yield document_store_1.DocumentStoreFactory.connect(documentStore, provider);
        const intermediateHashes = (0, exports.getIntermediateHashes)(targetHash, proofs);
        const revokedHash = yield (0, exports.isAnyHashRevoked)(documentStoreContract, intermediateHashes);
        return revokedHash
            ? {
                revoked: true,
                address: documentStore,
                reason: {
                    message: `Document ${merkleRoot} has been revoked under contract ${documentStore}`,
                    code: error_1.OpenAttestationEthereumDocumentStoreStatusCode.DOCUMENT_REVOKED,
                    codeString: error_1.OpenAttestationEthereumDocumentStoreStatusCode[error_1.OpenAttestationEthereumDocumentStoreStatusCode.DOCUMENT_REVOKED],
                },
            }
            : {
                revoked: false,
                address: documentStore,
            };
    }
    catch (error) {
        // If error can be decoded and it's because of document is not revoked, we return false
        // Else allow error to continue to bubble up
        return {
            revoked: true,
            address: documentStore,
            reason: {
                message: (0, exports.decodeError)(error),
                code: error_1.OpenAttestationEthereumDocumentStoreStatusCode.DOCUMENT_REVOKED,
                codeString: error_1.OpenAttestationEthereumDocumentStoreStatusCode[error_1.OpenAttestationEthereumDocumentStoreStatusCode.DOCUMENT_REVOKED],
            },
        };
    }
});
exports.isRevokedOnDocumentStore = isRevokedOnDocumentStore;
const isRevokedByOcspResponder = ({ merkleRoot, targetHash, proofs, location, }) => __awaiter(void 0, void 0, void 0, function* () {
    const intermediateHashes = (0, exports.getIntermediateHashes)(targetHash, proofs);
    for (const hash of intermediateHashes) {
        const res = yield fetch(`${location}/${hash}`).catch((e) => {
            throw new error_2.CodedError(`Invalid or unexpected response from OCSP Responder - ${e}`, error_1.OpenAttestationDidSignedDocumentStatusCode.OCSP_RESPONSE_INVALID, "OCSP_RESPONSE_INVALID");
        });
        const data = yield res.json();
        if (didSignedDocumentStatus_type_1.ValidOcspResponse.guard(data)) {
            continue;
        }
        else if (didSignedDocumentStatus_type_1.ValidOcspResponseRevoked.guard(data)) {
            const { reasonCode } = data;
            return {
                revoked: true,
                address: location,
                reason: {
                    message: `Document ${merkleRoot} has been revoked under OCSP Responder: ${location}`,
                    code: reasonCode,
                    codeString: revocation_types_1.OcspResponderRevocationReason[reasonCode],
                },
            };
        }
        else {
            throw new error_2.CodedError(`Invalid or unexpected response from OCSP Responder`, error_1.OpenAttestationDidSignedDocumentStatusCode.OCSP_RESPONSE_INVALID, "OCSP_RESPONSE_INVALID");
        }
    }
    return {
        revoked: false,
        address: location,
    };
});
exports.isRevokedByOcspResponder = isRevokedByOcspResponder;
