// @tradetrust-tt/tt-verify/dist/cjs/verifiers/issuerIdentity/dnsTxt/openAttestationDnsTxt.type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DnsTxtVerificationStatusDataV3 = exports.InvalidDnsTxtVerificationStatusDataV3 = exports.ValidDnsTxtVerificationStatusDataV3 = exports.DnsTxtVerificationStatusArray = exports.DnsTxtVerificationStatus = exports.InvalidDnsTxtVerificationStatus = exports.ValidDnsTxtVerificationStatusArray = exports.ValidDnsTxtVerificationStatus = void 0;
const runtypes_1 = require("runtypes");
const error_1 = require("../../../types/error");
/**
 * DNS-TXT verification status
 */
exports.ValidDnsTxtVerificationStatus = (0, runtypes_1.Record)({
    status: (0, runtypes_1.Literal)("VALID"),
    location: runtypes_1.String,
    value: runtypes_1.String,
});
exports.ValidDnsTxtVerificationStatusArray = (0, runtypes_1.Array)(exports.ValidDnsTxtVerificationStatus).withConstraint((elements) => elements.length > 0 || "Expect at least one valid element");
exports.InvalidDnsTxtVerificationStatus = (0, runtypes_1.Record)({
    status: (0, runtypes_1.Literal)("INVALID"),
    location: (0, runtypes_1.Optional)(runtypes_1.String),
    value: (0, runtypes_1.Optional)(runtypes_1.String),
    reason: error_1.Reason,
});
exports.DnsTxtVerificationStatus = (0, runtypes_1.Union)(exports.ValidDnsTxtVerificationStatus, exports.InvalidDnsTxtVerificationStatus);
exports.DnsTxtVerificationStatusArray = (0, runtypes_1.Array)(exports.DnsTxtVerificationStatus);
/**
 * Data for v3 Fragments
 */
exports.ValidDnsTxtVerificationStatusDataV3 = (0, runtypes_1.Record)({
    identifier: runtypes_1.String,
    value: runtypes_1.String,
});
// by design runtypes will validate arrays when an object has only partial properties
// https://github.com/pelotom/runtypes/issues/32
exports.InvalidDnsTxtVerificationStatusDataV3 = exports.ValidDnsTxtVerificationStatusDataV3.asPartial().withConstraint((value) => !Array.isArray(value) || "can't be an array");
exports.DnsTxtVerificationStatusDataV3 = (0, runtypes_1.Union)(exports.ValidDnsTxtVerificationStatusDataV3, exports.InvalidDnsTxtVerificationStatusDataV3);
