// @tradetrust-tt/tt-verify/dist/cjs/verifiers/issuerIdentity/dnsTxt/openAttestationDnsTxt.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.openAttestationDnsTxtIdentityProof = void 0;
const tradetrust_1 = require("@tradetrust-tt/tradetrust");
const dnsprove_1 = require("@tradetrust-tt/dnsprove");
const error_1 = require("../../../types/error");
const errorHandler_1 = require("../../../common/errorHandler");
const error_2 = require("../../../common/error");
const openAttestationDnsTxt_type_1 = require("./openAttestationDnsTxt.type");
const name = "OpenAttestationDnsTxtIdentityProof";
const type = "ISSUER_IDENTITY";
// Resolve identity of an issuer, currently supporting only DNS-TXT
// DNS-TXT is explained => https://github.com/Open-Attestation/adr/blob/master/decentralized_identity_proof_DNS-TXT.md
const resolveIssuerIdentity = (location, smartContractAddress, options) => __awaiter(void 0, void 0, void 0, function* () {
    const network = yield options.provider.getNetwork();
    const records = yield (0, dnsprove_1.getDocumentStoreRecords)(location);
    const matchingRecord = records.find((record) => record.addr.toLowerCase() === smartContractAddress.toLowerCase() &&
        record.netId === network.chainId.toString(10) &&
        record.type === "openatts" &&
        record.net === "ethereum");
    return matchingRecord
        ? {
            status: "VALID",
            location,
            value: smartContractAddress,
        }
        : {
            status: "INVALID",
            location,
            value: smartContractAddress,
            reason: {
                message: `Matching DNS record not found for ${smartContractAddress}`,
                code: error_1.OpenAttestationDnsTxtCode.MATCHING_RECORD_NOT_FOUND,
                codeString: error_1.OpenAttestationDnsTxtCode[error_1.OpenAttestationDnsTxtCode.MATCHING_RECORD_NOT_FOUND],
            },
        };
});
const skip = () => __awaiter(void 0, void 0, void 0, function* () {
    return {
        status: "SKIPPED",
        type,
        name,
        reason: {
            code: error_1.OpenAttestationDnsTxtCode.SKIPPED,
            codeString: error_1.OpenAttestationDnsTxtCode[error_1.OpenAttestationDnsTxtCode.SKIPPED],
            message: `Document issuers doesn't have "documentStore" / "tokenRegistry" property or doesn't use ${tradetrust_1.v3.IdentityProofType.DNSTxt} type`,
        },
    };
});
const test = (document) => {
    if (tradetrust_1.utils.isWrappedV2Document(document)) {
        const documentData = (0, tradetrust_1.getData)(document);
        // at least one issuer uses DNS-TXT
        return documentData.issuers.some((issuer) => {
            var _a;
            return ((issuer.documentStore || issuer.tokenRegistry || issuer.certificateStore) &&
                ((_a = issuer.identityProof) === null || _a === void 0 ? void 0 : _a.type) === tradetrust_1.v2.IdentityProofType.DNSTxt);
        });
    }
    else if (tradetrust_1.utils.isWrappedV3Document(document)) {
        return document.openAttestationMetadata.identityProof.type === tradetrust_1.v3.IdentityProofType.DNSTxt;
    }
    return false;
};
const verifyV2 = (document, options) => __awaiter(void 0, void 0, void 0, function* () {
    const documentData = (0, tradetrust_1.getData)(document);
    const identities = yield Promise.all(documentData.issuers.map((issuer) => {
        var _a;
        if (((_a = issuer.identityProof) === null || _a === void 0 ? void 0 : _a.type) === tradetrust_1.v2.IdentityProofType.DNSTxt) {
            const { location } = issuer.identityProof;
            const smartContractAddress = issuer.documentStore || issuer.tokenRegistry || issuer.certificateStore;
            if (!location)
                throw new error_2.CodedError("Location not found in identity proof", error_1.OpenAttestationDnsTxtCode.INVALID_ISSUERS, error_1.OpenAttestationDnsTxtCode[error_1.OpenAttestationDnsTxtCode.INVALID_ISSUERS]);
            if (!smartContractAddress)
                throw new error_2.CodedError("Smart contract address not found in identity proof", error_1.OpenAttestationDnsTxtCode.INVALID_ISSUERS, error_1.OpenAttestationDnsTxtCode[error_1.OpenAttestationDnsTxtCode.INVALID_ISSUERS]);
            return resolveIssuerIdentity(location, smartContractAddress, options);
        }
        const invalidResponse = {
            status: "INVALID",
            reason: {
                message: "Issuer is not using DNS-TXT identityProof type",
                code: error_1.OpenAttestationDnsTxtCode.INVALID_ISSUERS,
                codeString: error_1.OpenAttestationDnsTxtCode[error_1.OpenAttestationDnsTxtCode.INVALID_ISSUERS],
            },
        };
        return invalidResponse; // eslint is happy, so am I (https://github.com/bradzacher/eslint-plugin-typescript/blob/master/docs/rules/no-object-literal-type-assertion.md)
    }));
    if (openAttestationDnsTxt_type_1.ValidDnsTxtVerificationStatusArray.guard(identities)) {
        return {
            name,
            type,
            data: identities,
            status: "VALID",
        };
    }
    const invalidIdentity = identities.find(openAttestationDnsTxt_type_1.InvalidDnsTxtVerificationStatus.guard);
    if (openAttestationDnsTxt_type_1.InvalidDnsTxtVerificationStatus.guard(invalidIdentity)) {
        return {
            name,
            type,
            data: identities,
            reason: invalidIdentity.reason,
            status: "INVALID",
        };
    }
    throw new error_2.CodedError("Unable to retrieve the reason of the failure", error_1.OpenAttestationDnsTxtCode.UNEXPECTED_ERROR, "UNEXPECTED_ERROR");
});
const verifyV3 = (document, options) => __awaiter(void 0, void 0, void 0, function* () {
    if (document.openAttestationMetadata.proof.method !== tradetrust_1.v3.Method.DocumentStore &&
        document.openAttestationMetadata.proof.method !== tradetrust_1.v3.Method.TokenRegistry)
        throw new error_2.CodedError("DNS-TXT is only supported with documents issued using document store or token registry", error_1.OpenAttestationDnsTxtCode.UNSUPPORTED, error_1.OpenAttestationDnsTxtCode[error_1.OpenAttestationDnsTxtCode.UNSUPPORTED]);
    const smartContractAddress = document.openAttestationMetadata.proof.value;
    const { identifier } = document.openAttestationMetadata.identityProof;
    const issuerIdentity = yield resolveIssuerIdentity(identifier, smartContractAddress, options);
    if (openAttestationDnsTxt_type_1.ValidDnsTxtVerificationStatus.guard(issuerIdentity)) {
        return {
            name,
            type,
            data: {
                identifier: issuerIdentity.location,
                value: issuerIdentity.value,
            },
            status: "VALID",
        };
    }
    return {
        name,
        type,
        data: {
            identifier: issuerIdentity.location,
            value: issuerIdentity.value,
        },
        reason: issuerIdentity.reason,
        status: "INVALID",
    };
});
exports.openAttestationDnsTxtIdentityProof = {
    skip,
    test,
    verify: (0, errorHandler_1.withCodedErrorHandler)((document, options) => __awaiter(void 0, void 0, void 0, function* () {
        if (tradetrust_1.utils.isWrappedV2Document(document))
            return verifyV2(document, options);
        else if (tradetrust_1.utils.isWrappedV3Document(document))
            return verifyV3(document, options);
        // this code is actually unreachable because of the test function
        throw new error_2.CodedError("Document does not match either v2 or v3 formats", error_1.OpenAttestationDnsTxtCode.UNRECOGNIZED_DOCUMENT, error_1.OpenAttestationDnsTxtCode[error_1.OpenAttestationDnsTxtCode.UNRECOGNIZED_DOCUMENT]);
    }), {
        name,
        type,
        unexpectedErrorCode: error_1.OpenAttestationDnsTxtCode.UNEXPECTED_ERROR,
        unexpectedErrorString: error_1.OpenAttestationDnsTxtCode[error_1.OpenAttestationDnsTxtCode.UNEXPECTED_ERROR],
    }),
};
